---
title: "22q11"
output: html_notebook
---


```{r}
library(Seurat)
library(dittoSeq)
library(harmony)
library(SeuratWrappers)
library(scCustomize)
library(presto)
library(viridis)
library(stringr)
library(FastCAR)
library(Matrix)
library(DropletQC)
library(SingleCellExperiment)
library(dreamlet)
library(plot1cell)
library(igraph)
library(JASPAR2022)
library(motifmatchr)
library(TFBSTools)
library(EnsDb.Hsapiens.v86)
library(GenomicRanges)
library(ggplot2)
library(hdWGCNA)
library(ggrepel)
library(enrichR)
library(GeneOverlap)
library(dplyr)


ReadSTAR <- function(data.dir, vireo,  proj = "SeuratObject", feats = 1000, 
                     ratio=.1, counts = 1000, sample="Sample") {
  data.dir.genefull = paste0(data.dir, "GeneFull_Ex50pAS/raw")
  mtx = file.path(data.dir.genefull, "UniqueAndMult-EM.mtx")
  cells = file.path(data.dir.genefull, "barcodes.tsv")
  features = file.path(data.dir.genefull, "features.tsv")
  raw = ReadMtx(mtx = mtx, cells = cells, features = features)
  
  seurat = CreateSeuratObject(raw, project=proj, min.features = feats)
  
  data.dir.velo = paste0(data.dir, "/Velocyto/raw")
  cells = file.path(data.dir.velo, "barcodes.tsv")
  features = file.path(data.dir.velo, "features.tsv")
  spliced = file.path(data.dir.velo, "spliced.mtx")
  unspliced = file.path(data.dir.velo, "unspliced.mtx")
  ambi = file.path(data.dir.velo, "ambiguous.mtx")
  spliced = ReadMtx(mtx = spliced, cells = cells, features = features)
  unspliced = ReadMtx(mtx = unspliced, cells = cells, features = features)
  ambi = ReadMtx(mtx = ambi, cells = cells, features = features)
  spliced = spliced + ambi

  seurat[['unspliced']] = CreateAssayObject(unspliced[,colnames(seurat)])
  seurat[['spliced']] = CreateAssayObject(spliced[,colnames(seurat)])
  seurat$Sample = sample
  
  vir = read.table(paste0(vireo, "donor_ids.tsv"), sep='\t', 
                     header=T, row.names = 1)
  cbc = intersect(rownames(vir), colnames(seurat))
  seurat = subset(seurat, cells=cbc)
  vir = vir[cbc,]
  seurat@meta.data = cbind(seurat@meta.data, vir)
  seurat = subset(seurat, subset=donor_id != "unassigned")

  ambProfile = describe.ambient.RNA.sequence(fullMatrix = raw, 
                                             start = 10, 
                                             stop = 200, 
                                             by = 10, 
                                             contaminationChanceCutoff = 0.05)
  emptyDropletCutoff = recommend.empty.cutoff(ambProfile)
  ambientProfile = determine.background.to.remove(raw, emptyDropletCutoff, .05)
  seurat[['cleaned']] = CreateAssayObject(remove.background(seurat@assays$RNA@counts, ambientProfile))
  seurat[['percent.mt']] = PercentageFeatureSet(seurat, pattern = "MT-",assay = 'cleaned')
  seurat = subset(seurat, subset=donor_id != "doublet" & percent.mt <= 20)

  return(seurat)
}

org = ReadSTAR("/media/chang/HDD-3/chang/david/DS_org/org_cr/outs/org_Solo.out/",
               "/media/chang/HDD-3/chang/david/DS_org/org_cr/outs/vireo/", 
               proj = "org")

org$donor_id[org$donor_id == "donor0"] = "540135"
org$donor_id[org$donor_id == "donor1"] = "6OC2"
org$donor_id[org$donor_id == "donor2"] = "13234"

org1 = ReadSTAR("/media/chang/HDD-3/chang/david/org_1_2/org_1/outs/DS_org_1_Solo.out/",              "/media/chang/HDD-3/chang/david/org_1_2/org_1/outs/vireo/",
                proj = 'org1')

org1$donor_id[org1$donor_id == "donor0"] = "35773"
org1$donor_id[org1$donor_id == "donor1"] = "28126"
org1$donor_id[org1$donor_id == "donor2"] = "28115"
org1$donor_id[org1$donor_id == "donor3"] = "540135"

org2 = ReadSTAR("/media/chang/HDD-3/chang/david/org_1_2/org_2/outs/DS_org_2_Solo.out/",              "/media/chang/HDD-3/chang/david/org_1_2/org_2/outs/vireo/",
                proj = 'org2')

org2$donor_id[org2$donor_id == "donor0"] = "35773"
org2$donor_id[org2$donor_id == "donor1"] = "28115"
org2$donor_id[org2$donor_id == "donor2"] = "540135"
org2$donor_id[org2$donor_id == "donor3"] = "28126"

org3 = ReadSTAR("/media/chang/HDD-3/chang/david/org3/org3_Solo.out/",              "/media/chang/HDD-3/chang/david/org3/vireo_5/", proj = 'org3')

org3$donor_id[org3$donor_id == "donor0"] = "35773"
org3$donor_id[org3$donor_id == "donor1"] = "540135"
org3$donor_id[org3$donor_id == "donor2"] = "13234"
org3$donor_id[org3$donor_id == "donor3"] = "6OC2"
org3$donor_id[org3$donor_id == "donor4"] = "28126"

org_apr = ReadSTAR("/media/chang/HDD-3/chang/david/DSTJN01_10XThalOrgWK10_Apr2021/ThalOrgWk10_Apr2021_Solo.out/", "/media/chang/HDD-3/chang/david/DSTJN01_10XThalOrgWK10_Apr2021/vireo/", proj = 'org_apr')

org_apr$donor_id[org_apr$donor_id == "donor0"] = "WTC11"
org_apr$donor_id[org_apr$donor_id == "donor1"] = "28126"
org_apr$donor_id[org_apr$donor_id == "donor2"] = "6OC2"

ab2 = ReadSTAR("/media/chang/HDD-3/chang/david/latest_org/DSTJN01/ThalOrgWithAB2/ThalOrgWithAB2_Solo.out/","/media/chang/HDD-3/chang/david/latest_org/DSTJN01/ThalOrgWithAB2/vireo/", proj='ThalOrgWithAB2')

ab2$donor_id[ab2$donor_id == "donor0"] = "X"
ab2$donor_id[ab2$donor_id == "donor1"] = "AB2"
ab2$donor_id[ab2$donor_id == "donor2"] = "540135"
ab2$donor_id[ab2$donor_id == "donor3"] = "Y"
ab2$donor_id[ab2$donor_id == "donor4"] = "35773"
ab2$donor_id[ab2$donor_id == "donor5"] = "Z"
ab2$donor_id[ab2$donor_id == "donor6"] = "6OC2"
ab2$donor_id[ab2$donor_id == "donor7"] = "13234"

wtc10 = ReadSTAR("/media/chang/HDD-3/chang/david/latest_org/DSTJN01/ThalOrgWithWTC10/ThalOrgWithWTC10_Solo.out/","/media/chang/HDD-3/chang/david/latest_org/DSTJN01/ThalOrgWithWTC10/vireo/", proj='ThalOrgWithWTC10')

wtc10$donor_id[wtc10$donor_id == "donor0"] = "540135"
wtc10$donor_id[wtc10$donor_id == "donor1"] = "35773"
wtc10$donor_id[wtc10$donor_id == "donor2"] = "Y"
wtc10$donor_id[wtc10$donor_id == "donor3"] = "X"
wtc10$donor_id[wtc10$donor_id == "donor4"] = "Z"
wtc10$donor_id[wtc10$donor_id == "donor5"] = "6OC2"
wtc10$donor_id[wtc10$donor_id == "donor6"] = "13234"

wtc11 = ReadSTAR("/media/chang/HDD-3/chang/david/latest_org/DSTJN01/ThalOrgWithWTC11/ThalOrgWithWTC11_Solo.out/","/media/chang/HDD-3/chang/david/latest_org/DSTJN01/ThalOrgWithWTC11/vireo/", proj='ThalOrgWithWTC11')

wtc11$donor_id[wtc11$donor_id == "donor0"] = "6OC2"
wtc11$donor_id[wtc11$donor_id == "donor1"] = "540135"
wtc11$donor_id[wtc11$donor_id == "donor2"] = "WTC11"
wtc11$donor_id[wtc11$donor_id == "donor3"] = "X"
wtc11$donor_id[wtc11$donor_id == "donor4"] = "Y"
wtc11$donor_id[wtc11$donor_id == "donor5"] = "13234"
wtc11$donor_id[wtc11$donor_id == "donor6"] = "35773"
wtc11$donor_id[wtc11$donor_id == "donor7"] = "Z"

lane1 = ReadSTAR("/media/chang/HDD-3/chang/david/DS_org_fastq/DS_Lane1/DS_Lane1_Solo.out/", "/media/chang/HDD-3/chang/david/DS_org_fastq/DS_Lane1/vireo/", proj="Lane1")

lane1$donor_id[lane1$donor_id == "donor0"] = "28126"
lane1$donor_id[lane1$donor_id == "donor1"] = "13234"
lane1$donor_id[lane1$donor_id == "donor2"] = "35773"
lane1$donor_id[lane1$donor_id == "donor3"] = "6OC2"

lane2 = ReadSTAR("/media/chang/HDD-3/chang/david/DS_org_fastq/DS_Lane2/DS_Lane2_Solo.out/", "/media/chang/HDD-3/chang/david/DS_org_fastq/DS_Lane2/vireo/", proj="Lane2")

lane2$donor_id[lane2$donor_id == "donor0"] = "28126"
lane2$donor_id[lane2$donor_id == "donor1"] = "6OC2"
lane2$donor_id[lane2$donor_id == "donor2"] = "35773"
lane2$donor_id[lane2$donor_id == "donor3"] = "13234"

lane3 = ReadSTAR("/media/chang/HDD-3/chang/david/DS_org_fastq/DS_Lane3/DS_Lane3_Solo.out/", "/media/chang/HDD-3/chang/david/DS_org_fastq/DS_Lane3/vireo/", proj="Lane3")

lane3$donor_id[lane3$donor_id == "donor0"] = "13234"
lane3$donor_id[lane3$donor_id == "donor1"] = "28126"
lane3$donor_id[lane3$donor_id == "donor2"] = "35773"
lane3$donor_id[lane3$donor_id == "donor3"] = "6OC2"

lane4 = ReadSTAR("/media/chang/HDD-3/chang/david/DS_org_fastq/DS_Lane4/DS_Lane4_Solo.out/", "/media/chang/HDD-3/chang/david/DS_org_fastq/DS_Lane4/vireo/", proj="Lane4")

lane4$donor_id[lane4$donor_id == "donor0"] = "540135"
lane4$donor_id[lane4$donor_id == "donor1"] = "13234"
lane4$donor_id[lane4$donor_id == "donor2"] = "6OC2"
lane4$donor_id[lane4$donor_id == "donor3"] = "28126"
lane4$donor_id[lane4$donor_id == "donor4"] = "35773"

lane5 = ReadSTAR("/media/chang/HDD-3/chang/david/DS_org_fastq/DS_Lane5/DS_Lane5_Solo.out/", "/media/chang/HDD-3/chang/david/DS_org_fastq/DS_Lane5/vireo/", proj="Lane5")

lane5$donor_id[lane5$donor_id == "donor0"] = "28126"
lane5$donor_id[lane5$donor_id == "donor1"] = "540135"
lane5$donor_id[lane5$donor_id == "donor2"] = "6OC2"
lane5$donor_id[lane5$donor_id == "donor3"] = "35773"
lane5$donor_id[lane5$donor_id == "donor4"] = "13234"

lane6 = ReadSTAR("/media/chang/HDD-3/chang/david/DS_org_fastq/DS_Lane6/DS_Lane6_Solo.out/", "/media/chang/HDD-3/chang/david/DS_org_fastq/DS_Lane6/vireo/", proj="Lane6")

lane6$donor_id[lane6$donor_id == "donor0"] = "6OC2"
lane6$donor_id[lane6$donor_id == "donor1"] = "28126"
lane6$donor_id[lane6$donor_id == "donor2"] = "13234"
lane6$donor_id[lane6$donor_id == "donor3"] = "540135"
lane6$donor_id[lane6$donor_id == "donor4"] = "35773"

lane7 = ReadSTAR("/media/chang/HDD-3/chang/david/DS_org_fastq/DS_Lane7/DS_Lane7_Solo.out/", "/media/chang/HDD-3/chang/david/DS_org_fastq/DS_Lane7/vireo/", proj="Lane7")

lane7$donor_id[lane7$donor_id == "donor0"] = "6OC2"
lane7$donor_id[lane7$donor_id == "donor1"] = "540135"
lane7$donor_id[lane7$donor_id == "donor2"] = "28126"
lane7$donor_id[lane7$donor_id == "donor3"] = "13234"
lane7$donor_id[lane7$donor_id == "donor4"] = "35773"

org_list = list(org, org_apr, org1, org2, org3, ab2, wtc10, wtc11, lane1, lane2,
                lane3, lane4, lane5, lane6, lane7)

saveRDS(org_list, file='/media/chang/HDD-3/chang/david/org.list.rds')

merged = merge(org_list[[1]], org_list[2:length(org_list)])

saveRDS(merged, file='/media/chang/HDD-3/chang/david/merged.rds')
```

```{r}
merged = readRDS("/media/chang/HDD-3/chang/david/merged.rds")

merged = SCTransform(merged, vst.flavor='v2', assay = "cleaned")
VariableFeatures(merged) = VariableFeatures(merged)[!grepl("MT-", VariableFeatures(merged))]
VariableFeatures(merged) = VariableFeatures(merged)[!grepl("MALAT1", VariableFeatures(merged))]
VariableFeatures(merged) = VariableFeatures(merged)[!grepl("RPS", VariableFeatures(merged))]
VariableFeatures(merged) = VariableFeatures(merged)[!grepl("RPL", VariableFeatures(merged))]

merged = RunPCA(merged, npcs = 30)
merged = RunUMAP(merged, dims=1:30)

merged$Batch = merged$orig.ident
merged$Batch[grepl("Thal", merged$Batch)] = "Thal"
merged$Batch[grepl("org_apr", merged$Batch)] = "org_apr"
merged$Batch[grepl("org1", merged$Batch)] = "org1_2"
merged$Batch[grepl("org2", merged$Batch)] = "org1_2"
merged$Batch[grepl("org3", merged$Batch)] = "org3"
merged$Batch[grepl("org_cr", merged$Batch)] = "org_cr"
merged$Batch[grepl("Lane1", merged$Batch)] = "Lane1_3"
merged$Batch[grepl("Lane2", merged$Batch)] = "Lane1_3"
merged$Batch[grepl("Lane3", merged$Batch)] = "Lane1_3"
merged$Batch[grepl("Lane4", merged$Batch)] = "Lane4_7"
merged$Batch[grepl("Lane5", merged$Batch)] = "Lane4_7"
merged$Batch[grepl("Lane6", merged$Batch)] = "Lane4_7"
merged$Batch[grepl("Lane7", merged$Batch)] = "Lane4_7"
merged$Batch[merged$Batch == "org"] = "org"

merged$Condition = "CTRL"
merged$Condition[grepl("6OC2", merged$donor_id)] = "22Q"
merged$Condition[grepl("540135", merged$donor_id)] = "22Q"
merged$Condition[grepl("35773", merged$donor_id)] = "22Q"
merged$Condition[grepl("AB2", merged$donor_id)] = "22Q"

merged$donor_id[grepl("Y", merged$donor_id)] = "29098A"
merged$Condition[grepl("29098A", merged$donor_id)] = "CTRL"
merged$donor_id[grepl("Z", merged$donor_id)] = "721532"
merged$Condition[grepl("721532", merged$donor_id)] = "22Q"
merged$donor_id[grepl("X", merged$donor_id)] = "20961B"
merged$Condition[grepl("20961B", merged$donor_id)] = "CTRL"

merged = RunHarmony(merged, group.by.vars = "Batch", kmeans_init_nstart=100, kmeans_init_iter_max=10000, assay.use = "SCT", plot_convergence = T, theta=0)
merged = RunUMAP(merged, dims=1:30, reduction = 'harmony')
merged = FindNeighbors(merged, dims=1:30, reduction = 'harmony')
merged = FindClusters(merged, algorithm=2, resolution=.2)
merged = FindClusters(merged, algorithm=2, resolution=.3)
merged = FindClusters(merged, algorithm=2, resolution=.4)
merged = FindClusters(merged, algorithm=2, resolution=.5)

saveRDS(merged, file='/media/chang/HDD-3/chang/david/merged_proc.rds')

DimPlot(merged, group.by = c("donor_id",'orig.ident','Batch',"Condition"), 
        cols = dittoColors()) & NoAxes()

DimPlot(merged, group.by = c("SCT_snn_res.0.2",'SCT_snn_res.0.3',
                             "SCT_snn_res.0.4","SCT_snn_res.0.5"), 
        cols = dittoColors()) & NoAxes()

FeaturePlot(merged ,features = c("FOXG1","TCF7L2","SOX2","POU4F1","GBX2",
                                 "CRABP1","FOXP2","EN1","LHX1"),
            ncol = 3, order=T, max.cutoff = 'q99') & NoAxes()

SCpubr::do_CorrelationPlot(sample = merged,group.by = "SCT_snn_res.0.5" ,
                           cell_size = 10)

markers = wilcoxauc(merged, seurat_assay = "SCT")

merged$clusters = as.character(merged$SCT_snn_res.0.5)
merged$clusters[merged$clusters == 10] = "CP1"
merged$clusters[merged$clusters == 22] = "FB2"
merged$clusters[merged$clusters == 4] = "Cor_EN"
merged$clusters[merged$clusters == 21] = "Cor_dlEN"
merged$clusters[merged$clusters == 15] = "Cor_RG"
merged$clusters[merged$clusters == 17] = "Cor_Div"
merged$clusters[merged$clusters == 7] = "Thal_EN1"
merged$clusters[merged$clusters == 14] = "Thal_IPC"
merged$clusters[merged$clusters == 3] = "Thal_EN2"
merged$clusters[merged$clusters == 0] = "Thal_IN1"
merged$clusters[merged$clusters == 6] = "Thal_IN2"
merged$clusters[merged$clusters == 19] = "Thal_IN3"
merged$clusters[merged$clusters == 9] = "Thal_Div"
merged$clusters[merged$clusters == 20] = "FB1"
merged$clusters[merged$clusters == 11] = "AC1"
merged$clusters[merged$clusters == 8] = "AC2"
merged$clusters[merged$clusters == 1] = "Thal_GL1"
merged$clusters[merged$clusters == 16] = "Thal_GL1"
merged$clusters[merged$clusters == 18] = "Thal_GL1"
merged$clusters[merged$clusters == 2] = "Thal_GL2"
merged$clusters[merged$clusters == 5] = "Thal_GL3"
merged$clusters[merged$clusters == 13] = "Thal_GL3"
merged$clusters[merged$clusters == 12] = "CP2"

Idents(merged) = merged$clusters

DimPlot(merged, label = T, group.by = 'clusters', cols = dittoColors())
```

```{r}
table(merged$donor_id, merged$Batch)
merged = readRDS("/media/chang/HDD-3/chang/david/merged_proc.rds")

cortex = subset(merged, subset=Batch %in% c("org1_2", "org3", "Thal"))
cortex$orig.ident = paste0(cortex$Batch, "_", cortex$donor_id)

cortex = subset(cortex, subset=orig.ident %in% c(names(table(cortex$orig.ident)[1:9]), 
                                                 "Thal_29098A","Thal_WTC11","Thal_AB2"))
thal3 = subset(cortex, subset=orig.ident=="Thal_13234")
cortex = subset(cortex, subset=orig.ident != "Thal_13234")
saveRDS(cortex, file='/media/chang/HDD-3/chang/david/merged_cortex.rds')


thal = subset(merged, subset=Batch %in% c("Lane1_3", "Lane4_7","org","org_apr"))
thal2 = subset(cortex, subset=orig.ident %in% c("Thal_6OC2", "Thal_540135", "Thal_35773",
                                                 "Thal_721532","Thal_13234"))
thal = merge(thal, thal2, thal3)
saveRDS(thal, file='/media/chang/HDD-3/chang/david/merged_thal.rds')
```

```{r}
cortex = readRDS('/media/chang/HDD-3/chang/david/merged_cortex.rds')
cortex = SCTransform(cortex, vst.flavor='v2', assay = "cleaned")
VariableFeatures(cortex) = VariableFeatures(cortex)[!grepl("MT-", VariableFeatures(cortex))]
VariableFeatures(cortex) = VariableFeatures(cortex)[!grepl("MALAT1", VariableFeatures(cortex))]
VariableFeatures(cortex) = VariableFeatures(cortex)[!grepl("RPS", VariableFeatures(cortex))]
VariableFeatures(cortex) = VariableFeatures(cortex)[!grepl("RPL", VariableFeatures(cortex))]
cortex = RunPCA(cortex, npcs = 20)

saveRDS(cortex, file='/media/chang/HDD-3/chang/david/merged_cortex.rds')

DimPlot(cortex, group.by = c("donor_id","Batch"), cols=dittoColors()) & NoAxes()

cortex = RunUMAP(cortex, dims=1:20)
cortex = FindNeighbors(cortex, dims=1:20)

cortex = FindClusters(cortex, algorithm = 2, resolution = .3)

thal2 = subset(cortex, subset=SCT_snn_res.0.3 %in% c(0,1,4,6,10,13))
saveRDS(thal2, 
        file='/media/chang/HDD-3/chang/david/thal_from_cortical_org1_2_3_latest.rds')
cortex = subset(cortex, subset=SCT_snn_res.0.3 %in% c(2,3,5,7,8,9,11,12))

DimPlot(cortex, group.by = c("SCT_snn_res.0.4","Batch"), cols=dittoColors()) & NoAxes()
FeaturePlot(cortex ,features = c("FOXG1","TCF7L2","TTR","SOX2","MKI67","EOMES"),ncol = 3, order=T, max.cutoff = 'q99') & NoAxes()

cortex = SCTransform(cortex, vst.flavor='v2', assay = "cleaned")
VariableFeatures(cortex) = VariableFeatures(cortex)[!grepl("MT-", VariableFeatures(cortex))]
VariableFeatures(cortex) = VariableFeatures(cortex)[!grepl("MALAT1", VariableFeatures(cortex))]
VariableFeatures(cortex) = VariableFeatures(cortex)[!grepl("RPS", VariableFeatures(cortex))]
VariableFeatures(cortex) = VariableFeatures(cortex)[!grepl("RPL", VariableFeatures(cortex))]
cortex = RunPCA(cortex, npcs = 20)
cortex = RunHarmony(cortex, group.by.vars = 'Batch', assay.use = "SCT", 
                    theta = 1, plot_convergence = T)
cortex = RunUMAP(cortex, dims=1:20, reduction='harmony')
cortex = FindNeighbors(cortex, dims=1:20, reduction='harmony')
saveRDS(cortex, 
        file='/media/chang/HDD-3/chang/david/merged_cortex_filtered.rds')

FeaturePlot(cortex, features = c("FOXG1","AQP4","NEUROD6","SOX2","MKI67","EOMES","HOPX","FEZF2","SATB2",
                                 "CRYM","OPCML","NRP1"),
            ncol = 4, order=T, max.cutoff = 'q99') & NoAxes()

cortex = FindClusters(cortex, algorithm = 2, resolution = .3)

cortex.markers = FindAllMarkers(cortex, only.pos = T)
saveRDS(cortex.markers, file="/media/chang/HDD-3/chang/david/cortex_markers.rds")
cortex.markers %>%
    group_by(cluster) %>%
    top_n(n = 5, wt = avg_log2FC) -> top10
DoHeatmap(cortex, features = top10$gene) + NoLegend()

cortex$clusters = as.character(cortex$SCT_snn_res.0.3)
cortex$clusters[cortex$clusters == "0"] = "RG"
cortex$clusters[cortex$clusters == "7"] = "Div"
cortex$clusters[cortex$clusters == "8"] = "Div"
cortex$clusters[cortex$clusters == "5"] = "IPC"
cortex$clusters[cortex$clusters == "3"] = "ulEN"
cortex$clusters[cortex$clusters == "4"] = "dlEN"
cortex$clusters[cortex$clusters == "1"] = "EN"
cortex$clusters[cortex$clusters == "2"] = "nEN1"
cortex$clusters[cortex$clusters == "6"] = "nEN2"

Idents(cortex) = cortex$clusters
saveRDS(cortex, 
        file='/media/chang/HDD-3/chang/david/merged_cortex_filtered.rds')
DimPlot(cortex, group.by = 'clusters', label=T, cols=dittoColors())




thal = readRDS('/media/chang/HDD-3/chang/david/merged_thal.rds')
thal = SCTransform(thal, vst.flavor='v2', assay = "cleaned")
VariableFeatures(thal) = VariableFeatures(thal)[!grepl("MT-", VariableFeatures(thal))]
VariableFeatures(thal) = VariableFeatures(thal)[!grepl("MALAT1", VariableFeatures(thal))]
VariableFeatures(thal) = VariableFeatures(thal)[!grepl("RPS", VariableFeatures(thal))]
VariableFeatures(thal) = VariableFeatures(thal)[!grepl("RPL", VariableFeatures(thal))]
thal = RunPCA(thal, npcs = 30)
thal = RunHarmony(thal, group.by.vars = 'Batch', assay.use = "SCT", 
                    theta = 1, plot_convergence = T)
thal = RunUMAP(thal, dims=1:30, reduction = 'harmony')
thal = FindNeighbors(thal, dims=1:30, reduction = 'harmony')
saveRDS(thal, file='/media/chang/HDD-3/chang/david/merged_thal.rds')

thal = FindClusters(thal, algorithm = 2, resolution = .5)

thal$clusters = as.character(thal$SCT_snn_res.0.5)

thal$clusters[thal$clusters == 6] = "Div"
thal$clusters[thal$clusters == 23] = "Div"
thal$clusters[thal$clusters == 10] = "CP1"
thal$clusters[thal$clusters == 24] = "CP2"
thal$clusters[thal$clusters == 12] = "CP3"

thal$clusters[thal$clusters == 9] = "EN1"
thal$clusters[thal$clusters == 2] = "EN2"
thal$clusters[thal$clusters == 0] = "IN1"
thal$clusters[thal$clusters == 25] = "IN1"
thal$clusters[thal$clusters == 4] = "IN2"
thal$clusters[thal$clusters == 13] = "IN3"
thal$clusters[thal$clusters == 22] = "IN4"

thal$clusters[thal$clusters == 14] = "IPC1"
thal$clusters[thal$clusters == 26] = "IPC2"

thal$clusters[thal$clusters == 20] = "CTX"
thal$clusters[thal$clusters == 15] = "CTX"

thal$clusters[thal$clusters == 7] = "GL1"
thal$clusters[thal$clusters == 8] = "GL2"
thal$clusters[thal$clusters == 17] = "GL2"
thal$clusters[thal$clusters == 11] = "AC"
thal$clusters[thal$clusters == 1] = "GL3"
thal$clusters[thal$clusters == 3] = "GL4"

thal$clusters[thal$clusters == 5] = "GL5"
thal$clusters[thal$clusters == 16] = "GL5"

thal$clusters[thal$clusters == 19] = "FB"
thal$clusters[thal$clusters == 21] = "UNK"
thal$clusters[thal$clusters == 18] = "LQ"

Idents(thal) = thal$clusters

thal = subset(thal, subset=clusters %in% c("Div", "GL1","GL2","UNK","IN1",
                                        "IN2","AC","FB","LQ","IPC1","IPC2","EN1",
                                        "EN2","IN3","IN4","GL3","GL4","GL5","LQ"));gc();

thal = subset(thal, subset=donor_id %in% c("13234", "28126","35773","540135",
                                           "6OC2", "721532"));gc();

thal = SCTransform(thal, vst.flavor='v2', assay = "cleaned")
VariableFeatures(thal) = VariableFeatures(thal)[!grepl("MT-", VariableFeatures(thal))]
VariableFeatures(thal) = VariableFeatures(thal)[!grepl("MALAT1", VariableFeatures(thal))]
VariableFeatures(thal) = VariableFeatures(thal)[!grepl("RPS", VariableFeatures(thal))]
VariableFeatures(thal) = VariableFeatures(thal)[!grepl("RPL", VariableFeatures(thal))]
thal = RunPCA(thal, npcs = 30)
thal = RunHarmony(thal, group.by.vars = 'Batch', assay.use = "SCT", 
                    theta = 1, plot_convergence = T)
thal = RunUMAP(thal, dims=1:30, reduction = 'harmony')
thal = FindNeighbors(thal, dims=1:30, reduction = 'harmony')

saveRDS(thal,file="/media/chang/HDD-3/chang/david/merged_thal_filtered_cp.rds")


SCpubr::do_CorrelationPlot(sample = thal,group.by = "clusters" ,
                           cell_size = 10)

FeaturePlot(thal ,features = c("FOXG1","TCF7L2","EYA2","SOX2","MKI67","GAD1", "PAX6",
                               "LHX1","RNF220","GAD1","CRABP1","PAX3","EN1"),
            ncol = 4, order=T, max.cutoff = 'q99') & NoAxes()

FeaturePlot(thal ,features = c("AQP4","DCN","EYA2","FOXG1","HOPX","SPARCL1",
                               "OLIG1","PCDH15","LUM"),ncol = 3, 
            order=T, max.cutoff = 'q99') & NoAxes()

DimPlot(thal, group.by = c("donor_id","Batch"), cols=dittoColors()) & NoAxes()
DimPlot(thal, group.by = c("clusters"),
        label=T, cols=dittoColors()) & NoAxes()

dittoBarPlot(thal, var='Condition', group.by = 'clusters') + dittoBarPlot(thal, var='donor_id', group.by = 'clusters')


second = readRDS("/media/chang/HDD-7/chang/aparna/second.rds")

anchors <- FindTransferAnchors(
  reference = second,
  query = thal,
  recompute.residuals = T,
  normalization.method = "SCT",  ## SCT
  query.assay = 'SCT', ##
  reference.reduction = "harmony",
  reference.assay = "SCT", ## SCT
  features = VariableFeatures(second),
  dims = 1:30
)
second = RunHarmony(second, group.by.vars=c("Sample"), assay.use="SCT",
                    kmeans_init_nstart=20, kmeans_init_iter_max=100)
second <- RunUMAP(second, dims = 1:30, 
                  reduction = "harmony", 
                  return.model = TRUE)

second$class = second$clusters
second$class[grepl ("EN", second$class)] = 'Glut'
second$class[grepl ("IN", second$class)] = 'Gaba'
second$class[grepl ("GL", second$class)] = 'GL'

thal <- MapQuery(anchorset = anchors,
  query = thal,
  reference = second,
  refdata = list(celltype.l1 = "class",
    celltype.l2 = "clusters"),
  reference.reduction = "harmony",
  reduction.model = "umap")

gc()

saveRDS(thal, file="/media/chang/HDD-3/chang/david/thal_labelled.rds")
saveRDS(second, file="/media/chang/HDD-3/chang/david/second_labelled.rds")

DimPlot(second, label = T, cols = dittoColors()) +DimPlot(thal, cols=dittoColors(),reduction = 'ref.umap', group.by = "predicted.celltype.l1", label=T) & NoAxes()

FeaturePlot(thal, features = c("TTR","FOXG1"),max.cutoff = 'q99',
            order=T, reduction = 'ref.umap') & NoAxes()
```

```{r}
second = readRDS("/media/chang/HDD-3/chang/david/second_labelled.rds")
thal = readRDS("/media/chang/HDD-3/chang/david/thal_labelled.rds")

thal@reductions$umap_denovo <- thal@reductions$umap
thal@reductions$umap <- thal@reductions$ref.umap
Idents(thal) = thal$predicted.celltype.l1

circ_data <- prepare_circlize_data(thal, scale = 0.8 )
set.seed(1234)
cluster_colors<-dittoColors()[1:length(unique(thal$predicted.celltype.l1))]
group_colors<-dittoColors()[1:length(unique(thal$Condition))]
rep_colors<-dittoColors()[1:length(unique(thal$donor_id))]

###plot and save figures
pdf('~/22q_figs/label_transfer_circle.pdf', width = 6, height = 6)
plot_circlize(circ_data,do.label = T, pt.size = 0.01, col.use = cluster_colors ,
              bg.color = 'white', kde2d.n = 200, repel = T, label.cex = 0.6)
add_track(circ_data, group = "Condition", colors = group_colors, track_num = 2) 
add_track(circ_data, group = "donor_id",colors = rep_colors, track_num = 3) 
dev.off()

Idents(second) = second$class
circ_data <- prepare_circlize_data(second, scale = 0.8 )
set.seed(1234)
cluster_colors<-dittoColors()[1:length(unique(second$class))]
group_colors<-dittoColors()[1:length(unique(second$Sample))]
age_colors<-dittoColors()[1:length(unique(second$Age))]
area_colors<-dittoColors()[1:length(unique(second$Area))]

###plot and save figures
pdf('~/22q_figs/second_circle.pdf', width = 6, height = 6)
plot_circlize(circ_data,do.label = T, pt.size = 0.01, col.use = cluster_colors ,
              bg.color = 'white', kde2d.n = 200, repel = T, label.cex = 0.6)
add_track(circ_data, group = "Sample", colors = group_colors, track_num = 2) 
add_track(circ_data, group = "Age", colors = age_colors, track_num = 3) 
add_track(circ_data, group = "Area", colors = area_colors, track_num = 4) 
dev.off()


pdf('~/22q_figs/line_bar.pdf', width = 4, height = 6)
dittoBarPlot(thal, var='predicted.celltype.l1', group.by = 'donor_id')
dev.off()
#dittoBarPlot(thal, var='Condition', group.by = 'predicted.celltype.l1')

FeaturePlot(thal, 
            features = c("SLC17A6","SLC32A1","TCF7L2","GAD1",'GAD2',"EYA2"),
            ncol=3,max.cutoff = 'q99',order=T) &NoAxes()

thal@reductions$umap <- thal@reductions$umap_denovo


thal = RunHarmony(thal, group.by.vars = "Batch",assay.use = "SCT",
                  plot_convergence = T, theta=0)
thal = RunUMAP(thal, dims=1:30, reduction='harmony')
thal = FindNeighbors(thal, dims=1:30, reduction = 'harmony')
thal = FindClusters(thal, algorithm = 2, resolution = .3)

Idents(thal) = thal$SCT_snn_res.0.3
thal$clusters = as.character(thal$SCT_snn_res.0.3)
thal$clusters[thal$clusters == "7"] = "Div"
thal$clusters[thal$clusters == "14"] = "Div"
thal$clusters[thal$clusters == "12"] = "FB"
thal$clusters[thal$SCT_snn_res.0.3 == 1] = "GL1"
thal$clusters[thal$SCT_snn_res.0.3 == 2] = "GL2"
thal$clusters[thal$clusters == "5"] = "IPC2"
thal$clusters[thal$clusters == "6"] = "IPC1"
thal$clusters[thal$clusters == "4"] = "AC"
thal$clusters[thal$clusters == "9"] = "AC"
thal$clusters[thal$clusters == "18"] = "AC"
thal$clusters[thal$clusters == "0"] = "Gaba"
thal$clusters[thal$clusters == "15"] = "Gaba"
thal$clusters[thal$clusters == "17"] = "Gaba"
thal$clusters[thal$SCT_snn_res.0.3 == "3"] = "nGlut"
thal$clusters[thal$clusters == "8"] = "Glut"
thal$clusters[thal$clusters == "19"] = "nGlut"
thal$clusters[thal$SCT_snn_res.0.3 == "16"] = "nGlut"
thal$clusters[thal$SCT_snn_res.0.3 == 10] = "IPC3"
thal$clusters[thal$SCT_snn_res.0.5 == 20] = "GL3"
thal$clusters[thal$SCT_snn_res.0.3 == "13"] = "GL4"
thal$clusters[thal$SCT_snn_res.0.3 == "11"] = "GL4"
Idents(thal) = thal$clusters

saveRDS(thal, file="/media/chang/HDD-3/chang/david/thal_labelled.rds")


p1 = DimPlot(thal, group.by = c("clusters"),
        cols=dittoColors(), label=T,pt.size=.1) & NoAxes()
cells.located <- CellSelector(plot = p1)
thal$clusters[cells.located] = "Gaba"

test= wilcoxauc(thal, seurat_assay = "SCT")

DimPlot(thal, group.by = c("clusters",'SCT_snn_res.0.3','predicted.celltype.l1'),
        cols=dittoColors(), label=T) & NoAxes()

pdf("~/22q_figs/org_final_cluster.pdf")
DimPlot(thal, group.by = c("clusters"),
        cols=dittoColors(), label=T) & NoAxes()
dev.off()

pdf("~/22q_figs/org_feature.pdf", height = 4)
FeaturePlot(thal, features = c("SLC32A1","GAD1","SLC17A6","TCF7L2",
                               "GLI3","EYA2"), max.cutoff = 'q99', order=T, ncol=3)&NoAxes() & scale_color_viridis()
dev.off()

dittoBarPlot(thal, var="predicted.celltype.l1",group.by = "clusters")

Idents(thal) = thal$clusters
markers = FindAllMarkers(thal, only.pos = T)
saveRDS(markers, file='/media/chang/HDD-3/chang/david/markers.rds')


markers = markers[!grepl("MIR", markers$gene),]
markers = markers[!grepl("AL1", markers$gene),]
markers = markers[!grepl("AL3", markers$gene),]



markers %>%
    group_by(cluster) %>%
    top_n(n = 6, wt = avg_log2FC) -> top10

genes = c("SLC17A6", "LHX9", "LHX2", "GBX2", "GAD1", "LHX1", "LHX5", "MEIS2","STC2",
  "GLI3", "PAX6", "OTX2", "EYA2","MKI67","TOP2A","HSPA1A","HSPA6","PDK1",
  "COL1A1","DCN","SPARCL1","SSTR2","RELN","MGP",'HES6',"CRABP1","SOX3","CRYAB",
  "APOE","ATF5")

pdf("~/22q_figs/dotplot.pdf", width = 6)
Clustered_DotPlot(seurat_object = thal, features = genes, group.by = "clusters",
                  colors_use_idents = dittoColors(), x_lab_rotate = 90)
dev.off()



thal <- SetupForWGCNA(
  thal,
  features = VariableFeatures(thal),
  wgcna_name = "WGCNA"
)

thal <- MetacellsByGroups(
  seurat_obj = thal,
  group.by = c("clusters", "Condition"),
  k = 30,
  max_shared = 15 ,
  min_cells = 50,
  target_metacells= 1000,
  reduction = 'harmony',
  ident.group = 'clusters',
  slot = 'counts',
  assay = 'cleaned',
  wgcna_name="WGCNA"
)
thal <- NormalizeMetacells(thal)
thal <- SetDatExpr(
  thal,
  group_name = "Glut", 
  group.by='clusters', 
  assay = 'cleaned', 
  slot = 'data' # using normalized data
)
# Test different soft powers:
thal <- TestSoftPowers(
  thal,
  networkType = 'signed'
)

#saveRDS(thal, file="/media/chang/HDD-3/chang/david/wgcna.rds")
#thal = readRDS("/media/chang/HDD-3/chang/david/wgcna.rds")

plot_list <- PlotSoftPowers(thal)
print(wrap_plots(plot_list, ncol=2))

thal <- ConstructNetwork(
  thal, soft_power=6,
  setDatExpr=FALSE,overwrite_tom = T,
  tom_name = 'Glut' # name of the topoligical overlap matrix written to disk
)

PlotDendrogram(thal, main='Glut hdWGCNA Dendrogram')

thal <- ScaleData(thal, features=VariableFeatures(thal),
                           assay = "cleaned")
thal <- ModuleEigengenes(
 thal,assay = 'cleaned'#,group.by.vars='Batch'
)

MEs <- GetMEs(thal, harmonized=FALSE)

thal <- ModuleConnectivity(
  thal,harmonized = F,
  group.by = 'clusters', group_name = 'Glut'
)

thal <- ResetModuleNames(
  thal,
  new_name = "Glut-M"
)

modules <- GetModules(thal)
hub_df <- GetHubGenes(thal, n_hubs = 10)

p <- PlotKMEs(thal, ncol=5)

plot_list <- ModuleFeaturePlot(
  thal,
  features='MEs', # plot the hMEs
  order=TRUE # order so the points with highest MEs are on top
)
wrap_plots(plot_list, ncol=4)

saveRDS(thal, file='/media/chang/HDD-3/chang/david/wgcna.rds')



group1 <- thal@meta.data %>% subset(clusters == 'Glut' & Condition == "CTRL") %>% rownames
group2 <- thal@meta.data %>% subset(clusters == 'Glut' & Condition == "22Q") %>% rownames

DMEs <- FindDMEs(
  thal,harmonized = F,
  barcodes1 = group1,
  barcodes2 = group2,
  test.use='wilcox',
  wgcna_name='WGCNA'
)

PlotDMEsLollipop(
  thal, 
  DMEs, 
  wgcna_name='WGCNA', 
  pvalue = "p_val_adj"
)

PlotDMEsVolcano(
  thal,
  DMEs,
  wgcna_name = 'WGCNA'
)

DMEs_all <- FindAllDMEs(
  thal,harmonized = F,
  group.by = 'clusters',
  wgcna_name = 'WGCNA'
)

p <- PlotDMEsVolcano(
  thal,
  DMEs_all,
  wgcna_name = 'WGCNA',
  plot_labels=FALSE,
  show_cutoff=FALSE
)

# facet wrap by each cell type
p + facet_wrap(~group, ncol=4)



ModuleNetworkPlot(thal)

HubGeneNetworkPlot(
  thal, 
  n_hubs = 50, n_other=50,
  edge_prop = 0.75,
  mods = "Glut-M3"
)

thal <- RunModuleUMAP(
  thal,
  n_hubs = 25, # number of hub genes to include for the UMAP embedding
  n_neighbors=15, # neighbors parameter for UMAP
  min_dist=0.3 # min distance between points in UMAP space
)

ModuleUMAPPlot(
  thal,label_genes = c("FOXP2"),
  edge.alpha=0.25,
  sample_edges=TRUE,
  edge_prop=0.1, # proportion of edges to sample (20% here)
  label_hubs=2,# how many hub genes to plot per module?
  keep_grey_edges=FALSE
)




dbs <- c('GO_Biological_Process_2021','GO_Cellular_Component_2021','GO_Molecular_Function_2021')
# perform enrichment tests
thal <- RunEnrichr(
  thal,
  dbs=dbs, # character vector of enrichr databases to test
  max_genes = 100 # number of genes per module to test
)

EnrichrDotPlot(
  thal,
  mods = "all", # use all modules (this is the default behavior)
  database = "GO_Molecular_Function_2021", # this has to be one of the lists we used above!!!
  n_terms=1 # number of terms for each module
)
EnrichrBarPlot(
  thal,
  outdir = "enrichr_plots", # name of output directory
  n_terms = 10, # number of enriched terms to show (sometimes more show if there are ties!!!)
  plot_size = c(5,7), # width, height of the output .pdfs
  logscale=TRUE # do you want to show the enrichment as a log scale?
)
```

```{r}
pfm <- TFBSTools::getMatrixSet(
  x = JASPAR2022,
  opts = list(collection = "CORE", tax_group = 'vertebrates', all_versions = FALSE)
)


##################
EnsDb= EnsDb.Hsapiens.v86
  
  motif_df <- data.frame(
    motif_name = purrr::map(1:length(pfm), function(i){pfm[[i]]@name}) %>% unlist,
    motif_ID = purrr::map(1:length(pfm), function(i){pfm[[i]]@ID}) %>% unlist
  )

  # get promoters of protein coding genes from the given Ensdb
  # note: everything breaks if I try to use X & Y chromosomes.
  gene.promoters <- ensembldb::promoters(EnsDb, filter = ~ gene_biotype == "protein_coding") %>%
    subset(seqnames %in% c(1:100))
  gene.coords <- ensembldb::genes(EnsDb, filter = ~ gene_biotype == "protein_coding") %>%
    subset(seqnames %in% c(1:100))

  # add the gene name to the promoter object
  gene.promoters$symbol <- gene.coords$symbol[match(gene.promoters$gene_id, names(gene.coords))]

  # drop unnecessary chromosomes
  gene.promoters <- keepSeqlevels(gene.promoters, value= levels(droplevels(seqnames(gene.promoters))))

  # rename seqlevels to add 'chr', remove X&Y chromosomes because they break everything
  old_levels <- levels(seqnames(gene.promoters))
  new_levels <- ifelse(old_levels %in% c('X', 'Y'), old_levels, paste0('chr', old_levels))
  gene.promoters <- renameSeqlevels(gene.promoters, new_levels)

  # set the genome (not sure if we NEED to do this...)
  genome(seqinfo(gene.promoters)) <- 'hg38'

  # set up promoters object that only has the necessary info for motifmatchr
  my_promoters <- GRanges(
    seqnames =  droplevels(seqnames(gene.promoters)),
    IRanges(
      start = start(gene.promoters),
      end = end(gene.promoters)
    ),
    symbol = gene.promoters$symbol,
    genome="hg38"
  )

  # scan these promoters for motifs:
  print('Matching motifs...')
  motif_ix <- motifmatchr::matchMotifs(pfm, my_promoters, genome="hg38")

  # get the matches
  tf_match <- motifmatchr::motifMatches(motif_ix)
  rownames(tf_match) <- my_promoters$symbol

  # use motif names as the column names:
  motif_df <- motif_df[motif_df$motif_ID %in% colnames(tf_match), ] # removes motif IDs not present in tf_match
  colnames(tf_match) <- motif_df$motif_name

  # only keep genes that are in the Seurat object and in the given EnsDb:
  gene_list <- rownames(thal)
  gene_list <- gene_list[gene_list %in% rownames(tf_match)]
  tf_match <- tf_match[gene_list,]

  print('Getting TF target genes...')
  tfs <- motif_df$motif_name
  tf_targets <- list()
  n_targets <- list()
  for(cur_tf in tfs){
    tf_targets[[cur_tf]] <- names(tf_match[,cur_tf][tf_match[,cur_tf]])
    n_targets[[cur_tf]] <- length(tf_targets[[cur_tf]] )
  }
  n_targets <- unlist(n_targets)

  # add number of target genes to motif df
  motif_df = motif_df[!duplicated(motif_df$motif_name),]
  rownames(motif_df) = motif_df$motif_name
  motif_df = motif_df[names(n_targets),]
  motif_df$n_targets <- n_targets
  # add info to seurat object
  thal <- SetMotifMatrix(thal, tf_match)
  thal <- SetMotifs(thal, motif_df)
  thal <- SetMotifTargets(thal, tf_targets)
  thal <- SetPFMList(thal, pfm)
#############
  
  
target_genes <- GetMotifTargets(thal)
thal <- OverlapModulesMotifs(thal)

head(GetMotifOverlap(thal))

MotifOverlapBarPlot(
  thal,
  #motif_font = 'xkcd_regular',
  outdir = '~/motifs/',
  plot_size=c(5,6)
)


ModuleTFNetwork(
  thal,
  edge.alpha=0.75,
  cor_thresh = 0.75,
  tf_name = "FOXP2",
  tf_gene_name = "FOXP2",
  tf_x = -7,
  tf_y = -7
)

head(target_genes$FOXP2)


ggplot() + ggseqlogo::geom_logo( as.matrix(pfm[["MA0593.1"]])) +
        ggseqlogo::theme_logo() +
        xlab('') + ylab("FOXP2") + theme(
          axis.text.x=element_blank(),
          axis.text.y=element_blank(),
          axis.title.y = element_text(angle=0))
          

library(UCell)
thal <- MotifTargetScore(
  thal,maxRank=10000,
  method='UCell'
)

df <- GetMotifOverlap(thal)

cur_df <- df %>% subset(tf == 'FOXP2')

plot_var <- 'odds_ratio'
p <- cur_df %>%
  ggplot(aes(y=reorder(module, odds_ratio), x=odds_ratio)) +
  geom_bar(stat='identity', fill=cur_df$color) +
  geom_vline(xintercept = 1, linetype='dashed', color='gray') +
  geom_text(aes(label=Significance), color='black', size=3.5, hjust='center') +
  ylab('') +
  xlab("Odds Ratio") +
  ggtitle("FOXP2 overlap") +
  theme(
    plot.title = element_text(hjust = 0.5)
  ) + theme_bw()


png(paste0(fig_dir, 'Sox9_motif_overlap_or.png'), width=3, height=4, units='in', res=400)
p
dev.off()

saveRDS(thal, file='/media/chang/HDD-3/chang/david/thal_labelled.rds')

```


```{r}
seurat_obj = thal
MotifOverlapBarPlot2 <- function(
  seurat_obj,
  n_tfs = 10,
  plot_size = c(5,6),
  outdir = "MotifOverlaps/",
  motif_font = 'helvetica_regular',
  module_names = NULL,
  wgcna_name=NULL
){

  if(is.null(wgcna_name)){wgcna_name <- seurat_obj@misc$active_wgcna}

  # make output dir if it doesn't exist:
  if(!dir.exists(outdir)){dir.create(outdir)}

  # get Modules
  modules <- GetModules(seurat_obj)
  mods <- levels(modules$module)
  mods <- mods[mods != 'grey']

  if(is.null(module_names)){module_names <- mods}

  # get overlap info from Seurat obj
  overlap_df <- GetMotifOverlap(seurat_obj, wgcna_name)
  motif_df <- GetMotifs(seurat_obj)

  # get pfm list from Seurat obj
  pfm <- GetPFMList(seurat_obj)

  # add motif ID to the overlap_df
  overlap_df$motif_ID <- motif_df$motif_ID[match(overlap_df$tf, motif_df$motif_name)]

  # subset by the modules that we are using:
  overlap_df <- overlap_df %>% subset(module %in% module_names)

  for(cur_mod in module_names){
    print(cur_mod)
    # get df for cur mod
    plot_df <- overlap_df %>% subset(module == cur_mod) %>% top_n(n_tfs, wt=odds_ratio) %>%
      arrange(desc(odds_ratio))

    # make the barplot
    p1 <- plot_df %>% ggplot(aes(y=reorder(tf, odds_ratio), fill=odds_ratio, x=odds_ratio)) +
      geom_bar(stat='identity', width=0.7) + NoLegend() +
      scale_fill_gradient(high=unique(plot_df$color), low='grey90') +
      ylab('') + theme(
        axis.line.y = element_blank(),
        axis.text.y = element_blank(),
        plot.margin = margin(
          t = 0, r = 0, b = 0, l = 0
        )
      )

    # make the motif logo plots:
    plot_list <- list()
    for(i in 1:nrow(plot_df)){
      cur_id <- plot_df[i,'motif_ID']
      cur_name <- plot_df[i,'tf']
      plot_list[[cur_id]] <- ggplot() +
        ggseqlogo::geom_logo( as.matrix(pfm[[cur_id]]), font=motif_font) +
        ggseqlogo::theme_logo() +
        xlab('') + ylab(cur_name) + theme(
          axis.text.x=element_blank(),
          axis.text.y=element_blank(),
          axis.title.y = element_text(angle=0),
          plot.margin = margin(t = 0,  # Top margin
                                 r = 0,  # Right margin
                                 b = 0,  # Bottom margin
                                 l = 0) # Left margin
        )
    }

    # wrap the motif logos
    patch1 <- wrap_plots(plot_list, ncol=1)

    # assemble the plot with patchwork
    outplot <- (patch1 | p1) +
          plot_layout(ncol=2, widths=c(1,2)) +
          plot_annotation(title=paste0('Motif overlaps with ', cur_mod),
          theme = theme(plot.title=element_text(hjust=0.5))
        )

    pdf(paste0('~/motifs/', cur_mod, '_motif_overlaps.pdf'), width=plot_size[1], height=plot_size[2], useDingbats=FALSE)
    print(outplot)
    dev.off()

  }

}


```

```{r}
network <- get_collectri(organism='human', split_complexes=FALSE)
# Run weighted means algorithm.
activities <- decoupleR::run_wmean(mat = as.matrix(thal@assays[["SCT"]]@data),
                                   network = network,
                                   .source = "source",
                                   .targe = "target",
                                   .mor = "mor",
                                   times = 100,
                                   minsize = 5)
saveRDS(activities, file='/media/chang/HDD-3/chang/david/activities.rds')

Idents(thal) = thal$clusters
glut = subset(thal, idents = "Glut")
DefaultAssay(glut) = 'cleaned'
glut = NormalizeData(glut)
markers = FindMarkers(glut, max.cells.per.ident = 6404, group.by = "Condition",
                   ident.1 = "22Q", ident.2 = "CTRL", logfc.threshold = .5)

markers = markers[!grepl("^AC[0-9]",rownames(markers)),]
markers = markers[!grepl("^AL[0-9]",rownames(markers)),]
markers = markers[!grepl("LINC",rownames(markers)),]
markers = markers[!grepl("^MT",rownames(markers)),]

SCpubr::do_VolcanoPlot(sample = glut,n_genes = 30,
                            de_genes = markers)

form <- ~ (1|donor_id) + (1|Condition)
varPart <- fitExtractVarPartModel(as.matrix(glut@assays$cleaned@data[rownames(markers),]), form, glut@meta.data)

vp = varPart[order(varPart$Condition, decreasing = T),]
plotPercentBars( vp[1:80,] )
plotVarPart( vp )


seuratToH5AD(thal,h5ad_output = "/media/chang/HDD-3/chang/david/thal.h5ad",
             pca=T,umap=T)
```

```{r}
seuratToH5AD <- function(seuratObj, h5ad_output, pca=TRUE, umap=FALSE){
  Sys.setenv(RETICULATE_PYTHON = "/home/chang/miniconda3/envs/venv/bin/python")
  library(reticulate)
  #use_condaenv('venv')
  library(Matrix)
  writeMM(t(seuratObj@assays$cleaned@counts), file='combined.mtx')
  #writeMM(t(seuratObj@assays$SCT@data), file='sct.mtx')
  writeMM(t(seuratObj@assays$spliced@counts), file='spliced.mtx')
  writeMM(t(seuratObj@assays$unspliced@counts), file='unspliced.mtx')
  write.csv(rownames(seuratObj@assays$spliced@counts), file = "genes.csv", row.names = FALSE)
  if(umap == TRUE){
  	write.csv(seuratObj@reductions$umap@cell.embeddings, file = "umap.csv", row.names = FALSE)
  }
  if(pca == TRUE){
	  write.csv(seuratObj@reductions$harmony@cell.embeddings, file = "pca.csv", row.names = FALSE)
  }
  write.csv(colnames(seuratObj@assays$spliced@counts), file = "cells.csv", row.names = FALSE)
  write.csv(seuratObj@meta.data, file = "meta.csv", row.names = FALSE)
  source_python('/home/chang/build_combined.py')
  build(h5ad_output, pca = pca, umap = umap)
  file.remove('combined.mtx')
  file.remove('spliced.mtx')
  file.remove('unspliced.mtx')
  file.remove('genes.csv')
  file.remove('cells.csv')
  file.remove('meta.csv')
  #file.remove("sct.csv")
  if(umap == TRUE){
  	file.remove('umap.csv')
  }
  if(pca == TRUE){
	  file.remove('pca.csv')
  }
}

```

```{r}




sce = as.SingleCellExperiment(thal, assay='SCT')
sce$id <- paste0(thal$Condition, thal$donor_id)
# Create pseudobulk data by specifying cluster_id and sample_id
# Count data for each cell type is then stored in the `assay` field
# assay: entry in assayNames(sce) storing raw counts
# cluster_id: variable in colData(sce) indicating cell clusters
# sample_id: variable in colData(sce) indicating sample id for aggregating cells
pb <- aggregateToPseudoBulk(sce,
    assay = "counts",
    cluster_id = "clusters",  
    sample_id = "id",
    verbose = FALSE)

res.proc = processAssays( pb, ~ Condition, min.count=5)

plotVoom( res.proc)

vp.lst = fitVarPart( res.proc, ~ Condition)

genes = vp.lst$gene[2:4]
plotPercentBars(vp.lst[vp.lst$gene %in% genes,])

plotVarPart(vp.lst, label.angle=60,) 

res.dl = dreamlet( res.proc, ~ Condition)
coefNames(res.dl) 

plotVolcano( res.dl, coef = 'ConditionCTRL' ,ncol = 6)
```
