---
title: "Vasc"
output: html_notebook
---

```{r setup, include=FALSE}
library(Seurat)
library(dittoSeq)
library(harmony)
library(SeuratWrappers)
library(scCustomize)
library(scds)
library(Scissor)
library(presto)
library(WGCNA)
library(hdWGCNA)
library(MAGMA.Celltyping)
library(EWCE)
library(MungeSumstats)
library(cowplot)
library(patchwork)
library(tidyverse)
library(UCell)
library(MetBrewer)
library(igraph)
library(enrichR)
library(CellChat)
library(UCell)
library(pathfindR)

#load('/media/chang/HDD-1/chang/fastqs/aneu/solo.RData')
```

```{r}
# data.dir should be Solo.out dir
ReadSTAR <- function(data.dir, proj = "SeuratObject", feats = 1000, cutoff=1) {
  # test dir /media/chang/HDD-1/chang/fastqs/ctrl/ctrl085_L_Solo.out/
  data.dir.genefull = paste0(data.dir, "/GeneFull_Ex50pAS/raw")
  mtx = file.path(data.dir.genefull, "UniqueAndMult-EM.mtx")
  cells = file.path(data.dir.genefull, "barcodes.tsv")
  features = file.path(data.dir.genefull, "features.tsv")
  seurat = as.Seurat(cxds_bcds_hybrid(as.SingleCellExperiment
                                      (CreateSeuratObject(ReadMtx
                                      (mtx = mtx, cells = cells, 
                                      features = features), 
                                      project = proj, min.features = feats))))
  data.dir.velo = paste0(data.dir, "/Velocyto/raw")
  cells = file.path(data.dir.velo, "barcodes.tsv")
  features = file.path(data.dir.velo, "features.tsv")
  spliced = file.path(data.dir.velo, "spliced.mtx")
  unspliced = file.path(data.dir.velo, "unspliced.mtx")
  ambi = file.path(data.dir.velo, "ambiguous.mtx")
  spliced = ReadMtx(mtx = spliced, cells = cells, features = features)
  unspliced = ReadMtx(mtx = unspliced, cells = cells, features = features)
  ambi = ReadMtx(mtx = ambi, cells = cells, features = features)
  spliced = spliced + ambi
  spliced = spliced[,colnames(seurat)]
  unspliced = unspliced[,colnames(seurat)]
  seurat[['spliced']] = CreateAssayObject(spliced)
  seurat[['unspliced']] = CreateAssayObject(unspliced)
  seurat[['percent.mt']] = PercentageFeatureSet(seurat, pattern = "MT-",assay = 'RNA')
  seurat = subset(seurat, subset=hybrid_score < cutoff & percent.mt < 30)
  return(seurat)
}
```

```{r}
dir = "/media/chang/HDD-1/chang/fastqs/aneu_bulk/"
files = list.files(path = dir ,pattern = "\\Gene.out.tab$")
samples = unlist(lapply(strsplit(files, split="_"), function(i) i[[1]]))
mtx = lapply(files, function(i){
  tmp = read.table(paste0(dir, i), sep = '\t', header = FALSE,row.names = 1)
  tmp = tmp[5:dim(tmp)[1],]
  tmp2 = tmp$V2 # Unstranded
  names(tmp2) = rownames(tmp)
  tmp2
})
names(mtx) = samples
mtx = do.call(cbind, mtx)
meta = read.table(paste0(dir,"meta.txt"), sep='\t',header=T)
meta$pheno = 0
meta$pheno[meta$Condition == "R"] = 1
mtx = mtx[,meta$Sample]

gtf  = rtracklayer::import('/media/chang/HDD-1/reference_cellranger/gencode.v39.primary_assembly.annotation.gtf.filtered.gtf')
gtf_df = as.data.frame(gtf)
genes = unique(gtf_df[ ,c("gene_id","gene_name")])
rownames(genes) = genes$gene_id
genes = genes[rownames(mtx),]
rownames(mtx) = genes$gene_name
```

```{r}
ctrl085_L = ReadSTAR("/media/chang/HDD-1/chang/fastqs/ctrl/ctrl085_L_Solo.out/", "ctrl085_L",cutoff = 1.5)
ctrl085_S = ReadSTAR("/media/chang/HDD-1/chang/fastqs/ctrl/ctrl085_S_Solo.out/", "ctrl085_S",cutoff = 2)
ctrl086_1 = ReadSTAR("/media/chang/HDD-1/chang/fastqs/ctrl/ctrl086_1_Solo.out/", "ctrl086_1")
ctrl086_2 = ReadSTAR("/media/chang/HDD-1/chang/fastqs/ctrl/ctrl086_2_Solo.out/", "ctrl086_2", cutoff = 1.5)
ctrl099_21 = ReadSTAR("/media/chang/HDD-1/chang/fastqs/ctrl/ctrl099_21_Solo.out/", "ctrl099_21")
ctrl099_22 = ReadSTAR("/media/chang/HDD-1/chang/fastqs/ctrl/ctrl099_22_Solo.out/", "ctrl099_22")
ctrl12 = ReadSTAR("/media/chang/HDD-7/chang/122120_czb/AVM_Paper/EW_Peri_Ctr/reads/ctrl12_Solo.out/", "ctrl12", cutoff=2)
a1211 = ReadSTAR("/media/chang/HDD-1/chang/fastqs/aneu/aneu_sc/a1211_Solo.out/", "a1211", 500, cutoff = 1)
a1217 = ReadSTAR("/media/chang/HDD-1/chang/fastqs/aneu/aneu_sc/a1217_Solo.out/", "a1217", 500, cutoff = 1)
a219 = ReadSTAR("/media/chang/HDD-1/chang/fastqs/aneu/aneu_sc/a219_Solo.out/", "a219", 500, cutoff=2)
#cav = ReadSTAR("/media/chang/HDD-7/chang/122120_czb/Cav_Paper/Cav_Solo.out/", "cav")
```

```{r}
merged = merge(ctrl085_L, c(ctrl085_S, ctrl086_1, ctrl086_2, ctrl099_21, ctrl099_22, ctrl12, a1211, a1217, a219))
merged$hybrid_score[is.na(merged$hybrid_score)] = 0
merged$Condition = "CTRL"
merged$Condition[grepl("a", merged$orig.ident)] = "Aneu"
merged$Sample = merged$orig.ident
merged$Sample[grepl('ctrl085', merged$orig.ident)] = "ctrl085"
merged$Sample[grepl('ctrl086', merged$orig.ident)] = "ctrl086"
```

```{r}
merged = SCTransform(merged, vst.flavor='v2', n_genes=NULL, ncells=NULL,variable.features.n = 3000)
VariableFeatures(merged) = VariableFeatures(merged)[!grepl("RPS", VariableFeatures(merged))] 
VariableFeatures(merged) = VariableFeatures(merged)[!grepl("RPL", VariableFeatures(merged))] 
VariableFeatures(merged) = VariableFeatures(merged)[!grepl("MT-", VariableFeatures(merged))] 
VariableFeatures(merged) = VariableFeatures(merged)[!grepl("MALAT1", VariableFeatures(merged))] 

merged = RunPCA(merged)
merged = RunHarmony(merged, group.by.vars = c("Sample",'Condition'), plot_convergence = T, assay.use = "SCT")
merged = RunUMAP(merged, dims=1:50, reduction='harmony')
merged = FindNeighbors(merged, dims=1:50, reduction = 'harmony')
merged = FindClusters(merged, resolution = 1.2, algorithm = 2)

DimPlot(merged, cols=dittoColors(), label=T, group.by='SCT_snn_res.0.4') + NoAxes()
DimPlot(merged, cols=dittoColors(), label=F, group.by='Condition') + NoAxes()

FeaturePlot(merged, features = c("HIGD1B","DCN","LUM","KCNT2","CNN1","KCNJ8","ACTA2","RGS5","CLDN5"),max.cutoff = 'q99',ncol = 3, order=T) & NoAxes() & scale_color_viridis_c()
```

```{r}
test = wilcoxauc(merged_subset, group_by = "clusters",seurat_assay = 'SCT')
test = FindMarkers(merged_subset, ident.1 = "Aneu", group.by = "clusters", only.pos = T, assay = "RNA")
test2 = FindMarkers(merged_subset, ident.1 = "Aneu", ident.2 = 'FB',
                    group.by = "clusters", only.pos = T, assay = "RNA")
head(test[order(test$avg_log2FC, decreasing = T),],20)
head(test2[order(test2$avg_log2FC, decreasing = T),],20)


Stacked_VlnPlot(merged_subset, features = gwas_genes, x_lab_rotate = TRUE,
    colors_use = dittoColors(), split.by = "Condition", group.by = 'clusters') 
```

```{r}
Rcpp::sourceCpp(code='
#include <Rcpp.h>
using namespace Rcpp;
// [[Rcpp::export]]
IntegerMatrix asMatrix(NumericVector rp,
                       NumericVector cp,
                       NumericVector z,
                       int nrows,
                       int ncols){
  int k = z.size() ;
  IntegerMatrix  mat(nrows, ncols);
  for (int i = 0; i < k; i++){
      mat(rp[i],cp[i]) = z[i];
  }
  return mat;
}
' )


as_matrix <- function(mat){
  
  row_pos <- mat@i
  col_pos <- findInterval(seq(mat@x)-1,mat@p[-1])
  
  tmp <- asMatrix(rp = row_pos, cp = col_pos, z = mat@x,
                  nrows =  mat@Dim[1], ncols = mat@Dim[2])
  
  row.names(tmp) <- mat@Dimnames[[1]]
  colnames(tmp) <- mat@Dimnames[[2]]
  return(tmp)
}
```

```{r}
merged@graphs$RNA_snn = merged@graphs$SCT_snn
DefaultAssay(merged) = "RNA"
merged = NormalizeData(merged)
merged@assays$RNA@data = merged@assays$SCT@data

infos = Scissor(bulk_dataset = mtx, sc_dataset= merged, phenotype = meta$pheno,
                 alpha = NULL, family = "binomial", tag = c("UNR","RUP"),cutoff = .3,
                 Save_file = '/media/chang/HDD-1/chang/fastqs/aneu/Scissor.RData')
merged$Scissor = "BG"
merged$Scissor[infos$Scissor_pos] = "Pos"
merged$Scissor[infos$Scissor_neg] = "Neg"

save.image('/media/chang/HDD-1/chang/fastqs/aneu/solo.RData')

evaluate_summary = evaluate.cell('/media/chang/HDD-1/chang/fastqs/aneu/Scissor.RData', 
                                  infos, FDR = 0.05, bootstrap_n = 100)
evaluate_summary[1:5,1:4]

DimPlot_scCustom(seurat_object = merged, group.by = "Scissor", pt.size = 1) + NoAxes()
DimPlot_scCustom(seurat_object = merged_subset, group.by = "Scissor", pt.size = 1) + NoAxes()
dittoBarPlot(merged, var='Condition', group.by = 'Scissor')
dittoBarPlot(merged_subset, var='Condition', group.by = 'Scissor')
dittoBarPlot(merged_subset, var='Scissor', group.by = 'clusters')

```


```{r}
merged$clusters2 = as.character(merged$SCT_snn_res.1.2)
merged$clusters2[merged$SCT_snn_res.1.2 == 1] = "SMC1"
merged$clusters2[merged$SCT_snn_res.1.2 == 2] = "SMC1"
merged$clusters2[merged$SCT_snn_res.1.2 == 3] = "SMC1"
merged$clusters2[merged$SCT_snn_res.1.2 == 5] = "SMC1"
merged$clusters2[merged$SCT_snn_res.1.2 == 19] = "SMC1"
merged$clusters2[merged$SCT_snn_res.1.2 == 36] = "SMC1"
merged$clusters2[merged$SCT_snn_res.1.2 == 8] = "FBMC1"
merged$clusters2[merged$SCT_snn_res.1.2 == 26] = "FBMC2"
merged$clusters2[merged$SCT_snn_res.1.2 == 40] = "LQ"
merged$clusters2[merged$SCT_snn_res.1.2 == 42] = "LQ"
merged$clusters2[merged$SCT_snn_res.1.2 == 11] = "FB1"
merged$clusters2[merged$SCT_snn_res.1.2 == 16] = "FB2"
merged$clusters2[merged$SCT_snn_res.1.2 == 34] = "FB3"
merged$clusters2[merged$SCT_snn_res.1.2 == 13] = "Aneu1"
merged$clusters2[merged$SCT_snn_res.1.2 == 20] = "Aneu2"
merged$clusters2[merged$SCT_snn_res.1.2 == 25] = "PC"
merged$clusters2[merged$SCT_snn_res.1.2 == 7] = "SMC2"
merged$clusters2[merged$SCT_snn_res.1.2 == 21] = "SMC3"
merged$clusters2[merged$SCT_snn_res.1.2 == 30] = "OL"
merged$clusters2[merged$SCT_snn_res.1.2 == 33] = "OC"
merged$clusters2[merged$SCT_snn_res.1.2 == 37] = "AC"
merged$clusters2[merged$SCT_snn_res.1.2 == 10] = "Artery"
merged$clusters2[merged$SCT_snn_res.1.2 == 12] = "Venule"
merged$clusters2[merged$SCT_snn_res.1.2 == 28] = "Venous"
merged$clusters2[merged$SCT_snn_res.1.2 == 39] = "LQ"
merged$clusters2[merged$SCT_snn_res.1.2 == 31] = "RBC"
merged$clusters2[merged$SCT_snn_res.1.2 == 4] = "Ly1"
merged$clusters2[merged$SCT_snn_res.1.2 == 6] = "Ly2"
merged$clusters2[merged$SCT_snn_res.1.2 == 35] = "Ly2"
merged$clusters2[merged$SCT_snn_res.1.2 == 18] = "Ly3"
merged$clusters2[merged$SCT_snn_res.1.2 == 24] = "Ly4"
merged$clusters2[merged$SCT_snn_res.1.2 == 29] = "Neu"
merged$clusters2[merged$SCT_snn_res.1.2 == 41] = "Neu"
merged$clusters2[merged$SCT_snn_res.1.2 == 38] = "pDC"
merged$clusters2[merged$SCT_snn_res.1.2 == 43] = "BC"
merged$clusters2[merged$SCT_snn_res.1.2 == 32] = "BC"
merged$clusters2[merged$SCT_snn_res.1.2 == 14] = "BC"
merged$clusters2[merged$SCT_snn_res.1.2 == 23] = "cDC"
merged$clusters2[merged$SCT_snn_res.1.2 == 0] = "PVM1"
merged$clusters2[merged$SCT_snn_res.1.2 == 9] = "PVM1"
merged$clusters2[merged$SCT_snn_res.1.2 == 17] = "PVM2"
merged$clusters2[merged$SCT_snn_res.1.2 == 15] = "PVM3"
merged$clusters2[merged$SCT_snn_res.1.2 == 27] = "PVM3"
merged$clusters2[merged$SCT_snn_res.1.2 == 22] = "PVM4"

Idents(merged) = merged$clusters2
DimPlot_scCustom(seurat_object = merged, group.by = "clusters", 
                 label=T, pt.size = .1) + NoAxes()
```

```{r}
unk = FindMarkers(merged, ident.1 = 'UNK', only.pos = T)
write.table(unk[order(unk$avg_log2FC, decreasing = T),],'unk_markers.csv', 
            quote=F, sep='\t')
```

```{r}
merged@assays$RNA@data <- shifted_log_transform(merged@assays$RNA@counts, 
                                                overdispersion = 0.01, on_disk = TRUE)
DefaultAssay(merged) = "RNA"

merged.split <- SplitObject(merged, split.by = 'Condition')

passed = names(table(merged.split$Aneu$clusters)[table(merged.split$Aneu$clusters)>= 20])
passed = passed[passed!="LQ"]
merged.split$CTRL = subset(merged.split$CTRL, idents=passed)
merged.split$Aneu = subset(merged.split$Aneu, idents=passed)

cellchat.aneu <- createCellChat(object = merged.split$Aneu, group.by = "clusters", assay = "SCT")
cellchat.ctrl <- createCellChat(object = merged.split$CTRL, group.by = "clusters", assay = "SCT")
cellchat.aneu@DB <- CellChatDB.human
cellchat.ctrl@DB <- CellChatDB.human
cellchat.aneu <- subsetData(cellchat.aneu)
cellchat.ctrl <- subsetData(cellchat.ctrl)
cellchat.aneu <- identifyOverExpressedGenes(cellchat.aneu)
cellchat.aneu <- identifyOverExpressedInteractions(cellchat.aneu)
cellchat.aneu <- projectData(cellchat.aneu, PPI.human)
cellchat.ctrl <- identifyOverExpressedGenes(cellchat.ctrl)
cellchat.ctrl <- identifyOverExpressedInteractions(cellchat.ctrl)
cellchat.ctrl <- projectData(cellchat.ctrl, PPI.human)
cellchat.ctrl <- computeCommunProb(cellchat.ctrl)
cellchat.ctrl <- filterCommunication(cellchat.ctrl, min.cells = 20)
cellchat.aneu <- computeCommunProb(cellchat.aneu)
cellchat.aneu <- filterCommunication(cellchat.aneu, min.cells = 20)
cellchat.aneu <- computeCommunProbPathway(cellchat.aneu)
cellchat.ctrl <- computeCommunProbPathway(cellchat.ctrl)
cellchat.ctrl <- aggregateNet(cellchat.ctrl)
cellchat.aneu <- aggregateNet(cellchat.aneu)
cellchat.aneu <- netAnalysis_computeCentrality(cellchat.aneu, slot.name = "netP") 
cellchat.ctrl <- netAnalysis_computeCentrality(cellchat.ctrl, slot.name = "netP") 

object.list = list(cellchat.aneu,cellchat.ctrl)
cellchat <- mergeCellChat(object.list,
                          add.names = c("ANEU","CTRL"))

gg1 <- compareInteractions(cellchat, show.legend = F, group = c(1,2))
gg2 <- compareInteractions(cellchat, show.legend = F, group = c(1,2), measure = "weight")
gg1 + gg2

gg1 <- rankNet(cellchat, mode = "comparison", stacked = T, do.stat = TRUE)
gg2 <- rankNet(cellchat, mode = "comparison", stacked = F, do.stat = TRUE)
gg1 + gg2


pathways.show <- c("PERIOSTIN") 
weight.max <- getMaxWeight(object.list, slot.name = c("netP"), 
                           attribute = pathways.show) 
par(mfrow = c(1,2), xpd=TRUE)
for (i in 1:length(object.list)) {
  netVisual_aggregate(object.list[[i]], signaling = pathways.show, layout = "circle", 
                      edge.width.max = 10, signaling.name = paste(pathways.show, names(object.list)[i]))
}

num.link <- sapply(object.list, function(x) {rowSums(x@net$count) + colSums(x@net$count)-diag(x@net$count)})
weight.MinMax <- c(min(num.link), max(num.link)) # control the dot size in the different datasets
gg <- list()
for (i in 1:length(object.list)) {
  gg[[i]] <- netAnalysis_signalingRole_scatter(object.list[[i]], title = names(object.list)[i], weight.MinMax = weight.MinMax)
}
#> Signaling role analysis on the aggregated cell-cell communication network from all signaling pathways
#> Signaling role analysis on the aggregated cell-cell communication network from all signaling pathways
patchwork::wrap_plots(plots = gg)
```


```{r}
merged.split$Aneu$Scissor2 = merged.split$Aneu$Scissor
merged.split$Aneu$Scissor2[merged.split$Aneu$Scissor2 == "BG"] = "Neg"
merged.split.rupture <- SplitObject(merged.split$Aneu, split.by = 'Scissor2')
passed = names(table(merged.split.rupture$Pos$clusters)[table(merged.split.rupture$Pos$clusters)>= 20])
merged.split.rupture$Neg = subset(merged.split.rupture$Neg, idents=passed)
merged.split.rupture$Pos = subset(merged.split.rupture$Pos, idents=passed)

cellchat.rupt <- createCellChat(object = merged.split.rupture$Pos, group.by = "clusters")
cellchat.unrupt <- createCellChat(object = merged.split.rupture$Neg, group.by = "clusters")
cellchat.rupt@DB <- CellChatDB.human
cellchat.unrupt@DB <- CellChatDB.human
cellchat.rupt <- subsetData(cellchat.rupt)
cellchat.unrupt <- subsetData(cellchat.unrupt)
cellchat.rupt <- identifyOverExpressedGenes(cellchat.rupt)
cellchat.rupt <- identifyOverExpressedInteractions(cellchat.rupt)
cellchat.rupt <- projectData(cellchat.rupt, PPI.human)
cellchat.unrupt <- identifyOverExpressedGenes(cellchat.unrupt)
cellchat.unrupt <- identifyOverExpressedInteractions(cellchat.unrupt)
cellchat.unrupt <- projectData(cellchat.unrupt, PPI.human)
cellchat.unrupt <- computeCommunProb(cellchat.unrupt)
cellchat.unrupt <- filterCommunication(cellchat.unrupt, min.cells = 20)
cellchat.rupt <- computeCommunProb(cellchat.rupt)
cellchat.rupt <- filterCommunication(cellchat.rupt, min.cells = 20)
cellchat.rupt <- computeCommunProbPathway(cellchat.rupt)
cellchat.unrupt <- computeCommunProbPathway(cellchat.unrupt)
cellchat.unrupt <- aggregateNet(cellchat.unrupt)
cellchat.rupt <- aggregateNet(cellchat.rupt)
cellchat.rupt <- netAnalysis_computeCentrality(cellchat.rupt, slot.name = "netP") 
cellchat.unrupt <- netAnalysis_computeCentrality(cellchat.unrupt, slot.name = "netP") 

object.list2 = list(cellchat.rupt,cellchat.unrupt)
cellchat2 <- mergeCellChat(object.list2,
                          add.names = c("Pos","Neg"))

gg1 <- compareInteractions(cellchat2, show.legend = F, group = c(1,2))
gg2 <- compareInteractions(cellchat2, show.legend = F, group = c(1,2), measure = "weight")
gg1 + gg2

gg1 <- rankNet(cellchat2, mode = "comparison", stacked = T, do.stat = TRUE, font.size = 15)
gg2 <- rankNet(cellchat2, mode = "comparison", stacked = F, do.stat = TRUE, font.size = 15)
gg1 + gg2


pathways.show <- c("ICAM") 
weight.max <- getMaxWeight(object.list2, slot.name = c("netP"), 
                           attribute = pathways.show) 
par(mfrow = c(1,2), xpd=TRUE)
for (i in 1:length(object.list)) {
  netVisual_aggregate(object.list2[[i]], signaling = pathways.show, layout = "circle", 
                      edge.width.max = 10, signaling.name = paste(pathways.show, names(object.list2)[i]))
}


gg1 <- netAnalysis_signalingChanges_scatter(cellchat2, idents.use = "OC", signaling.exclude = c("SPP1","GALECTIN"))
gg2 <- netAnalysis_signalingChanges_scatter(cellchat2, idents.use = "Aneu1", signaling.exclude = c("FN1","COLLAGEN","SPP1", "MIF", "LAMININ"))
patchwork::wrap_plots(plots = list(gg1,gg2))
```

```{r}
merged.vasc2 = merged.vasc
merged3 = merge(merged.aneu, ctrl12)
cells = intersect(colnames(merged.vasc2), colnames(merged3))
merged3 = subset(merged3, cells=cells)

cells2 = setdiff(colnames(merged.vasc2), cells)
tmp = merged.vasc2@meta.data
merged.vasc2 = subset(merged.vasc2, cells=cells2)
merged.vasc2 = merge(merged.vasc2, merged3)
tmp = tmp[colnames(merged.vasc2),]
merged.vasc2$clusters = tmp$clusters
merged.vasc2$Scissor = tmp$Scissor
merged.vasc2$Condition = tmp$Condition
merged.vasc2$Sample = tmp$Sample
merged.vasc = merged.vasc2

merged.vasc = subset(merged2, idents=c("FB1","Aneu1","Aneu2","FBMC1","FBMC2",
                                      "FB2","SMC2","PC"))

DefaultAssay(merged.vasc) = 'RNA'
merged.vasc = NormalizeData(merged.vasc)
merged.vasc = SCTransform(merged.vasc, vst.flavor='v2', ncells=NULL, n_genes=NULL)
VariableFeatures(merged.vasc) = VariableFeatures(merged.vasc)[!grepl("RPS", VariableFeatures(merged.vasc))] 
VariableFeatures(merged.vasc) = VariableFeatures(merged.vasc)[!grepl("RPL", VariableFeatures(merged.vasc))] 
VariableFeatures(merged.vasc) = VariableFeatures(merged.vasc)[!grepl("MT-", VariableFeatures(merged.vasc))] 
VariableFeatures(merged.vasc) = VariableFeatures(merged.vasc)[!grepl("MALAT1", VariableFeatures(merged.vasc))] 
merged.vasc@assays$SCT@scale.data = merged.vasc@assays$SCT@scale.data[VariableFeatures(merged.vasc),] 


merged.vasc = RunPCA(merged.vasc, npcs = 30)
merged.vasc = RunHarmony(merged.vasc, group.by.vars = c('Sample', "Condition"), 
                         plot_convergence = T, assay.use = "SCT")
merged.vasc = RunUMAP(merged.vasc, dims=1:30, reduction = 'harmony')
merged.vasc = FindNeighbors(merged.vasc, dims=1:30, reduction = 'harmony')
merged.vasc = FindClusters(merged.vasc , resolution = .1, algorithm = 2)

seuratToH5AD(merged.vasc, "aneu.h5ad", rownames(merged.vasc@assays$SCT@data))

DimPlot(merged.vasc, label=T, pt.size = 1, cols=dittoColors(),
        group.by = 'clusters') + NoAxes()
DimPlot(merged.vasc, label=T, pt.size = 1, cols=dittoColors(), 
        group.by = 'Condition') + NoAxes()
DimPlot(merged.vasc, label=T, pt.size = 1, cols=dittoColors(), 
        group.by = 'Scissor') + NoAxes()
FeaturePlot(merged.vasc, features = 'KCNJ8', order=T, max.cutoff = 'q99') + 
  NoAxes() + scale_color_viridis_c()

markers = wilcoxauc(merged.vasc, seurat_assay = 'SCT')
top_markers(markers)
```

```{r}
gwas_genes = c('SLC22A5','SLC22A4','P4HA2','SOX17','NT5C2','MARCKSL1P1',
                'FGD6','NR2C1','PSMA4','BCAR1','RP11.252K23.2')
gwas = list()
gwas$gwas = gwas_genes
gwas$background = setdiff(rownames(merged@assays$SCT@counts) ,gwas$gwas)

FeaturePlot(merged, features = gwas,max.cutoff = 'q99',ncol=3, order=T, pt.size = .1) &   NoAxes() & scale_color_viridis_c()

merged = AddModuleScore_UCell(merged, features = gwas, assay = 'SCT', maxRank = 34000)

FeaturePlot(merged, features = c("gwas_UCell"), max.cutoff = 'q99', order=T, 
            pt.size = .1, min.cutoff = 'q25') & NoAxes() & scale_color_viridis_c()
``` 

# Plotting for the paper
```{r}
hsamgi <- buildMSIGDB(species="human",keytype="SYMBOL",anntype="GO")
hsakom <- buildAnnot(species = "human",keytype="SYMBOL", anntype = "KEGGM")

setwd("~/Aneu_paper")

id = unique(Idents(merged))
id = id[id !="LQ"]
merged2 = subset(merged, idents=id)
merged2$clusters[merged2$clusters == "FB3"] = "FBMC3"
merged2$clusters[merged2$clusters == "Artery"] = "EC1"
merged2$clusters[merged2$clusters == "Venule"] = "EC2"
merged2$clusters[merged2$clusters == "Venous"] = "EC3"
Idents(merged2) = merged2$clusters

merged2@assays$RNA@data <- shifted_log_transform(merged2@assays$RNA@counts, 
                                                overdispersion = 0.01, on_disk = TRUE)


saveRDS(merged2,file="/media/chang/HDD-1/chang/fastqs/aneu/merged.rds")

pdf("UMAP_merged.pdf",width = 12)
DimPlot_scCustom(merged2, label=F,group.by = 'clusters', raster = F, pt.size = .1) + DimPlot_scCustom(merged2, group.by = 'Condition', colors_use = c("red","lightgray"),raster = F, pt.size = .1) & NoAxes() & NoLegend()
dev.off()

pdf("UMAP_legend_merged.pdf",width = 12)
DimPlot_scCustom(merged2, label=F, group.by = 'clusters', raster = T, pt.size = .2) + DimPlot_scCustom(merged2, group.by = 'Condition', colors_use = c("red","lightgray"),raster = T, pt.size = .2) & NoAxes()
dev.off()


pdf("bar_clusters_condition.pdf")
dittoBarPlot(merged2, var='Condition', group.by = 'clusters', color.panel = c("red","lightgray"))
dev.off()

write.table(data.frame(rbind(table(merged2$Sample, merged2$clusters))), sep=',',
            "sample_cluster_cell_count.csv",row.names =TRUE, quote=F)

markers = wilcoxauc(merged2, seurat_assay = 'SCT',group_by = "clusters")
top5 = top_markers(markers, n = 3)
top5 = unlist(top5)
top5 = unique(top5[4:length(top5)])
top5 = top5[!grepl("RPS",top5)]
top5 = top5[!grepl("RPL",top5)]
top5 = top5[!grepl("ENSG",top5)]
top5 = top5[!grepl("HSPA",top5)]
top5 = top5[!grepl("HLA",top5)]

pdf('dotplot.pdf', width = 14, height = 10)
top5 = c("DCN",'LUM',"KCNJ8",'ACTA2','CNN1','LYZ','MRC1','F13A1','CD1C','CD8A','CD8B','KCNT2',
         'HIGD1B','CLDN5','ARL15','MECOM','VWF','MFSD2A','IGKC','PTPRC','MMP9','ACP5',"CD3D",
         'SPP1','GNLY','PLP1','PTPRZ1','NRXN1','GPC6','LAMA2','CCL19','IGFBP5','TSHZ2',"GFAP",
         'HBB','HBA1','HBA2','APOD','MYH11','DLC1','NKG7','CCL5','IGFBP7','COL1A2','OLIG1',"GPNMB",
         'SPARC',"BGN","PTGDS","JCHAIN","ZFHX3",'CXCR4',"CD79A","CCL2","PTPRK","VCAN","P2RY13")
Clustered_DotPlot(merged2, features = top5, k = 9,row_label_size = 13)
dev.off()


merged2$group = merged2$Condition
merged2$sample = merged2$Sample
meta2 = propeller(merged2)
pdf('prop_volcano.pdf', width = 14, height = 10)
EnhancedVolcano(meta2,
    lab = rownames(meta2),
    x = 'Tstatistic',
    y = 'FDR',
    title = 'Aneu versus CTRL',
    pCutoff = .05,
    FCcutoff = 1,
    pointSize = 3.0,
    labSize = 5.0, ylim=c(0,6),xlim=c(-8,8),
    col=c('black', 'black', 'black', 'red3'),
    colAlpha = 1)
dev.off()


DefaultAssay(merged2) = "SCT"
avg = AverageExpression(merged2, assays = 'SCT', group.by = 'clusters')
avg = avg$SCT[,c("SMC1","SMC2","SMC3","PC","FB1","FB2","FBMC1","FBMC2","FBMC3",
                 "Aneu1","Aneu2")]
cor.matrix = cor(avg)
pdf("heatmap_cor.pdf", width = 10)
pheatmap::pheatmap(cor.matrix, cluster_rows = T, cluster_cols = T, color=viridis(30)) 
dev.off()

fb1.markers = FindMarkers(merged2, ident.1 = "Aneu1", ident.2 = "FB1", only.pos = T)
fb1.markers = fb1.markers[!grepl("RPS", rownames(fb1.markers)),]
fb1.markers = fb1.markers[!grepl("MT-", rownames(fb1.markers)),]
fb1.markers = fb1.markers[!grepl("RPL", rownames(fb1.markers)),]
fb1.markers$gene = rownames(fb1.markers)
output_df <- run_pathfindR(fb1.markers[,c('gene','avg_log2FC','p_val_adj')])

aneu2.markers = FindMarkers(merged2, ident.1 = "Aneu2", ident.2 = "SMC2", only.pos = T)
aneu2.markers = aneu2.markers[!grepl("RPS", rownames(aneu2.markers)),]
aneu2.markers = aneu2.markers[!grepl("MT-", rownames(aneu2.markers)),]
aneu2.markers = aneu2.markers[!grepl("RPL", rownames(aneu2.markers)),]
aneu2.markers$gene = rownames(aneu2.markers)
output_df2 <- run_pathfindR(aneu2.markers[,c('gene','avg_log2FC','p_val_adj')])

combined_df <- combine_pathfindR_results(result_A = output_df, 
                                         result_B = output_df2)
combined_df2 = combined_df[combined_df$combined_p < .000000001,]
combined_results_graph(combined_df2)

aneu.markers = FindMarkers(merged2, ident.1 = "Aneu1", ident.2 = "Aneu2")
aneu.markers = FindMarkers(merged2, ident.1 = "Pos", ident.2 = "Neg", 
                           subset.ident = "Aneu1", group.by = 'Scissor')
aneu.markers$gene = rownames(aneu.markers)
aneu.markers = aneu.markers[!grepl("RPS", rownames(aneu.markers)),]
aneu.markers = aneu.markers[!grepl("MT-", rownames(aneu.markers)),]
aneu.markers = aneu.markers[!grepl("RPL", rownames(aneu.markers)),]

resgo <- richGO(rownames(aneu.markers[aneu.markers$avg_log2FC > 0.25 & aneu.markers$p_val_adj < .01,]),godata = hsamgi,ontology ="BP")
resgo2 <- richGO(rownames(aneu.markers[aneu.markers$avg_log2FC < -.25 & aneu.markers$p_val_adj < .01,]),godata = hsamgi,ontology ="BP")

res <- compareResult(list(Aneu1=resgo,Aneu2=resgo2))
res2 = res[res$Padj < .000001,]

pdf("aneu_dotplot.pdf",width = 12, height = 10)
comparedot(res2,alpha = 1,low = 'lightgrey', high = 'red')
dev.off()

output_df3 <- run_pathfindR(aneu.markers[,c('gene','avg_log2FC','p_val_adj')])
EnhancedVolcano(aneu.markers,
    lab = rownames(aneu.markers),
    x = 'avg_log2FC',
    y = 'p_val_adj',
    title = 'Aneu1 versus Aneu2',
    pCutoff = 10e-50,
    FCcutoff = 1,
    pointSize = 3.0,
    labSize = 6.0,
    col=c('black', 'black', 'black', 'red3'),
    colAlpha = 1)


DimPlot_scCustom(merged2, group.by = 'Scissor', pt.size = .2, 
                 colors_use = c("lightgrey","orange", "red")) + NoAxes()
pdf("scissor_barplot.pdf")
dittoBarPlot(merged2, var='Scissor', group.by = 'clusters', 
             color.panel  =  c("lightgrey","orange", "red"))
dev.off()

numbers <- length(infos$Scissor_pos) + length(infos$Scissor_neg)
result1 <- reliability.test(X, Y, network, alpha = 0.05, family = "cox", cell_num = numbers, n = 10, nfold = 10)


c("EC1","EC2","EC3","PC","SMC1","SMC2","SMC3","FBMC1","FBMC2","FBMC3","FB1","FB2",
  "BC","cDC","pDC","Ly1","Ly2","Ly3","Ly4","PVM1","PVM2","PVM4","PVM3","OC","AC",
  "Neur","OL")
```


```{r}
sce = as.SingleCellExperiment(merged2)
```
