---
title: "Vasc"
output: html_notebook
---

```{r setup, include=FALSE}
library(Seurat)
library(dittoSeq)
library(harmony)
library(SeuratWrappers)
library(scCustomize)
library(scds)
library(Scissor)
library(presto)
library(WGCNA)
library(scWGCNA)
library(MAGMA.Celltyping)
library(EWCE)
library(MungeSumstats)
library(cowplot)
library(patchwork)
library(tidyverse)
library(UCell)
library(MetBrewer)
library(igraph)
library(enrichR)
library(CellChat)

#load('/media/chang/HDD-1/chang/fastqs/aneu/solo.RData')
```

```{r}
# data.dir should be Solo.out dir
ReadSTAR <- function(data.dir, proj = "SeuratObject", feats = 1000, cutoff=1) {
  # test dir /media/chang/HDD-1/chang/fastqs/ctrl/ctrl085_L_Solo.out/
  data.dir.genefull = paste0(data.dir, "/GeneFull_Ex50pAS/raw")
  mtx = file.path(data.dir.genefull, "UniqueAndMult-EM.mtx")
  cells = file.path(data.dir.genefull, "barcodes.tsv")
  features = file.path(data.dir.genefull, "features.tsv")
  seurat = as.Seurat(cxds_bcds_hybrid(as.SingleCellExperiment
                                      (CreateSeuratObject(ReadMtx
                                      (mtx = mtx, cells = cells, 
                                      features = features), 
                                      project = proj, min.features = feats))))
  data.dir.velo = paste0(data.dir, "/Velocyto/raw")
  cells = file.path(data.dir.velo, "barcodes.tsv")
  features = file.path(data.dir.velo, "features.tsv")
  spliced = file.path(data.dir.velo, "spliced.mtx")
  unspliced = file.path(data.dir.velo, "unspliced.mtx")
  ambi = file.path(data.dir.velo, "ambiguous.mtx")
  spliced = ReadMtx(mtx = spliced, cells = cells, features = features)
  unspliced = ReadMtx(mtx = unspliced, cells = cells, features = features)
  ambi = ReadMtx(mtx = ambi, cells = cells, features = features)
  spliced = spliced + ambi
  spliced = spliced[,colnames(seurat)]
  unspliced = unspliced[,colnames(seurat)]
  seurat[['spliced']] = CreateAssayObject(spliced)
  seurat[['unspliced']] = CreateAssayObject(unspliced)
  seurat[['percent.mt']] = PercentageFeatureSet(seurat, pattern = "MT-",assay = 'RNA')
  seurat = subset(seurat, subset=hybrid_score < cutoff & percent.mt < 30)
  return(seurat)
}

'''
cxds_bcds_hybrid_seurat_batch <- function(obj, batch = "orig.ident", ... ) {
  sub = SplitObject(obj, split.by = batch)
  sub = lapply(sub, function(x){
    x = as.SingleCellExperiment(x)
    x = cxds_bcds_hybrid(x)
    as.Seurat(x)
  })
  return(merge(sub[[1]], sub[2:length(sub)]))
}  

cxds_bcds_hybrid_seurat <- function(obj, ... ) {
  return(as.Seurat(cxds_bcds_hybrid(as.SingleCellExperiment(obj))))
} 
'''
```

```{r}
dir = "/media/chang/HDD-1/chang/fastqs/aneu_bulk/"
files = list.files(path = dir ,pattern = "\\Gene.out.tab$")
samples = unlist(lapply(strsplit(files, split="_"), function(i) i[[1]]))
mtx = lapply(files, function(i){
  tmp = read.table(paste0(dir, i), sep = '\t', header = FALSE,row.names = 1)
  tmp = tmp[5:dim(tmp)[1],]
  tmp2 = tmp$V2 # Unstranded
  names(tmp2) = rownames(tmp)
  tmp2
})
names(mtx) = samples
mtx = do.call(cbind, mtx)
meta = read.table(paste0(dir,"meta.txt"), sep='\t',header=T)
meta$pheno = 0
meta$pheno[meta$Condition == "R"] = 1
mtx = mtx[,meta$Sample]

gtf  = rtracklayer::import('/media/chang/HDD-1/reference_cellranger/gencode.v39.primary_assembly.annotation.gtf.filtered.gtf')
gtf_df = as.data.frame(gtf)
genes = unique(gtf_df[ ,c("gene_id","gene_name")])
rownames(genes) = genes$gene_id
genes = genes[rownames(mtx),]
rownames(mtx) = genes$gene_name
```

```{r}
genome_ref_path = get_genome_ref(
  storage_dir = "/media/chang/HDD-1/chang/fastqs/aneu/aneu_gwas/",
  method = c("magma"), population = c("eur", "afr", "amr", "eas", "sas"))

path_formatted = format_sumstats(path="/media/chang/HDD-1/chang/fastqs/aneu/aneu_gwas/",
                                                 save_path = tempfile(fileext = ".formatted.tsv.gz"),
                                                 ref_genome ="GRCh37")

genesOutPath = map_snps_to_genes(path_formatted = path_formatted,
                                                    genome_build = "GRCh37",
                                                    genome_ref_path = genome_ref_path)
```

```{r}
ctrl085_L = ReadSTAR("/media/chang/HDD-1/chang/fastqs/ctrl/ctrl085_L_Solo.out/", "ctrl085_L",cutoff = 1.5)
ctrl085_S = ReadSTAR("/media/chang/HDD-1/chang/fastqs/ctrl/ctrl085_S_Solo.out/", "ctrl085_S",cutoff = 2)
ctrl086_1 = ReadSTAR("/media/chang/HDD-1/chang/fastqs/ctrl/ctrl086_1_Solo.out/", "ctrl086_1")
ctrl086_2 = ReadSTAR("/media/chang/HDD-1/chang/fastqs/ctrl/ctrl086_2_Solo.out/", "ctrl086_2", cutoff = 1.5)
ctrl099_21 = ReadSTAR("/media/chang/HDD-1/chang/fastqs/ctrl/ctrl099_21_Solo.out/", "ctrl099_21")
ctrl099_22 = ReadSTAR("/media/chang/HDD-1/chang/fastqs/ctrl/ctrl099_22_Solo.out/", "ctrl099_22")
ctrl12 = ReadSTAR("/media/chang/HDD-7/chang/122120_czb/AVM_Paper/EW_Peri_Ctr/reads/ctrl12_Solo.out/", "ctrl12", cutoff=2)
a1211 = ReadSTAR("/media/chang/HDD-1/chang/fastqs/aneu/a1211_Solo.out/", "a1211", 500)
a1217 = ReadSTAR("/media/chang/HDD-1/chang/fastqs/aneu/a1217_Solo.out/", "a1217", 500)
a219 = ReadSTAR("/media/chang/HDD-1/chang/fastqs/aneu/a219_Solo.out/", "a219", 500, cutoff=2)
#cav = ReadSTAR("/media/chang/HDD-7/chang/122120_czb/Cav_Paper/Cav_Solo.out/", "cav")
```

```{r}
merged = merge(ctrl085_L, c(ctrl085_S, ctrl086_1, ctrl086_2, ctrl099_21, ctrl099_22, ctrl12, a1211, a1217, a219))
merged$hybrid_score[is.na(merged$hybrid_score)] = 0
merged$Condition = "CTRL"
merged$Condition[grepl("a", merged$orig.ident)] = "Aneu"
merged$Sample = merged$orig.ident
merged$Sample[grepl('ctrl085', merged$orig.ident)] = "ctrl085"
merged$Sample[grepl('ctrl086', merged$orig.ident)] = "ctrl086"
```

```{r}
merged = SCTransform(merged, vst.flavor='v2', n_genes=NULL, ncells=NULL)
merged = RunPCA(merged)
merged = RunHarmony(merged, group.by.vars = c("Sample",'Condition'), plot_convergence = T, assay.use = "SCT")
merged = RunUMAP(merged, dims=1:50, reduction='harmony')
merged = FindNeighbors(merged, dims=1:50, reduction = 'harmony')

merged = FindClusters(merged, resolution = .1)

merged_subset = subset(merged, idents=c(0,3,4,5))
merged_subset = RunUMAP(merged_subset, dims=1:50, reduction='harmony')
merged_subset = FindNeighbors(merged_subset, dims=1:50, reduction = 'harmony')

merged_subset = FindClusters(merged_subset, resolution = .4)

DimPlot(merged_subset, cols=dittoColors(), label=T, group.by='SCT_snn_res.0.4') + NoAxes()
DimPlot(merged_subset, cols=dittoColors(), label=F, group.by='Condition') + NoAxes()

FeaturePlot(merged_subset, features = 'hybrid_score', order=T,max.cutoff = 'q100')

FeaturePlot(merged_subset, features = c("HIGD1B","DCN","LUM","KCNT2","CNN1","KCNJ8","ACTA2","RGS5","CLDN5"),max.cutoff = 'q99',ncol = 3, order=T) & NoAxes() & scale_color_viridis_c()

merged_subset$clusters = as.character(merged_subset$SCT_snn_res.0.4)
merged_subset$clusters[merged_subset$clusters == 0] = "SMC"
merged_subset$clusters[merged_subset$clusters == 1] = "SMC"
merged_subset$clusters[merged_subset$clusters == 2] = "SMC"
merged_subset$clusters[merged_subset$clusters == 13] = "PC"
merged_subset$clusters[merged_subset$clusters == 12] = "PC"
merged_subset$clusters[merged_subset$clusters == 10] = "Venous"
merged_subset$clusters[merged_subset$clusters == 7] = "Artery"
merged_subset$clusters[merged_subset$clusters == 8] = "Venule"
merged_subset$clusters[merged_subset$clusters == 5] = "FBMC"
merged_subset$clusters[merged_subset$clusters == 4] = "FB"
merged_subset$clusters[merged_subset$clusters == 3] = "vSMC"
merged_subset$clusters[merged_subset$clusters == 6] = "Aneu"
merged_subset$clusters[merged_subset$clusters == 9] = "ExVivo"
merged_subset$clusters[merged_subset$clusters == 11] = "Aneu"

test = wilcoxauc(merged_subset, group_by = "clusters",seurat_assay = 'SCT')
dittoBarPlot(merged_subset, var='Condition', group.by = 'clusters')
```

```{r}
merged_subset2 = merged_subset
merged_subset2 = RunPCA(merged_subset2)
merged_subset2 = RunHarmony(merged_subset2, group.by.vars = c("Sample",'Condition'),
                            plot_convergence = T, assay.use = "SCT")
merged_subset2 = RunUMAP(merged_subset2, dims=1:50, reduction='harmony')
merged_subset2 = FindNeighbors(merged_subset2, dims=1:50, reduction = 'harmony')
DimPlot(merged_subset2, cols=dittoColors(), label=F, group.by='Condition') + NoAxes()

```

```{r}
merged_ctrl = subset(merged, subset=Condition == "CTRL")
#merged_ctrl = SCTransform(merged_ctrl, vst.flavor='v2', n_genes=NULL, ncells=NULL)
#merged_ctrl = RunPCA(merged_ctrl)

merged_aneu = subset(merged, subset=Condition == "Aneu")

VariableFeaturePlot_scCustom(merged_subset, num_features = 20, repel = TRUE,
    y_axis_log = TRUE, assay = 'SCT')


merged_subset_wgcna = merged_subset

merged_subset_wgcna = SetupForWGCNA(
  merged_subset_wgcna,
  gene_select = "fraction", # the gene selection approach
  fraction = 0.01, # fraction of cells that a gene needs to be expressed in order to be included
  wgcna_name = "Vasc" # the name of the scWGCNA experiment
)

merged_subset_wgcna <- MetacellsByGroups(
  seurat_obj = merged_subset_wgcna,
  group.by = c("clusters", "Sample"), # specify the columns in seurat_obj@meta.data to group by
  k = 20, # nearest-neighbors parameter
  ident.group = 'clusters' # set the Idents of the metacell seurat object
)

# normalize metacell expression matrix:
merged_subset_wgcna <- NormalizeMetacells(merged_subset_wgcna)

merged_subset_wgcna <- SetDatExpr(
  merged_subset_wgcna,
  group_name = unique(merged_subset_wgcna$clusters),
  group.by='clusters' # the metadata column containing the cell type info. This same column should have also been used in MetacellsByGroups
)

merged_subset_wgcna = TestSoftPowers(
  merged_subset_wgcna,
  setDatExpr = FALSE, # set this to FALSE since we did this above
)

plot_list <- PlotSoftPowers(merged_subset_wgcna)

wrap_plots(plot_list, ncol=2)

power_table <- GetPowerTable(merged_subset_wgcna)
head(power_table)

merged_subset_wgcna <- ConstructNetwork(
  merged_subset_wgcna, soft_power=6,
  setDatExpr=FALSE
)

PlotDendrogram(merged_subset_wgcna, main='scWGCNA Dendrogram')

# expression matrix for all the WGCNA genes:
merged_subset_wgcna <- ScaleData(
 merged_subset_wgcna,
 features = GetWGCNAGenes(merged_subset_wgcna)
)

merged_subset_wgcna <- GetResidual(
 merged_subset_wgcna,
 features = GetWGCNAGenes(merged_subset_wgcna)
)

# compute all MEs in the full single-cell dataset
merged_subset_wgcna <- ModuleEigengenes(
 merged_subset_wgcna,
 group.by.vars=c("Sample","Condition")
)

hMEs <- GetMEs(merged_subset_wgcna,harmonized = TRUE)

merged_subset_wgcna <- ModuleConnectivity(merged_subset_wgcna)

merged_subset_wgcna <- ResetModuleNames(
  merged_subset_wgcna,
  new_name = "Aneu"
)

modules <- GetModules(merged_subset_wgcna)

merged_subset_wgcna <- ModuleExprScore(
  merged_subset_wgcna,
  n_genes = 25,
  method='UCell'
)

plot_list <- ModuleFeaturePlot(
  merged_subset_wgcna,
  features='hMEs', # plot the hMEs
  order=TRUE # order so the points with highest hMEs are on top
)

# stitch together with patchwork
wrap_plots(plot_list, ncol=5)

ModuleCorrelogram(merged_subset_wgcna)

GGally::ggpairs(GetMEs(merged_subset_wgcna))

# add hMEs to Seurat meta-data:
merged_subset_wgcna@meta.data <- cbind(
  merged_subset_wgcna@meta.data,
  GetMEs(merged_subset_wgcna, harmonized=TRUE)
)
selected_mods <- paste0('Aneu', c(1:10))
VlnPlot(
  merged_subset_wgcna,
  features = selected_mods,
  group.by = 'clusters',
  pt.size = 0 # don't show actual data points
)

theme_set(theme_cowplot())

ModuleNetworkPlot(merged_subset_wgcna)

HubGeneNetworkPlot(
  merged_subset_wgcna,
  n_hubs = 3, n_other=5,
  edge_prop = 0.75,
  mods = 'all'
)

merged_subset_wgcna <- RunModuleUMAP(
  merged_subset_wgcna,
  n_hubs = 10, # number of hub genes to include for the UMAP embedding
  n_neighbors=15, # neighbors parameter for UMAP
  min_dist=0.1 # min distance between points in UMAP space
)

umap_df <- GetModuleUMAP(merged_subset_wgcna)
ggplot(umap_df, aes(x=UMAP1, y=UMAP2)) +
  geom_point(
   color=umap_df$color, # color each point by WGCNA module
   size=umap_df$kME*2 # size of each point based on intramodular connectivity
  ) + umap_theme()

p1 = ModuleUMAPPlot(
  merged_subset_wgcna,
  edge.alpha=0.15,
  sample_edges=TRUE,
  edge_prop=0.2, # proportion of edges to sample (20% here)
  label_hubs=4 # how many hub genes to plot per module?
)


merged_subset_wgcna$Condition <- as.factor(merged_subset_wgcna$Condition)
cur_traits <- c('hybrid_score', 'nCount_RNA', 'nFeature_RNA', 'percent.mt', 'Condition')
merged_subset_wgcna <- ModuleTraitCorrelation(
  merged_subset_wgcna,
  traits = cur_traits,
  group.by='clusters'
)
mt_cor <- GetModuleTraitCorrelation(merged_subset_wgcna)
PlotModuleTraitCorrelation(
  merged_subset_wgcna,
  label = 'fdr',
  label_symbol = 'stars',
  text_size = 2,
  text_digits = 2,
  text_color = 'white',
  high_color = 'yellow',
  mid_color = 'black',
  low_color = 'purple',
  plot_max = 0.2,
  combine=TRUE
)

dbs <- c('GO_Biological_Process_2021','GO_Cellular_Component_2021','GO_Molecular_Function_2021')
merged_subset_wgcna <- RunEnrichr(
  merged_subset_wgcna,
  dbs=dbs, # character vector of enrichr databases to test
  max_genes = 100 # number of genes per module to test
)

# retrieve the output table
enrich_df <- GetEnrichrTable(merged_subset_wgcna)

EnrichrDotPlot(
  merged_subset_wgcna,
  mods = "all", # use all modules (this is the default behavior)
  database = "GO_Biological_Process_2021", # this has to be one of the lists we used above!!!
  n_terms=1 # number of terms for each module
)
```

```{r}
gwas_genes = c('SLC22A5','SLC22A4','P4HA2','SOX17','NT5C2','MARCKSL1P1',
                'FGD6','NR2C1','PSMA4','BCAR1','RP11.252K23.2')

FeaturePlot(merged_subset, features = gwas_genes,max.cutoff = 'q99',ncol = 3, order=T) & NoAxes() & scale_color_viridis_c()

DimPlot_scCustom(merged_subset, split.by = "Condition", num_columns = 2, 
                 repel = TRUE, split_seurat = TRUE,group.by = 'clusters', 
                 colors_use=met.brewer(name = 'VanGogh1',type="continuous",n=10)) & NoAxes()

DimPlot_scCustom(merged_subset, split.by = "Condition", num_columns = 2, label = T,
                 repel = TRUE, split_seurat = TRUE,group.by = 'clusters', 
                 colors_use=dittoColors()) & NoAxes()


Stacked_VlnPlot(merged_subset, features = gwas_genes, x_lab_rotate = TRUE,
    colors_use = dittoColors(), split.by = "Condition", group.by = 'clusters') 



```

```{r}
test = wilcoxauc(merged_subset, group_by = "clusters",seurat_assay = 'SCT')
test = FindMarkers(merged_subset, ident.1 = "Aneu", group.by = "clusters", only.pos = T, assay = "RNA")
test2 = FindMarkers(merged_subset, ident.1 = "Aneu", ident.2 = 'FB',
                    group.by = "clusters", only.pos = T, assay = "RNA")
head(test[order(test$avg_log2FC, decreasing = T),],20)
head(test2[order(test2$avg_log2FC, decreasing = T),],20)


Stacked_VlnPlot(merged_subset, features = gwas_genes, x_lab_rotate = TRUE,
    colors_use = dittoColors(), split.by = "Condition", group.by = 'clusters') 
```

```{r}
Rcpp::sourceCpp(code='
#include <Rcpp.h>
using namespace Rcpp;
// [[Rcpp::export]]
IntegerMatrix asMatrix(NumericVector rp,
                       NumericVector cp,
                       NumericVector z,
                       int nrows,
                       int ncols){
  int k = z.size() ;
  IntegerMatrix  mat(nrows, ncols);
  for (int i = 0; i < k; i++){
      mat(rp[i],cp[i]) = z[i];
  }
  return mat;
}
' )


as_matrix <- function(mat){
  
  row_pos <- mat@i
  col_pos <- findInterval(seq(mat@x)-1,mat@p[-1])
  
  tmp <- asMatrix(rp = row_pos, cp = col_pos, z = mat@x,
                  nrows =  mat@Dim[1], ncols = mat@Dim[2])
  
  row.names(tmp) <- mat@Dimnames[[1]]
  colnames(tmp) <- mat@Dimnames[[2]]
  return(tmp)
}
```

```{r}
merged_subset@graphs$RNA_snn = merged_subset@graphs$SCT_snn
DefaultAssay(merged_subset) = "RNA"
merged_subset = NormalizeData(merged_subset)

table(merged_subset$Condition)[[1]]/table(merged_subset$Condition)[[2]]

infos1 = Scissor(bulk_dataset = mtx, sc_dataset= merged_subset, phenotype = meta$pheno,
                 alpha = 0.05, family = "binomial", tag = c("UNR","RUP"),
                 Save_file = '/media/chang/HDD-1/chang/fastqs/aneu/Scissor.RData')
merged_subset$Scissor = "Neg"
merged_subset$Scissor[infos1$Scissor_pos] = "Pos"

evaluate_summary = evaluate.cell('/media/chang/HDD-1/chang/fastqs/aneu/Scissor.RData', 
                                  infos1, FDR = 0.05, bootstrap_n = 100)
evaluate_summary[1:5,1:4]

infos2 = Scissor(bulk_dataset = mtx, sc_dataset= merged, phenotype = meta$pheno,
                 alpha = 0.05, family = "binomial", tag = c("UNR","RUP"),
                 Save_file = '/media/chang/HDD-1/chang/fastqs/aneu/Scissor.RData')
merged$Scissor = "Neg"
merged$Scissor[infos2$Scissor_pos] = "Pos"

evaluate_summary2 = evaluate.cell('/media/chang/HDD-1/chang/fastqs/aneu/Scissor2.RData', 
                                  infos2, FDR = 0.05, bootstrap_n = 100)
evaluate_summary2[1:5,1:4]

DimPlot_scCustom(seurat_object = merged, group.by = "Scissor", pt.size = 1) + NoAxes()
DimPlot_scCustom(seurat_object = merged_subset, group.by = "Scissor", pt.size = 1) + NoAxes()
dittoBarPlot(merged, var='Condition', group.by = 'Scissor')
dittoBarPlot(merged_subset, var='Condition', group.by = 'Scissor')
dittoBarPlot(merged_subset, var='Scissor', group.by = 'clusters')

```


```{r}
merged$clusters = as.character(merged$SCT_snn_res.0.6)
merged$clusters[merged$clusters == 0] = "SMC"
merged$clusters[merged$clusters == 1] = "SMC"
merged$clusters[merged$clusters == 5] = "SMC"
merged$clusters[merged$clusters == 9] = "FBMC"
merged$clusters[merged$clusters == 8] = "FB"
merged$clusters[merged$clusters == 10] = "Aneu"
merged$clusters[merged$clusters == 4] = "Ly_1"
merged$clusters[merged$clusters == 7] = "Ly_2"
merged$clusters[merged$clusters == 15] = "Ly_3"
merged$clusters[merged$clusters == 19] = "Ly_4"
merged$clusters[merged$clusters == 2] = "My_1"
merged$clusters[merged$clusters == 6] = "My_2"
merged$clusters[merged$clusters == 13] = "My_3"
merged$clusters[merged$clusters == 16] = "My_4"
merged$clusters[merged$clusters == 17] = "My_5"
merged$clusters[merged$clusters == 12] = "Venule"
merged$clusters[merged$clusters == 11] = "Artery"
merged$clusters[merged$clusters == 20] = "Venous"
merged$clusters[merged$clusters == 3] = "vSMC"
merged$clusters[merged$clusters == 24] = "RBC"
merged$clusters[merged$clusters == 25] = "BC"
merged$clusters[merged$clusters == 30] = "LQ"
merged$clusters[merged$clusters == 14] = "LQ"
merged$clusters[merged$clusters == 18] = "FBMC_2"
merged$clusters[merged$clusters == 22] = "OL"
merged$clusters[merged$clusters == 29] = "AC"
merged$clusters[merged$clusters == 28] = "PC"
merged$clusters[merged$clusters == 21] = "Neu"
merged$clusters[merged$clusters == 23] = "UNK"
merged$clusters[merged$clusters == 27] = "FB_2"
merged$clusters[merged$clusters == 26] = "vSMC"

Idents(merged) = merged$clusters
DimPlot_scCustom(seurat_object = merged, group.by = "clusters", 
                 label=T, pt.size = .1) + NoAxes()
```

```{r}
merged.split <- SplitObject(merged, split.by = 'Condition')

passed = names(table(merged.split$Aneu$clusters)[table(merged.split$Aneu$clusters)>= 20])
passed = passed[passed!="LQ"]
merged.split$CTRL = subset(merged.split$CTRL, idents=passed)
merged.split$Aneu = subset(merged.split$Aneu, idents=passed)


cellchat.aneu <- createCellChat(object = merged.split$Aneu, group.by = "clusters")
cellchat.ctrl <- createCellChat(object = merged.split$CTRL, group.by = "clusters")
cellchat.aneu@DB <- CellChatDB.human
cellchat.ctrl@DB <- CellChatDB.human
cellchat.aneu <- subsetData(cellchat.aneu)
cellchat.ctrl <- subsetData(cellchat.ctrl)
cellchat.aneu <- identifyOverExpressedGenes(cellchat.aneu)
cellchat.aneu <- identifyOverExpressedInteractions(cellchat.aneu)
cellchat.aneu <- projectData(cellchat.aneu, PPI.human)
cellchat.ctrl <- identifyOverExpressedGenes(cellchat.ctrl)
cellchat.ctrl <- identifyOverExpressedInteractions(cellchat.ctrl)
cellchat.ctrl <- projectData(cellchat.ctrl, PPI.human)
cellchat.ctrl <- computeCommunProb(cellchat.ctrl)
cellchat.ctrl <- filterCommunication(cellchat.ctrl, min.cells = 20)
cellchat.aneu <- computeCommunProb(cellchat.aneu)
cellchat.aneu <- filterCommunication(cellchat.aneu, min.cells = 20)
cellchat.aneu <- computeCommunProbPathway(cellchat.aneu)
cellchat.ctrl <- computeCommunProbPathway(cellchat.ctrl)
cellchat.ctrl <- aggregateNet(cellchat.ctrl)
cellchat.aneu <- aggregateNet(cellchat.aneu)
cellchat.aneu <- netAnalysis_computeCentrality(cellchat.aneu, slot.name = "netP") 
cellchat.ctrl <- netAnalysis_computeCentrality(cellchat.ctrl, slot.name = "netP") 

object.list = list(cellchat.aneu,cellchat.ctrl)
cellchat <- mergeCellChat(object.list,
                          add.names = c("ANEU","CTRL"))

gg1 <- compareInteractions(cellchat, show.legend = F, group = c(1,2))
gg2 <- compareInteractions(cellchat, show.legend = F, group = c(1,2), measure = "weight")
gg1 + gg2

gg1 <- rankNet(cellchat, mode = "comparison", stacked = T, do.stat = TRUE)
gg2 <- rankNet(cellchat, mode = "comparison", stacked = F, do.stat = TRUE)
gg1 + gg2


pathways.show <- c("PERIOSTIN") 
weight.max <- getMaxWeight(object.list, slot.name = c("netP"), 
                           attribute = pathways.show) 
par(mfrow = c(1,2), xpd=TRUE)
for (i in 1:length(object.list)) {
  netVisual_aggregate(object.list[[i]], signaling = pathways.show, layout = "circle", 
                      edge.width.max = 10, signaling.name = paste(pathways.show, names(object.list)[i]))
}
```


```{r}
merged.split.rupture <- SplitObject(merged.split$Aneu, split.by = 'Scissor')
passed = names(table(merged.split.rupture$Pos$clusters)[table(merged.split.rupture$Pos$clusters)>= 20])
merged.split.rupture$Neg = subset(merged.split.rupture$Neg, idents=passed)
merged.split.rupture$Pos = subset(merged.split.rupture$Pos, idents=passed)

cellchat.rupt <- createCellChat(object = merged.split.rupture$Pos, group.by = "clusters")
cellchat.unrupt <- createCellChat(object = merged.split.rupture$Neg, group.by = "clusters")
cellchat.rupt@DB <- CellChatDB.human
cellchat.unrupt@DB <- CellChatDB.human
cellchat.rupt <- subsetData(cellchat.rupt)
cellchat.unrupt <- subsetData(cellchat.unrupt)
cellchat.rupt <- identifyOverExpressedGenes(cellchat.rupt)
cellchat.rupt <- identifyOverExpressedInteractions(cellchat.rupt)
cellchat.rupt <- projectData(cellchat.rupt, PPI.human)
cellchat.unrupt <- identifyOverExpressedGenes(cellchat.unrupt)
cellchat.unrupt <- identifyOverExpressedInteractions(cellchat.unrupt)
cellchat.unrupt <- projectData(cellchat.unrupt, PPI.human)
cellchat.unrupt <- computeCommunProb(cellchat.unrupt)
cellchat.unrupt <- filterCommunication(cellchat.unrupt, min.cells = 20)
cellchat.rupt <- computeCommunProb(cellchat.rupt)
cellchat.rupt <- filterCommunication(cellchat.rupt, min.cells = 20)
cellchat.rupt <- computeCommunProbPathway(cellchat.rupt)
cellchat.unrupt <- computeCommunProbPathway(cellchat.unrupt)
cellchat.unrupt <- aggregateNet(cellchat.unrupt)
cellchat.rupt <- aggregateNet(cellchat.rupt)
cellchat.rupt <- netAnalysis_computeCentrality(cellchat.rupt, slot.name = "netP") 
cellchat.unrupt <- netAnalysis_computeCentrality(cellchat.unrupt, slot.name = "netP") 

object.list2 = list(cellchat.rupt,cellchat.unrupt)
cellchat2 <- mergeCellChat(object.list2,
                          add.names = c("Pos","Neg"))

gg1 <- compareInteractions(cellchat2, show.legend = F, group = c(1,2))
gg2 <- compareInteractions(cellchat2, show.legend = F, group = c(1,2), measure = "weight")
gg1 + gg2

gg1 <- rankNet(cellchat2, mode = "comparison", stacked = T, do.stat = TRUE, font.size = 15)
gg2 <- rankNet(cellchat2, mode = "comparison", stacked = F, do.stat = TRUE, font.size = 15)
gg1 + gg2


pathways.show <- c("SEMA4") 
weight.max <- getMaxWeight(object.list2, slot.name = c("netP"), 
                           attribute = pathways.show) 
par(mfrow = c(1,2), xpd=TRUE)
for (i in 1:length(object.list)) {
  netVisual_aggregate(object.list2[[i]], signaling = pathways.show, layout = "circle", 
                      edge.width.max = 10, signaling.name = paste(pathways.show, names(object.list2)[i]))
}

```
