---
title: "Vasc"
output: html_notebook
---

```{r setup, include=FALSE}
library(Seurat)
library(dittoSeq)
library(harmony)
library(SeuratWrappers)
library(scCustomize)
library(scds)
library(Scissor)
library(presto)
library(WGCNA)
library(hdWGCNA)
library(MAGMA.Celltyping)
library(EWCE)
library(MungeSumstats)
library(cowplot)
library(patchwork)
library(tidyverse)
library(UCell)
library(MetBrewer)
library(igraph)
library(richR)
library(CellChat)
library(UCell)
library(pathfindR)
library(UpSetR)
library(EnhancedVolcano)
library(escape)
library(msigdbr)
library(fgsea)
library(BayesPrism)
library(SpaceFold)
library(phateR)
library(DESeq2)
#load('/media/chang/HDD-1/chang/fastqs/aneu/solo.RData')
```

```{r}
# data.dir should be Solo.out dir
ReadSTAR <- function(data.dir, proj = "SeuratObject", feats = 1000, cutoff=1) {
  # test dir /media/chang/HDD-1/chang/fastqs/ctrl/ctrl085_L_Solo.out/
  data.dir.genefull = paste0(data.dir, "/GeneFull_Ex50pAS/raw")
  mtx = file.path(data.dir.genefull, "UniqueAndMult-EM.mtx")
  cells = file.path(data.dir.genefull, "barcodes.tsv")
  features = file.path(data.dir.genefull, "features.tsv")
  seurat = as.Seurat(cxds_bcds_hybrid(as.SingleCellExperiment
                                      (CreateSeuratObject(ReadMtx
                                      (mtx = mtx, cells = cells, 
                                      features = features), 
                                      project = proj, min.features = feats))))
  data.dir.velo = paste0(data.dir, "/Velocyto/raw")
  cells = file.path(data.dir.velo, "barcodes.tsv")
  features = file.path(data.dir.velo, "features.tsv")
  spliced = file.path(data.dir.velo, "spliced.mtx")
  unspliced = file.path(data.dir.velo, "unspliced.mtx")
  ambi = file.path(data.dir.velo, "ambiguous.mtx")
  spliced = ReadMtx(mtx = spliced, cells = cells, features = features)
  unspliced = ReadMtx(mtx = unspliced, cells = cells, features = features)
  ambi = ReadMtx(mtx = ambi, cells = cells, features = features)
  spliced = spliced + ambi
  spliced = spliced[,colnames(seurat)]
  unspliced = unspliced[,colnames(seurat)]
  seurat[['spliced']] = CreateAssayObject(spliced)
  seurat[['unspliced']] = CreateAssayObject(unspliced)
  seurat[['percent.mt']] = PercentageFeatureSet(seurat, pattern = "MT-",assay = 'RNA')
  seurat = subset(seurat, subset=hybrid_score < cutoff & percent.mt < 30)
  return(seurat)
}
```

```{r}
dir = "/media/chang/HDD-1/chang/fastqs/aneu/aneu_bulk/"
files = list.files(path = dir ,pattern = "\\Gene.out.tab$")
samples = unlist(lapply(strsplit(files, split="_"), function(i) i[[1]]))
mtx = lapply(files, function(i){
  tmp = read.table(paste0(dir, i), sep = '\t', header = FALSE,row.names = 1)
  tmp = tmp[5:dim(tmp)[1],]
  tmp2 = tmp$V2 # Unstranded
  names(tmp2) = rownames(tmp)
  tmp2
})
names(mtx) = samples
mtx = do.call(cbind, mtx)
meta = read.table("/media/chang/HDD-1/chang/fastqs/aneu/aneu_bulk/meta.txt", sep='\t',header=T)
meta$pheno = 0
meta$pheno[meta$Condition == "R"] = 1
mtx = mtx[,meta$Sample]

gtf  = rtracklayer::import('/media/chang/HDD-1/reference_cellranger/gencode.v39.primary_assembly.annotation.gtf.filtered.gtf')
gtf_df = as.data.frame(gtf)
genes = unique(gtf_df[ ,c("gene_id","gene_name")])
rownames(genes) = genes$gene_id
genes = genes[rownames(mtx),]
rownames(mtx) = genes$gene_name
```


```{r}
meta = meta[meta$Condition != "C",]
mtx = mtx[,meta$Sample]
rownames(meta) = meta$Sample
dds <- DESeqDataSetFromMatrix(countData = mtx,
                              colData = meta,
                              design = ~ Condition)

dds <- DESeq(dds)
res <- lfcShrink(dds, coef="Condition_U_vs_R", type="apeglm")
res= res[!is.na(res$padj),]
res = res[res$padj < .05,]
res = res[abs(res$log2FoldChange) >= 2,]
up = res[res$log2FoldChange > 0,]
down = res[res$log2FoldChange < 0,]
```

```{r}
ctrl085_L = ReadSTAR("/media/chang/HDD-1/chang/fastqs/ctrl/ctrl085_L_Solo.out/", "ctrl085_L",cutoff = 1.5)
ctrl085_S = ReadSTAR("/media/chang/HDD-1/chang/fastqs/ctrl/ctrl085_S_Solo.out/", "ctrl085_S",cutoff = 2)
ctrl086_1 = ReadSTAR("/media/chang/HDD-1/chang/fastqs/ctrl/ctrl086_1_Solo.out/", "ctrl086_1")
ctrl086_2 = ReadSTAR("/media/chang/HDD-1/chang/fastqs/ctrl/ctrl086_2_Solo.out/", "ctrl086_2", cutoff = 1.5)
ctrl099_21 = ReadSTAR("/media/chang/HDD-1/chang/fastqs/ctrl/ctrl099_21_Solo.out/", "ctrl099_21")
ctrl099_22 = ReadSTAR("/media/chang/HDD-1/chang/fastqs/ctrl/ctrl099_22_Solo.out/", "ctrl099_22")
ctrl12 = ReadSTAR("/media/chang/HDD-7/chang/122120_czb/AVM_Paper/EW_Peri_Ctr/reads/ctrl12_Solo.out/", "ctrl12", cutoff=2)
a1211 = ReadSTAR("/media/chang/HDD-1/chang/fastqs/aneu/aneu_sc/a1211_Solo.out/", "a1211", 500, cutoff = 1)
a1217 = ReadSTAR("/media/chang/HDD-1/chang/fastqs/aneu/aneu_sc/a1217_Solo.out/", "a1217", 500, cutoff = 1)
a219 = ReadSTAR("/media/chang/HDD-1/chang/fastqs/aneu/aneu_sc/a219_Solo.out/", "a219", 500, cutoff=2)
#cav = ReadSTAR("/media/chang/HDD-7/chang/122120_czb/Cav_Paper/Cav_Solo.out/", "cav")
```

```{r}
merged = merge(ctrl085_L, c(ctrl085_S, ctrl086_1, ctrl086_2, ctrl099_21, ctrl099_22, ctrl12, a1211, a1217, a219))
merged$hybrid_score[is.na(merged$hybrid_score)] = 0
merged$Condition = "CTRL"
merged$Condition[grepl("a", merged$orig.ident)] = "Aneu"
merged$Sample = merged$orig.ident
merged$Sample[grepl('ctrl085', merged$orig.ident)] = "ctrl085"
merged$Sample[grepl('ctrl086', merged$orig.ident)] = "ctrl086"
```

```{r}
merged = SCTransform(merged, vst.flavor='v2', n_genes=NULL, ncells=NULL,variable.features.n = 3000)
VariableFeatures(merged) = VariableFeatures(merged)[!grepl("RPS", VariableFeatures(merged))] 
VariableFeatures(merged) = VariableFeatures(merged)[!grepl("RPL", VariableFeatures(merged))] 
VariableFeatures(merged) = VariableFeatures(merged)[!grepl("MT-", VariableFeatures(merged))] 
VariableFeatures(merged) = VariableFeatures(merged)[!grepl("MALAT1", VariableFeatures(merged))] 

merged = RunPCA(merged)
merged = RunHarmony(merged, group.by.vars = c("Sample",'Condition'), plot_convergence = T, assay.use = "SCT")
merged = RunUMAP(merged, dims=1:50, reduction='harmony')
merged = FindNeighbors(merged, dims=1:50, reduction = 'harmony')
merged = FindClusters(merged, resolution = 1.2, algorithm = 2)

DimPlot(merged, cols=dittoColors(), label=T, group.by='SCT_snn_res.0.4') + NoAxes()
DimPlot(merged, cols=dittoColors(), label=F, group.by='Condition') + NoAxes()

FeaturePlot(merged, features = c("HIGD1B","DCN","LUM","KCNT2","CNN1","KCNJ8","ACTA2","RGS5","CLDN5"),max.cutoff = 'q99',ncol = 3, order=T) & NoAxes() & scale_color_viridis_c()
```

```{r}
test = wilcoxauc(merged_subset, group_by = "clusters",seurat_assay = 'SCT')
test = FindMarkers(merged_subset, ident.1 = "Aneu", group.by = "clusters", only.pos = T, assay = "RNA")
test2 = FindMarkers(merged_subset, ident.1 = "Aneu", ident.2 = 'FB',
                    group.by = "clusters", only.pos = T, assay = "RNA")
head(test[order(test$avg_log2FC, decreasing = T),],20)
head(test2[order(test2$avg_log2FC, decreasing = T),],20)


Stacked_VlnPlot(merged_subset, features = gwas_genes, x_lab_rotate = TRUE,
    colors_use = dittoColors(), split.by = "Condition", group.by = 'clusters') 
```

```{r}
Rcpp::sourceCpp(code='
#include <Rcpp.h>
using namespace Rcpp;
// [[Rcpp::export]]
IntegerMatrix asMatrix(NumericVector rp,
                       NumericVector cp,
                       NumericVector z,
                       int nrows,
                       int ncols){
  int k = z.size() ;
  IntegerMatrix  mat(nrows, ncols);
  for (int i = 0; i < k; i++){
      mat(rp[i],cp[i]) = z[i];
  }
  return mat;
}
' )


as_matrix <- function(mat){
  
  row_pos <- mat@i
  col_pos <- findInterval(seq(mat@x)-1,mat@p[-1])
  
  tmp <- asMatrix(rp = row_pos, cp = col_pos, z = mat@x,
                  nrows =  mat@Dim[1], ncols = mat@Dim[2])
  
  row.names(tmp) <- mat@Dimnames[[1]]
  colnames(tmp) <- mat@Dimnames[[2]]
  return(tmp)
}
```

```{r}
merged@graphs$RNA_snn = merged@graphs$SCT_snn
DefaultAssay(merged) = "RNA"
merged = NormalizeData(merged)
merged@assays$RNA@data = merged@assays$SCT@data

infos = Scissor(bulk_dataset = mtx, sc_dataset= merged, phenotype = meta$pheno,
                 alpha = NULL, family = "binomial", tag = c("UNR","RUP"),cutoff = .3,
                 Save_file = '/media/chang/HDD-1/chang/fastqs/aneu/Scissor.RData')
merged$Scissor = "BG"
merged$Scissor[infos$Scissor_pos] = "Pos"
merged$Scissor[infos$Scissor_neg] = "Neg"

save.image('/media/chang/HDD-1/chang/fastqs/aneu/solo.RData')



DimPlot_scCustom(seurat_object = merged, group.by = "Scissor", pt.size = 1) + NoAxes()
DimPlot_scCustom(seurat_object = merged_subset, group.by = "Scissor", pt.size = 1) + NoAxes()
dittoBarPlot(merged, var='Condition', group.by = 'Scissor')
dittoBarPlot(merged_subset, var='Condition', group.by = 'Scissor')
dittoBarPlot(merged_subset, var='Scissor', group.by = 'clusters')

```


```{r}
merged$clusters2 = as.character(merged$SCT_snn_res.1.2)
merged$clusters2[merged$SCT_snn_res.1.2 == 1] = "SMC1"
merged$clusters2[merged$SCT_snn_res.1.2 == 2] = "SMC1"
merged$clusters2[merged$SCT_snn_res.1.2 == 3] = "SMC1"
merged$clusters2[merged$SCT_snn_res.1.2 == 5] = "SMC1"
merged$clusters2[merged$SCT_snn_res.1.2 == 19] = "SMC1"
merged$clusters2[merged$SCT_snn_res.1.2 == 36] = "SMC1"
merged$clusters2[merged$SCT_snn_res.1.2 == 8] = "FBMC1"
merged$clusters2[merged$SCT_snn_res.1.2 == 26] = "FBMC2"
merged$clusters2[merged$SCT_snn_res.1.2 == 40] = "LQ"
merged$clusters2[merged$SCT_snn_res.1.2 == 42] = "LQ"
merged$clusters2[merged$SCT_snn_res.1.2 == 11] = "FB1"
merged$clusters2[merged$SCT_snn_res.1.2 == 16] = "FB2"
merged$clusters2[merged$SCT_snn_res.1.2 == 34] = "FB3"
merged$clusters2[merged$SCT_snn_res.1.2 == 13] = "Aneu1"
merged$clusters2[merged$SCT_snn_res.1.2 == 20] = "Aneu2"
merged$clusters2[merged$SCT_snn_res.1.2 == 25] = "PC"
merged$clusters2[merged$SCT_snn_res.1.2 == 7] = "SMC2"
merged$clusters2[merged$SCT_snn_res.1.2 == 21] = "SMC3"
merged$clusters2[merged$SCT_snn_res.1.2 == 30] = "OL"
merged$clusters2[merged$SCT_snn_res.1.2 == 33] = "OC"
merged$clusters2[merged$SCT_snn_res.1.2 == 37] = "AC"
merged$clusters2[merged$SCT_snn_res.1.2 == 10] = "Artery"
merged$clusters2[merged$SCT_snn_res.1.2 == 12] = "Venule"
merged$clusters2[merged$SCT_snn_res.1.2 == 28] = "Venous"
merged$clusters2[merged$SCT_snn_res.1.2 == 39] = "LQ"
merged$clusters2[merged$SCT_snn_res.1.2 == 31] = "RBC"
merged$clusters2[merged$SCT_snn_res.1.2 == 4] = "Ly1"
merged$clusters2[merged$SCT_snn_res.1.2 == 6] = "Ly2"
merged$clusters2[merged$SCT_snn_res.1.2 == 35] = "Ly2"
merged$clusters2[merged$SCT_snn_res.1.2 == 18] = "Ly3"
merged$clusters2[merged$SCT_snn_res.1.2 == 24] = "Ly4"
merged$clusters2[merged$SCT_snn_res.1.2 == 29] = "Neu"
merged$clusters2[merged$SCT_snn_res.1.2 == 41] = "Neu"
merged$clusters2[merged$SCT_snn_res.1.2 == 38] = "pDC"
merged$clusters2[merged$SCT_snn_res.1.2 == 43] = "BC"
merged$clusters2[merged$SCT_snn_res.1.2 == 32] = "BC"
merged$clusters2[merged$SCT_snn_res.1.2 == 14] = "BC"
merged$clusters2[merged$SCT_snn_res.1.2 == 23] = "cDC"
merged$clusters2[merged$SCT_snn_res.1.2 == 0] = "PVM1"
merged$clusters2[merged$SCT_snn_res.1.2 == 9] = "PVM1"
merged$clusters2[merged$SCT_snn_res.1.2 == 17] = "PVM2"
merged$clusters2[merged$SCT_snn_res.1.2 == 15] = "PVM3"
merged$clusters2[merged$SCT_snn_res.1.2 == 27] = "PVM3"
merged$clusters2[merged$SCT_snn_res.1.2 == 22] = "PVM4"

Idents(merged) = merged$clusters2
DimPlot_scCustom(seurat_object = merged, group.by = "clusters", 
                 label=T, pt.size = .1) + NoAxes()
```

```{r}
unk = FindMarkers(merged, ident.1 = 'UNK', only.pos = T)
write.table(unk[order(unk$avg_log2FC, decreasing = T),],'unk_markers.csv', 
            quote=F, sep='\t')
```

```{r}
merged@assays$RNA@data <- shifted_log_transform(merged@assays$RNA@counts, 
                                                overdispersion = 0.01, on_disk = TRUE)
DefaultAssay(merged) = "RNA"

merged.split <- SplitObject(merged, split.by = 'Condition')

passed = names(table(merged.split$Aneu$clusters)[table(merged.split$Aneu$clusters)>= 20])
passed = passed[passed!="LQ"]
merged.split$CTRL = subset(merged.split$CTRL, idents=passed)
merged.split$Aneu = subset(merged.split$Aneu, idents=passed)

cellchat.aneu <- createCellChat(object = merged.split$Aneu, group.by = "clusters", assay = "SCT")
cellchat.ctrl <- createCellChat(object = merged.split$CTRL, group.by = "clusters", assay = "SCT")
cellchat.aneu@DB <- CellChatDB.human
cellchat.ctrl@DB <- CellChatDB.human
cellchat.aneu <- subsetData(cellchat.aneu)
cellchat.ctrl <- subsetData(cellchat.ctrl)
cellchat.aneu <- identifyOverExpressedGenes(cellchat.aneu)
cellchat.aneu <- identifyOverExpressedInteractions(cellchat.aneu)
cellchat.aneu <- projectData(cellchat.aneu, PPI.human)
cellchat.ctrl <- identifyOverExpressedGenes(cellchat.ctrl)
cellchat.ctrl <- identifyOverExpressedInteractions(cellchat.ctrl)
cellchat.ctrl <- projectData(cellchat.ctrl, PPI.human)
cellchat.ctrl <- computeCommunProb(cellchat.ctrl)
cellchat.ctrl <- filterCommunication(cellchat.ctrl, min.cells = 20)
cellchat.aneu <- computeCommunProb(cellchat.aneu)
cellchat.aneu <- filterCommunication(cellchat.aneu, min.cells = 20)
cellchat.aneu <- computeCommunProbPathway(cellchat.aneu)
cellchat.ctrl <- computeCommunProbPathway(cellchat.ctrl)
cellchat.ctrl <- aggregateNet(cellchat.ctrl)
cellchat.aneu <- aggregateNet(cellchat.aneu)
cellchat.aneu <- netAnalysis_computeCentrality(cellchat.aneu, slot.name = "netP") 
cellchat.ctrl <- netAnalysis_computeCentrality(cellchat.ctrl, slot.name = "netP") 

object.list = list(cellchat.aneu,cellchat.ctrl)
cellchat <- mergeCellChat(object.list,
                          add.names = c("ANEU","CTRL"))

gg1 <- compareInteractions(cellchat, show.legend = F, group = c(1,2))
gg2 <- compareInteractions(cellchat, show.legend = F, group = c(1,2), measure = "weight")
gg1 + gg2

gg1 <- rankNet(cellchat, mode = "comparison", stacked = T, do.stat = TRUE)
gg2 <- rankNet(cellchat, mode = "comparison", stacked = F, do.stat = TRUE)
gg1 + gg2


pathways.show <- c("PERIOSTIN") 
weight.max <- getMaxWeight(object.list, slot.name = c("netP"), 
                           attribute = pathways.show) 
par(mfrow = c(1,1), xpd=TRUE)
for (i in 1:length(object.list)) {
  netVisual_aggregate(object.list[[i]], signaling = pathways.show, layout = "circle", 
                      edge.width.max = 10, signaling.name = paste(pathways.show, names(object.list)[i]))
}

num.link <- sapply(object.list, function(x) {rowSums(x@net$count) + colSums(x@net$count)-diag(x@net$count)})
weight.MinMax <- c(min(num.link), max(num.link)) # control the dot size in the different datasets
gg <- list()
for (i in 1:length(object.list)) {
  gg[[i]] <- netAnalysis_signalingRole_scatter(object.list[[i]], title = names(object.list)[i], weight.MinMax = weight.MinMax)
}
#> Signaling role analysis on the aggregated cell-cell communication network from all signaling pathways
#> Signaling role analysis on the aggregated cell-cell communication network from all signaling pathways
patchwork::wrap_plots(plots = gg)
```


```{r}
merged.split$Aneu$Scissor2 = merged.split$Aneu$Scissor
merged.split$Aneu$Scissor2[merged.split$Aneu$Scissor2 == "BG"] = "Neg"
merged.split.rupture <- SplitObject(merged.split$Aneu, split.by = 'Scissor2')
passed = names(table(merged.split.rupture$Pos$clusters)[table(merged.split.rupture$Pos$clusters)>= 20])
merged.split.rupture$Neg = subset(merged.split.rupture$Neg, idents=passed)
merged.split.rupture$Pos = subset(merged.split.rupture$Pos, idents=passed)

cellchat.rupt <- createCellChat(object = merged.split.rupture$Pos, group.by = "clusters")
cellchat.unrupt <- createCellChat(object = merged.split.rupture$Neg, group.by = "clusters")
cellchat.rupt@DB <- CellChatDB.human
cellchat.unrupt@DB <- CellChatDB.human
cellchat.rupt <- subsetData(cellchat.rupt)
cellchat.unrupt <- subsetData(cellchat.unrupt)
cellchat.rupt <- identifyOverExpressedGenes(cellchat.rupt)
cellchat.rupt <- identifyOverExpressedInteractions(cellchat.rupt)
cellchat.rupt <- projectData(cellchat.rupt, PPI.human)
cellchat.unrupt <- identifyOverExpressedGenes(cellchat.unrupt)
cellchat.unrupt <- identifyOverExpressedInteractions(cellchat.unrupt)
cellchat.unrupt <- projectData(cellchat.unrupt, PPI.human)
cellchat.unrupt <- computeCommunProb(cellchat.unrupt)
cellchat.unrupt <- filterCommunication(cellchat.unrupt, min.cells = 20)
cellchat.rupt <- computeCommunProb(cellchat.rupt)
cellchat.rupt <- filterCommunication(cellchat.rupt, min.cells = 20)
cellchat.rupt <- computeCommunProbPathway(cellchat.rupt)
cellchat.unrupt <- computeCommunProbPathway(cellchat.unrupt)
cellchat.unrupt <- aggregateNet(cellchat.unrupt)
cellchat.rupt <- aggregateNet(cellchat.rupt)
cellchat.rupt <- netAnalysis_computeCentrality(cellchat.rupt, slot.name = "netP") 
cellchat.unrupt <- netAnalysis_computeCentrality(cellchat.unrupt, slot.name = "netP") 

object.list2 = list(cellchat.rupt,cellchat.unrupt)
cellchat2 <- mergeCellChat(object.list2,
                          add.names = c("Pos","Neg"))

gg1 <- compareInteractions(cellchat2, show.legend = F, group = c(1,2))
gg2 <- compareInteractions(cellchat2, show.legend = F, group = c(1,2), measure = "weight")
gg1 + gg2

gg1 <- rankNet(cellchat2, mode = "comparison", stacked = T, do.stat = TRUE, font.size = 15)
gg2 <- rankNet(cellchat2, mode = "comparison", stacked = F, do.stat = TRUE, font.size = 15)
gg1 + gg2


pathways.show <- c("ICAM") 
weight.max <- getMaxWeight(object.list2, slot.name = c("netP"), 
                           attribute = pathways.show) 
par(mfrow = c(1,2), xpd=TRUE)
for (i in 1:length(object.list)) {
  netVisual_aggregate(object.list2[[i]], signaling = pathways.show, layout = "circle", 
                      edge.width.max = 10, signaling.name = paste(pathways.show, names(object.list2)[i]))
}


gg1 <- netAnalysis_signalingChanges_scatter(cellchat2, idents.use = "OC", signaling.exclude = c("SPP1","GALECTIN"))
gg2 <- netAnalysis_signalingChanges_scatter(cellchat2, idents.use = "Aneu1", signaling.exclude = c("FN1","COLLAGEN","SPP1", "MIF", "LAMININ"))
patchwork::wrap_plots(plots = list(gg1,gg2))
```

```{r}
merged.vasc2 = merged.vasc
merged3 = merge(merged.aneu, ctrl12)
cells = intersect(colnames(merged.vasc2), colnames(merged3))
merged3 = subset(merged3, cells=cells)

cells2 = setdiff(colnames(merged.vasc2), cells)
tmp = merged.vasc2@meta.data
merged.vasc2 = subset(merged.vasc2, cells=cells2)
merged.vasc2 = merge(merged.vasc2, merged3)
tmp = tmp[colnames(merged.vasc2),]
merged.vasc2$clusters = tmp$clusters
merged.vasc2$Scissor = tmp$Scissor
merged.vasc2$Condition = tmp$Condition
merged.vasc2$Sample = tmp$Sample
merged.vasc = merged.vasc2

merged.vasc = subset(merged2, idents=c("FB1","Aneu1","Aneu2","FB2","SMC2","PC"))

DefaultAssay(merged.vasc) = 'RNA'
merged.vasc = NormalizeData(merged.vasc)
merged.vasc = SCTransform(merged.vasc, vst.flavor='v2', ncells=NULL, n_genes=NULL)
VariableFeatures(merged.vasc) = VariableFeatures(merged.vasc)[!grepl("RPS", VariableFeatures(merged.vasc))] 
VariableFeatures(merged.vasc) = VariableFeatures(merged.vasc)[!grepl("RPL", VariableFeatures(merged.vasc))] 
VariableFeatures(merged.vasc) = VariableFeatures(merged.vasc)[!grepl("MT-", VariableFeatures(merged.vasc))] 
VariableFeatures(merged.vasc) = VariableFeatures(merged.vasc)[!grepl("MALAT1", VariableFeatures(merged.vasc))] 
merged.vasc@assays$SCT@scale.data = merged.vasc@assays$SCT@scale.data[VariableFeatures(merged.vasc),] 

merged.vasc = RunPCA(merged.vasc, npcs = 30)
merged.vasc = RunHarmony(merged.vasc, group.by.vars = c('Sample', "Condition"), 
                         plot_convergence = T, assay.use = "SCT")
merged.vasc = RunUMAP(merged.vasc, dims=1:30, reduction = 'harmony')

seuratToH5AD(merged.vasc, "aneu.h5ad", rownames(merged.vasc@assays$SCT@data))

DimPlot(merged.vasc, label=T, pt.size = 1, cols=dittoColors(),
        group.by = 'clusters') + NoAxes()
DimPlot(merged.vasc, label=T, pt.size = 1, cols=dittoColors(), 
        group.by = 'Condition') + NoAxes()
DimPlot(merged.vasc, label=T, pt.size = 1, cols=dittoColors(), 
        group.by = 'Scissor') + NoAxes()
FeaturePlot(merged.vasc, features = 'KCNJ8', order=T, max.cutoff = 'q99') + 
  NoAxes() + scale_color_viridis_c()

markers = wilcoxauc(merged.vasc, seurat_assay = 'SCT')
top_markers(markers)
```

```{r}
gwas_genes = c('SLC22A5','SLC22A4','P4HA2','SOX17','NT5C2','MARCKSL1P1',
                'FGD6','NR2C1','PSMA4','BCAR1','RP11.252K23.2')
gwas = list()
gwas$gwas = gwas_genes
gwas$background = setdiff(rownames(merged@assays$SCT@counts) ,gwas$gwas)

FeaturePlot(merged, features = gwas,max.cutoff = 'q99',ncol=3, order=T, pt.size = .1) &   NoAxes() & scale_color_viridis_c()

merged = AddModuleScore_UCell(merged, features = gwas, assay = 'SCT', maxRank = 34000)

FeaturePlot(merged, features = c("gwas_UCell"), max.cutoff = 'q99', order=T, 
            pt.size = .1, min.cutoff = 'q25') & NoAxes() & scale_color_viridis_c()
``` 

```{r}
library(SpatialDecon)
rows = readLines("/media/chang/HDD-1/chang/fastqs/aneu/spatial/gene_names.txt")

neg = read.table("/media/chang/HDD-1/chang/fastqs/aneu/spatial/Neg_probes.csv",
                 sep=',', header=T, row.names = 1)
rows = c(rows, neg$ProbeDisplayName)
raw = read.table("/media/chang/HDD-1/chang/fastqs/aneu/spatial/count_matrix_all.csv",
                     sep=',', header=T)
norm = RelativeCounts(raw, scale.factor = 1e6)
rownames(norm) = rows
rownames(raw) = rows
norm_neg = colMeans(norm[grepl("NegProbe",rownames(norm)),])
raw_neg = colMeans(raw[grepl("NegProbe",rownames(raw)),])
raw = raw[!grepl("NegProbe", rownames(raw)),]
norm = norm[!grepl("NegProbe", rownames(norm)),]
norm = rbind(norm, norm_neg)
raw = rbind(raw, raw_neg)
rownames(norm)[length(rownames(norm))] = "NegProbe"
rownames(raw)[length(rownames(raw))] = "NegProbe"
norm = as.matrix(norm)
raw = as.matrix(raw)


bg2 = derive_GeoMx_background(norm = as.matrix(norm),
                             probepool = rep(1, nrow(norm)),
                             negnames = "NegProbe")



merged.vasc3 = subset(merged2, idents=c("FB1","Aneu1","Aneu2","FB2","SMC1","PC","BC","Ly1","Ly2",
                                        "FBMC1","FBMC2","FBMC3","SMC2","OC","EC1","OL","Ly3",
                                        "EC2","EC3","PVM1","PVM2","PVM3","PVM4",'cDC','pDC',"Ly4"))

merged.vasc3 = subset(merged2, idents=c("FB1","Aneu1","Aneu2","FB2","SMC1","PC",
                                        "FBMC1","FBMC2","FBMC3","SMC2","EC1",
                                        "EC2","EC3"))

merged.vasc3$class = merged.vasc3$clusters
merged.vasc3$class[grepl("SMC", merged.vasc3$clusters)] = "SMC"
merged.vasc3$class[grepl("FB", merged.vasc3$clusters)] = "FB"
merged.vasc3$class[grepl("FBMC", merged.vasc3$clusters)] = "FBMC"
merged.vasc3$class[grepl("EC", merged.vasc3$clusters)] = "EC"

merged.vasc3$class[grepl("PVM", merged.vasc3$clusters)] = "PVM"
merged.vasc3$class[grepl("cDC", merged.vasc3$clusters)] = "PVM"
merged.vasc3$class[grepl("Ly", merged.vasc3$clusters)] = "Ly"
#merged.vasc3$class[grepl("Aneu", merged.vasc3$clusters)] = "Aneu"

rows = readLines("/media/chang/HDD-1/chang/fastqs/aneu/spatial/gene_names.txt")
rows = intersect(rows, rownames(merged2@assays$SCT@data))
merged.vasc3$cellID = colnames(merged.vasc3)
#merged2$cellID = colnames(merged2)

custom_mtx_seurat <- create_profile_matrix(mtx = merged.vasc3@assays$RNA@counts[rows,], 
                                           cellAnnots = merged.vasc3@meta.data, 
                                           cellTypeCol = "class", 
                                           cellNameCol = "cellID", 
                                           matrixName = "vasc",
                                           outDir = NULL, 
                                           normalize = FALSE, 
                                           minCellNum = 5, 
                                           minGenes = 10)

heatmap(sweep(custom_mtx_seurat, 1, apply(custom_mtx_seurat, 1, max), "/"),
        labRow = NA, margins = c(10, 5))

nuclei = read.table("/media/chang/HDD-1/chang/fastqs/aneu/spatial/SegmentProperties.csv",
                    sep=',')

meta = read.table("/media/chang/HDD-1/chang/fastqs/aneu/spatial/Add_Meta.csv",
                    sep=',')

# vector identifying pure tumor segments:
meta$istumor = (meta$CONDITION == "Aneurysm")
nuclei$AOINucleiCount[34] = 34
#nuclei$AOINucleiCount = nuclei$AOINucleiCount *  5
# run spatialdecon with all the bells and whistles:
restils = spatialdecon(norm = norm,                     # normalized data
                       raw = raw,                       # raw data, used to down-weight low-count observations 
                       bg = bg2,                         # expected background counts for every data point in norm
                       X = custom_mtx_seurat,
                       maxit=10000,
                       cell_counts = nuclei$AOINucleiCount)     # nuclei counts, used to estimate total cells
                       #is_pure_tumor = meta$istumor,   # identities of the Tumor segments/observations
                       #n_tumor_clusters = 1)            # how many distinct tumor profiles to append to safeTME


heatmap(sweep(restils$X, 1, apply(restils$X, 1, max), "/"),
         labRow = NA, margins = c(10, 5))

layout(mat = (matrix(c(1, 2), 1)), widths = c(7, 3))

TIL_barplot(restils$cell.counts$cell.counts, draw_legend = TRUE, 
            cex.names = 0.5)


matching = list()
matching$SMC = c( "SMC1", "SMC2")
#matching$Ly = c("Ly1","Ly2", "Ly3", "Ly4")
#matching$BC = c("BC")
matching$OC = c("OC")
matching$OL = c("OL")
matching$PC = c("PC")
matching$pDC = c("pDC")
matching$AC = c("AC")
matching$Aneu1 = c("Aneu1")
matching$Aneu2 = c("Aneu2")
matching$EC = c("EC1","EC2","EC3")
matching$PVM = c("PVM1","PVM2","PVM3","PVM4","cDC")
matching$FB = c("FB1", "FB2")
matching$FBMC = c("FBMC1", "FBMC2","FBMC3")

collapsed = collapseCellTypes(fit = restils, 
                              matching = matching)

heatmap(collapsed$beta, cexRow = 0.85, cexCol = 0.75)



res = spatialdecon(norm = norm,
                   bg = bg2,
                   X = custom_mtx_seurat,
                   align_genes = TRUE)
heatmap(res$beta, cexCol = 0.5, cexRow = 0.7, margins = c(10,7))


rdecon = reverseDecon(norm = norm,
                      beta = restils$beta)
# look at the residuals:
heatmap(pmax(pmin(rdecon$resids, 2), -2))


```

```{r}
ctrl1 = read.table("/media/chang/HDD-1/chang/fastqs/aneu/spatial/count_matrix_ctrl1.csv",
                     sep=',', header=T)
ctrl2 = read.table("/media/chang/HDD-1/chang/fastqs/aneu/spatial/count_matrix_ctrl2.csv",
                     sep=',', header=T)
aneu1 = read.table("/media/chang/HDD-1/chang/fastqs/aneu/spatial/count_matrix_aneu1.csv",
                     sep=',', header=T)
aneu2 = read.table("/media/chang/HDD-1/chang/fastqs/aneu/spatial/count_matrix_aneu2.csv",
                     sep=',', header=T)

bk = list(ctrl1[,c(1:4)], ctrl1[,c(5:8)], ctrl1[,c(9:12)], ctrl2[,c(1:4)], ctrl2[,c(5:8)], 
          aneu1[,c(1:4)], aneu1[,c(5:8)], aneu1[,c(9:12)], aneu2)


bk.ctrl = data.frame(ctrl1, ctrl2)
bk.aneu = data.frame(aneu1, aneu2)
bk = list(bk.ctrl, bk.aneu)

merged.vasc3 = subset(merged2, idents=c("FB1","Aneu1","Aneu2","FB2","SMC1","pDC",
                                        "FBMC1","FBMC2","FBMC3","SMC2","OC","EC1",
                                        "EC2","EC3","PVM1","PVM2","PVM3","PVM4","cDC"))

merged.vasc3 = subset(merged2, idents=c("FB1","Aneu1","Aneu2","FB2","SMC1",
                                        "FBMC1","FBMC2","FBMC3","SMC2","EC1",
                                        "EC2","EC3",'PC'))

#merged.vasc3 = merged2
merged.vasc3$class = merged.vasc3$clusters
merged.vasc3$class[grepl("SMC", merged.vasc3$clusters)] = "SMC"
merged.vasc3$class[grepl("PC", merged.vasc3$clusters)] = "SMC"
merged.vasc3$class[grepl("FB", merged.vasc3$clusters)] = "FB"
merged.vasc3$class[grepl("FBMC", merged.vasc3$clusters)] = "FBMC"
merged.vasc3$class[grepl("EC", merged.vasc3$clusters)] = "EC"


merged.vasc3$class[grepl("PVM", merged.vasc3$clusters)] = "PVM"
merged.vasc3$class[grepl("PVM3", merged.vasc3$clusters)] = "PVM3"

rows = readLines("/media/chang/HDD-1/chang/fastqs/aneu/spatial/gene_names.txt")
inter = intersect(rows, rownames(merged.vasc3@assays$RNA@counts))

sc.dat = as.matrix(t(merged.vasc3@assays$RNA@counts[inter,]))
sc.dat.filtered <- cleanup.genes (input=sc.dat,
                                  input.type="count.matrix",
                                  species="hs", 
                                  gene.group=c("Rb","Mrp","other_Rb","chrM","MALAT1","chrX","chrY"),
                                  exp.cells=5)
rm(sc.dat);gc()

inter = intersect(rows, colnames(sc.dat.filtered))


# merged
###
cells = merged.vasc3@meta.data[merged.vasc3$class %in% 
                                 c("FB","SMC","EC","FBMC","PC"),]
cells = cells[cells$Condition %in% "CTRL",]
bk.ctrl = data.frame(ctrl1, ctrl2)
rownames(bk.ctrl) = rows
bk.ctrl = as.matrix(t(bk.ctrl[inter,]))
myPrism <- new.prism(
  reference=sc.dat.filtered[rownames(cells),inter], 
  mixture=bk.ctrl,
  input.type="count.matrix", 
  cell.type.labels = cells$class, 
  cell.state.labels = cells$clusters,
  outlier.cut=0.01,
  outlier.fraction=0.1,
  key=NULL
)
gc()
bp.res <- run.prism(prism = myPrism, n.cores=50)
bp.res

cells = merged.vasc3@meta.data[merged.vasc3$class %in% 
                                 c("FB","EC","FBMC","SMC","PC", "Aneu1","Aneu2"),]
cells = cells[cells$Condition %in% "Aneu",]
bk.aneu = data.frame(aneu1, aneu2)
rownames(bk.aneu) = rows
bk.aneu = as.matrix(t(bk.aneu[inter,]))
myPrism <- new.prism(
  reference=sc.dat.filtered[rownames(cells),inter], 
  mixture=bk.aneu,
  input.type="count.matrix", 
  cell.type.labels = cells$class, 
  cell.state.labels = cells$clusters,
  outlier.cut=0.01,
  outlier.fraction=0.1,
  key=NULL,
)
gc()
bp.res2 <- run.prism(prism = myPrism, n.cores=50)
bp.res2

cells = merged.vasc3@meta.data[merged.vasc3$class %in% 
                                 c("PVM",'pDC','cDC', "PVM3"),]
cells = cells[cells$Condition %in% "CTRL",]
bk.aneu = data.frame(aneu1, aneu2)
rownames(bk.aneu) = rows
bk.aneu = as.matrix(t(bk.aneu[inter,]))
myPrism <- new.prism(
  reference=sc.dat.filtered[rownames(cells),inter], 
  mixture=bk.aneu,
  input.type="count.matrix", 
  cell.type.labels = cells$class, 
  cell.state.labels = cells$clusters,
  outlier.cut=0.01,
  outlier.fraction=0.1,
  key=NULL,
)
gc()
bp.res3 <- run.prism(prism = myPrism, n.cores=50)
bp.res3


cells = merged.vasc3@meta.data[merged.vasc3$class %in% 
                                 c("PVM","OC",'pDC','cDC',"PVM3"),]
cells = cells[cells$Condition %in% "Aneu",]
bk.aneu = data.frame(aneu1, aneu2)
rownames(bk.aneu) = rows
bk.aneu = as.matrix(t(bk.aneu[inter,]))
myPrism <- new.prism(
  reference=sc.dat.filtered[rownames(cells),inter], 
  mixture=bk.aneu,
  input.type="count.matrix", 
  cell.type.labels = cells$class, 
  cell.state.labels = cells$clusters,
  outlier.cut=0.01,
  outlier.fraction=0.1,
  key=NULL,
)
gc()
bp.res4 <- run.prism(prism = myPrism, n.cores=50)
bp.res4
###################################

theta <- get.fraction(bp=bp.res,
            which.theta="final",
            state.or.type="state")

theta2 <- get.fraction(bp=bp.res2,
            which.theta="final",
            state.or.type="state")

theta3 <- get.fraction(bp=bp.res3,
            which.theta="final",
            state.or.type="state")

theta4 <- get.fraction(bp=bp.res4,
            which.theta="final",
            state.or.type="state")

ctrl_avg = colMeans(theta)
ctrl_avg["Aneu1"] = 0
ctrl_avg["Aneu2"] = 0
aneu_avg = colMeans(theta2)
avg_df = t(data.frame(ctrl_avg, aneu_avg))

pdf('nano_bar_prop.pdf', width = 3, height = 7)
ggplot(melt(avg_df,id.vars = 1),aes(x=Var1,y=value,fill=Var2)) + 
  geom_bar(position = "fill", stat = "identity") + theme_bw() + 
  scale_fill_manual(values=dittoColors()) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size=15),legend.text =element_text(size=15))
dev.off()

ctrl_avg = colMeans(theta3)
ctrl_avg["OC"] = 0
aneu_avg = colMeans(theta4)
avg_df = t(data.frame(ctrl_avg, aneu_avg))

pdf('nano_bar_immune_prop.pdf', width = 3, height = 7)
ggplot(melt(avg_df,id.vars = 1),aes(x=Var1,y=value,fill=Var2)) + 
  geom_bar(position = "fill", stat = "identity") + theme_bw() + 
  scale_fill_manual(values=dittoColors()[7:11]) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size=15),legend.text =element_text(size=15))
dev.off()




ggplot(melt(theta,id.vars = 1),aes(x=Var1,y=value,fill=Var2)) + 
  geom_bar(position = "fill", stat = "identity") + theme_bw() + 
  scale_fill_manual(values=dittoColors()) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size=15),legend.text =element_text(size=15))

ggplot(melt(theta2,id.vars = 1),aes(x=Var1,y=value,fill=Var2)) + 
  geom_bar(position = "fill", stat = "identity") + theme_bw() + 
  scale_fill_manual(values=dittoColors()) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size=15),legend.text =element_text(size=15))

ggplot(melt(theta3,id.vars = 1),aes(x=Var1,y=value,fill=Var2)) + 
  geom_bar(position = "fill", stat = "identity") + theme_bw() + 
  scale_fill_manual(values=dittoColors()) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size=15),legend.text =element_text(size=15))

ggplot(melt(theta4,id.vars = 1),aes(x=Var1,y=value,fill=Var2)) + 
  geom_bar(position = "fill", stat = "identity") + theme_bw() + 
  scale_fill_manual(values=dittoColors()) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size=15),legend.text =element_text(size=15))
####
###
aneu.expr <- data.frame(get.exp(bp=bp.res2,
                          state.or.type="type",
                          cell.name="Aneu1"))
aneu.meta = expand.grid(Area = c("EC", "M1", "M2", "Adv"),
                  replicates = c(1, 2, 3, 4))
aneu.expr$Area = aneu.meta$Area
#aneu.expr$Rep = as.character(aneu.meta$replicates)
aneu.expr2 = aneu.expr[,c("POSTN","THY1","CLDN5","ACTA2", "DCN","Area")]

ggplot(aneu.expr, aes(x=Area, y=POSTN, color=Rep)) + geom_point() +
  geom_smooth(method = "loess", se = F, span = 2,size=2,aes(color = Rep)) + theme_bw()

aneu.expr2 = melt(aneu.expr2, id="Area")  
#aneu.expr = aneu.expr[aneu.expr$variable != "Rep",]
p1 = ggplot(aneu.expr2, aes(x=Area, y=value, color=variable)) + geom_point() + ylim(0,40) +
  geom_smooth(method = "loess", se = F, span = 2,size=2,aes(color = variable)) + theme_bw()



aneu.expr <- data.frame(get.exp(bp=bp.res2,
                          state.or.type="type",
                          cell.name="Aneu2"))
aneu.meta = expand.grid(Area = c("EC", "M1", "M2", "Adv"),
                  replicates = c(1, 2, 3, 4))
aneu.expr$Area = aneu.meta$Area
#aneu.expr$Rep = as.character(aneu.meta$replicates)
aneu.expr2 = aneu.expr[,c("POSTN","THY1","CLDN5","ACTA2", "DCN","Area")]

ggplot(aneu.expr, aes(x=Area, y=POSTN, color=Rep)) + geom_point() +
  geom_smooth(method = "loess", se = F, span = 2,size=2,aes(color = Rep)) + theme_bw()

aneu.expr2 = melt(aneu.expr2, id="Area")  
#aneu.expr = aneu.expr[aneu.expr$variable != "Rep",]
p2 = ggplot(aneu.expr2, aes(x=Area, y=value, color=variable)) + geom_point() + ylim(0,40) +
  geom_smooth(method = "loess", se = F, span = 2,size=2,aes(color = variable)) + theme_bw()


aneu.expr <- data.frame(get.exp(bp=bp.res2,
                          state.or.type="type",
                          cell.name="SMC"))
aneu.meta = expand.grid(Area = c("EC", "M1", "M2", "Adv"),
                  replicates = c(1, 2, 3, 4))
aneu.expr$Area = aneu.meta$Area
#aneu.expr$Rep = as.character(aneu.meta$replicates)
aneu.expr2 = aneu.expr[,c("POSTN","THY1","CLDN5","ACTA2", "DCN","Area")]

ggplot(aneu.expr, aes(x=Area, y=POSTN, color=Rep)) + geom_point() +
  geom_smooth(method = "loess", se = F, span = 2,size=2,aes(color = Rep)) + theme_bw()

aneu.expr2 = melt(aneu.expr2, id="Area")  
#aneu.expr = aneu.expr[aneu.expr$variable != "Rep",]
p3 = ggplot(aneu.expr2, aes(x=Area, y=value, color=variable)) + geom_point() + ylim(0,40) +
  geom_smooth(method = "loess", se = F, span = 2,size=2,aes(color = variable)) + theme_bw()


aneu.expr <- data.frame(get.exp(bp=bp.res2,
                          state.or.type="type",
                          cell.name="FB"))
aneu.meta = expand.grid(Area = c("EC", "M1", "M2", "Adv"),
                  replicates = c(1, 2, 3, 4))
aneu.expr$Area = aneu.meta$Area
#aneu.expr$Rep = as.character(aneu.meta$replicates)
aneu.expr2 = aneu.expr[,c("POSTN","THY1","CLDN5","ACTA2","DCN", "Area")]

ggplot(aneu.expr, aes(x=Area, y=POSTN, color=Rep)) + geom_point() +
  geom_smooth(method = "loess", se = F, span = 2,size=2,aes(color = Rep)) + theme_bw()

aneu.expr2 = melt(aneu.expr2, id="Area")  
#aneu.expr = aneu.expr[aneu.expr$variable != "Rep",]
p4 = ggplot(aneu.expr2, aes(x=Area, y=value, color=variable)) + geom_point() + ylim(0,40) +
  geom_smooth(method = "loess", se = F, span = 2,size=2,aes(color = variable)) + theme_bw()


aneu.expr <- data.frame(get.exp(bp=bp.res2,
                          state.or.type="type",
                          cell.name="FBMC"))
aneu.meta = expand.grid(Area = c("EC", "M1", "M2", "Adv"),
                  replicates = c(1, 2, 3, 4))
aneu.expr$Area = aneu.meta$Area
#aneu.expr$Rep = as.character(aneu.meta$replicates)
aneu.expr2 = aneu.expr[,c("POSTN","THY1","CLDN5","ACTA2","DCN", "Area")]

ggplot(aneu.expr, aes(x=Area, y=POSTN, color=Rep)) + geom_point() +
  geom_smooth(method = "loess", se = F, span = 2,size=2,aes(color = Rep)) + theme_bw()

aneu.expr2 = melt(aneu.expr2, id="Area")  
#aneu.expr = aneu.expr[aneu.expr$variable != "Rep",]
p5 = ggplot(aneu.expr2, aes(x=Area, y=value, color=variable)) + geom_point() + ylim(0,40) +
  geom_smooth(method = "loess", se = F, span = 2,size=2,aes(color = variable)) + theme_bw()

aneu.expr <- data.frame(get.exp(bp=bp.res2,
                          state.or.type="type",
                          cell.name="PC"))
aneu.meta = expand.grid(Area = c("EC", "M1", "M2", "Adv"),
                  replicates = c(1, 2, 3, 4))
aneu.expr$Area = aneu.meta$Area
#aneu.expr$Rep = as.character(aneu.meta$replicates)
aneu.expr2 = aneu.expr[,c("POSTN","THY1","CLDN5","ACTA2","DCN", "Area")]

ggplot(aneu.expr, aes(x=Area, y=POSTN, color=Rep)) + geom_point() +
  geom_smooth(method = "loess", se = F, span = 2,size=2,aes(color = Rep)) + theme_bw()

aneu.expr2 = melt(aneu.expr2, id="Area")  
#aneu.expr = aneu.expr[aneu.expr$variable != "Rep",]
p6 = ggplot(aneu.expr2, aes(x=Area, y=value, color=variable)) + geom_point() + ylim(0,40) +
  geom_smooth(method = "loess", se = F, span = 2,size=2,aes(color = variable)) + theme_bw()


aneu.expr <- data.frame(get.exp(bp=bp.res2,
                          state.or.type="type",
                          cell.name="EC"))
aneu.meta = expand.grid(Area = c("EC", "M1", "M2", "Adv"),
                  replicates = c(1, 2, 3, 4))
aneu.expr$Area = aneu.meta$Area
#aneu.expr$Rep = as.character(aneu.meta$replicates)
aneu.expr2 = aneu.expr[,c("POSTN","THY1","CLDN5","ACTA2", "Area")]

ggplot(aneu.expr, aes(x=Area, y=POSTN, color=Rep)) + geom_point() +
  geom_smooth(method = "loess", se = F, span = 2,size=2,aes(color = Rep)) + theme_bw()

aneu.expr2 = melt(aneu.expr2, id="Area")  
#aneu.expr = aneu.expr[aneu.expr$variable != "Rep",]
p3 = ggplot(aneu.expr2, aes(x=Area, y=value, color=variable)) + geom_point() + ylim(0,40) +
  geom_smooth(method = "loess", se = F, span = 2,size=2,aes(color = variable)) + theme_bw()

###########

ctrl.expr <- data.frame(get.exp(bp=bp.res,
                        state.or.type="type",
                        cell.name="FB"))
ctrl.meta = expand.grid(Area = c("EC", "M1", "M2", "Adv"),
                        replicates = c(1, 2, 3, 4, 5))
ctrl.expr$Area = ctrl.meta$Area
#ctrl.expr$Rep = ctrl.meta$replicates
ctrl.expr2 = ctrl.expr[,c("POSTN","THY1","CLDN5", "Area", "PECAM1","ACTA2","DCN")]

p0 = ggplot(ctrl.expr, aes(x=Area, y=POSTN, color=Rep)) + geom_point() +
  geom_smooth(method = "loess", se = F, span = 2,size=2,aes(color = Rep)) + theme_bw()

ctrl.expr2 = melt(ctrl.expr2, id="Area")  
p1 = ggplot(ctrl.expr2, aes(x=Area, y=value, color=variable)) + geom_point() + ylim(0,40) +
  geom_smooth(method = "loess", se = F, span = 2,size=2,aes(color = variable)) + theme_bw()



ctrl.expr <- data.frame(get.exp(bp=bp.res,
                        state.or.type="type",
                        cell.name="FBMC"))
ctrl.meta = expand.grid(Area = c("EC", "M1", "M2", "Adv"),
                        replicates = c(1, 2, 3, 4, 5))
ctrl.expr$Area = ctrl.meta$Area
#ctrl.expr$Rep = ctrl.meta$replicates
ctrl.expr2 = ctrl.expr[,c("POSTN","THY1","CLDN5", "Area", "PECAM1","ACTA2","DCN")]

p0 = ggplot(ctrl.expr, aes(x=Area, y=POSTN, color=Rep)) + geom_point() +
  geom_smooth(method = "loess", se = F, span = 2,size=2,aes(color = Rep)) + theme_bw()

ctrl.expr2 = melt(ctrl.expr2, id="Area")  
p2 = ggplot(ctrl.expr2, aes(x=Area, y=value, color=variable)) + geom_point() + ylim(0,40) +
  geom_smooth(method = "loess", se = F, span = 2,size=2,aes(color = variable)) + theme_bw()

ctrl.expr <- data.frame(get.exp(bp=bp.res,
                        state.or.type="type",
                        cell.name="SMC"))
ctrl.meta = expand.grid(Area = c("EC", "M1", "M2", "Adv"),
                        replicates = c(1, 2, 3, 4, 5))
ctrl.expr$Area = ctrl.meta$Area
#ctrl.expr$Rep = ctrl.meta$replicates
ctrl.expr2 = ctrl.expr[,c("POSTN","THY1","CLDN5", "Area", "PECAM1","ACTA2","DCN")]

p0 = ggplot(ctrl.expr, aes(x=Area, y=POSTN, color=Rep)) + geom_point() +
  geom_smooth(method = "loess", se = F, span = 2,size=2,aes(color = Rep)) + theme_bw()

ctrl.expr2 = melt(ctrl.expr2, id="Area")  
p3 = ggplot(ctrl.expr2, aes(x=Area, y=value, color=variable)) + geom_point() + ylim(0,40) +
  geom_smooth(method = "loess", se = F, span = 2,size=2,aes(color = variable)) + theme_bw()

ctrl.expr <- data.frame(get.exp(bp=bp.res,
                        state.or.type="type",
                        cell.name="PC"))
ctrl.meta = expand.grid(Area = c("EC", "M1", "M2", "Adv"),
                        replicates = c(1, 2, 3, 4, 5))
ctrl.expr$Area = ctrl.meta$Area
#ctrl.expr$Rep = ctrl.meta$replicates
ctrl.expr2 = ctrl.expr[,c("POSTN","THY1","CLDN5", "Area", "PECAM1","ACTA2","DCN")]

p0 = ggplot(ctrl.expr, aes(x=Area, y=POSTN, color=Rep)) + geom_point() +
  geom_smooth(method = "loess", se = F, span = 2,size=2,aes(color = Rep)) + theme_bw()

ctrl.expr2 = melt(ctrl.expr2, id="Area")  
p4 = ggplot(ctrl.expr2, aes(x=Area, y=value, color=variable)) + geom_point() + ylim(0,40) +
  geom_smooth(method = "loess", se = F, span = 2,size=2,aes(color = variable)) + theme_bw()


ggplot(df, aes(x = "", y = perc, fill = answer)) +
  geom_col() +
  coord_polar(theta = "y")

##########################################

cells = get.exp(bp=bp.res,
        state.or.type="type", cell.name = c("SMC","FB","FBMC","EC"))
mtx = apply(cells, 2, function(i) i)
rn = sapply(c("SMC", "FB", "FBMC", "EC"), function(i) paste0(i,"_",rownames(cells[,,1])))
rownames(mtx) = rn
ctrl_wta = CreateSeuratObject(t(mtx))
ctrl.meta = expand.grid(Area = c("EC", "M1", "M2", "Adv"),
                        replicates = c(1:20))
ctrl_wta$Area = ctrl.meta$Area
ctrl_wta$Area_num = as.numeric(ctrl_wta$Area)
ctrl_wta$Replicate = substr(unlist(lapply(strsplit(rownames(ctrl_wta@meta.data), "_"), function(i) i[[2]])), 2,3)
ctrl_wta$ROI = substr(unlist(lapply(strsplit(rownames(ctrl_wta@meta.data), "_"), function(i) i[[2]])), 8,9)
ctrl_wta$ROI = str_remove(ctrl_wta$ROI, "^0+")
ctrl_wta$map = paste0(ctrl_wta$Replicate, "_", ctrl_wta$ROI)

nuclei$map = paste0(nuclei$SlideName, "_", nuclei$ROILabel)

new_meta = merge(ctrl_wta@meta.data, nuclei, by.x = "map", by.y = "map")
ctrl_wta@meta.data$ROI_counts = new_meta$AOINucleiCount
ctrl_wta@assays$RNA@data = ctrl_wta@assays$RNA@counts * (ctrl_wta$ROI_counts / max(ctrl_wta$ROI_counts))
ctrl_wta@assays$RNA@data = ctrl_wta@assays$RNA@counts

FeatureScatter(ctrl_wta, feature1 = "Area_num", feature2 = "CLDN5", pt.size = .5, 
               cols=dittoColors()) + geom_smooth(method = "loess", se = F, span = 2,size=2,
                                                 aes(color = ctrl_wta$orig.ident))

cells = get.exp(bp=bp.res3,
         state.or.type="type", cell.name = c("PVM","PVM3","cDC","pDC"))
mtx = apply(cells, 2, function(i) i)
rn = sapply(c("PVM", "PVM3", "cDC", "pDC"), function(i) paste0(i,"_",rownames(cells[,,1])))
rownames(mtx) = rn
ctrl_wta2 = CreateSeuratObject(t(mtx))
ctrl.meta2 = expand.grid(Area = c("EC", "M1", "M2", "Adv"),
                        replicates = c(1:16))
ctrl_wta2$Area = ctrl.meta2$Area
ctrl_wta2$Area_num = as.numeric(ctrl_wta2$Area)

FeatureScatter(ctrl_wta2, feature1 = "Area_num", feature2 = "ACP5", pt.size = .5, 
               cols=dittoColors()) + geom_smooth(method = "loess", se = F, span = 2,size=2,
                                                 aes(color = ctrl_wta2$orig.ident))

cells = get.exp(bp=bp.res2,
        state.or.type="type", cell.name = c("SMC","FB","FBMC","EC", "Aneu1", "Aneu2"))
mtx = apply(cells,2, function(i) i)
rn = sapply(c("SMC", "FB", "FBMC", "EC", "Aneu1", "Aneu2"), function(i) paste0(i,"_",rownames(cells[,,1])))
rownames(mtx) = rn
aneu_wta = CreateSeuratObject(t(mtx))
aneu.meta = expand.grid(Area = c("EC", "M1", "M2", "Adv"),
                        replicates = c(1:24))
aneu_wta$Area = aneu.meta$Area
aneu_wta$Area_num = as.numeric(aneu_wta$Area)
aneu_wta$Replicate = substr(unlist(lapply(strsplit(rownames(aneu_wta@meta.data), "_"), function(i) i[[2]])), 2,3)
aneu_wta$ROI = substr(unlist(lapply(strsplit(rownames(aneu_wta@meta.data), "_"), function(i) i[[2]])), 8,9)
aneu_wta$ROI = str_remove(aneu_wta$ROI, "^0+")
aneu_wta$map = paste0(aneu_wta$Replicate, "_", aneu_wta$ROI)

nuclei$map = paste0(nuclei$SlideName, "_", nuclei$ROILabel)

new_meta = merge(aneu_wta@meta.data, nuclei, by.x = "map", by.y = "map")
aneu_wta@meta.data$ROI_counts = new_meta$AOINucleiCount
aneu_wta@assays$RNA@data = aneu_wta@assays$RNA@counts * (aneu_wta$ROI_counts / max(aneu_wta$ROI_counts))
aneu_wta@assays$RNA@data = aneu_wta@assays$RNA@counts

FeatureScatter(aneu_wta, feature1 = "Area_num", feature2 = "POSTN", pt.size = .5, 
               cols=dittoColors()) + geom_smooth(method = "loess", se = F, span = 2,size=2,
                                                 aes(color = aneu_wta$orig.ident))

cells = get.exp(bp=bp.res4,
         state.or.type="type", cell.name = c("PVM","PVM3","cDC","pDC", "OC"))
mtx = apply(cells,2, function(i) i)
rn = sapply(c("PVM","PVM3","cDC","pDC", "OC"), function(i) paste0(i,"_",rownames(cells[,,1])))
rownames(mtx) = rn
aneu_wta2 = CreateSeuratObject(t(mtx))
aneu.meta2 = expand.grid(Area = c("EC", "M1", "M2", "Adv"),
                        replicates = c(1:20))
aneu_wta2$Area = aneu.meta2$Area
aneu_wta2$Area_num = as.numeric(aneu_wta2$Area)

FeatureScatter(aneu_wta2, feature1 = "Area_num", feature2 = "ACP5", pt.size = .5, 
               cols=dittoColors()) + geom_smooth(method = "loess", se = F, span = 2,size=2,
                                                 aes(color = aneu_wta2$orig.ident))


ctrl_wta$Condition = "CTRL"
aneu_wta$Condition = "Aneu"

merged_wta = merge(ctrl_wta, aneu_wta)
merged_wta@assays$RNA@data = merged_wta@assays$RNA@counts * 
  (merged_wta$ROI_counts / max(merged_wta$ROI_counts))

merged_wta = NormalizeData(merged_wta, scale.factor = 1e6)


top25 = head(aneu.markers[order(aneu.markers$avg_log2FC, decreasing = T),], 25)
bot25 = head(aneu.markers[order(aneu.markers$avg_log2FC, decreasing = F),], 25)

merged_wta@assays$RNA@data = merged_wta@assays$RNA@counts
merged_wta <- AddModuleScore(merged_wta, features = list(top25$gene, bot25$gene), 
                              name = c("Uncorrected_Aneu", "Uncorrected_Aneu"))
merged_wta <- AddModuleScore(merged_wta, features = list(rownames(up), rownames(down)), 
                              name = c("Unrupture", "Rupture"))

FeatureScatter(merged_wta, feature1 = "Area_num", feature2 = "POSTN", pt.size = 2,group.by = 'orig.ident',
               cols=dittoColors()) + geom_smooth(method = "loess", se = F, span = 2,size=2,
                                                 aes(color = merged_wta$orig.ident))

p1 = FeatureScatter(merged_wta, feature1 = "Area_num", feature2 = "Unrupture1", pt.size = 2,
               group.by = 'orig.ident',cols=dittoColors(), jitter = T) + 
  geom_smooth(method = "loess", se = F, span = 2,size=2, aes(color = merged_wta$orig.ident))
p2 = FeatureScatter(merged_wta, feature1 = "Area_num", feature2 = "Rupture1", pt.size = 2,
               group.by = 'orig.ident',cols=dittoColors(), jitter = T) + 
  geom_smooth(method = "loess", se = F, span = 2,size=2, aes(color = merged_wta$orig.ident))




p1 = FeatureScatter(merged_wta, feature1 = "Area_num", feature2 = "Uncorrected_Aneu1", pt.size = 2,
               group.by = 'orig.ident',cols=dittoColors(), jitter = T) + 
  geom_smooth(method = "loess", se = F, span = 2,size=2, aes(color = merged_wta$orig.ident))
p2 = FeatureScatter(merged_wta, feature1 = "Area_num", feature2 = "Uncorrected_Aneu2", pt.size = 2,
               group.by = 'orig.ident',cols=dittoColors(), jitter = T) + 
  geom_smooth(method = "loess", se = F, span = 2,size=2, aes(color = merged_wta$orig.ident))

pdf("nano_deconv_sig.pdf", width = 14)
p1+p2
dev.off()

p1 = FeatureScatter(merged_wta, feature1 = "Area_num", feature2 = "POSTN", pt.size = 2,
               group.by = 'orig.ident',cols=dittoColors(), jitter = T) + 
  geom_smooth(method = "loess", se = F, span = 2,size=2, aes(color = merged_wta$orig.ident))
p2 = FeatureScatter(merged_wta, feature1 = "Area_num", feature2 = "THY1", pt.size = 2,
               group.by = 'orig.ident',cols=dittoColors(), jitter = T) + 
  geom_smooth(method = "loess", se = F, span = 2,size=2, aes(color = merged_wta$orig.ident))

pdf("nano_deconv_markers.pdf", width = 14)
p1+p2
dev.off()

p1 = VlnPlot(merged_wta, features = c("CNN1","CLDN5","DCN","CCL19","MYH11","LUM"),
             cols = dittoColors())
pdf("nano_cell_markers.pdf", width = 14)
p1
dev.off()



ctrl_wta2$Condition = "CTRL"
aneu_wta2$Condition = "Aneu"

merged_wta2 = merge(ctrl_wta2, aneu_wta2)

p1 = FeatureScatter(merged_wta2, feature1 = "Area_num", feature2 = "ACP5", pt.size = 2,group.by = 'orig.ident',
               cols=dittoColors()) + geom_smooth(method = "loess", se = F, span = 2,size=2,
                                                 aes(color = merged_wta2$orig.ident))

p2 = FeatureScatter(merged_wta2, feature1 = "Area_num", feature2 = "SPP1", pt.size = 2,group.by = 'orig.ident',
               cols=dittoColors()) + geom_smooth(method = "loess", se = F, span = 2,size=2,
                                                 aes(color = merged_wta2$orig.ident))

p3 = FeatureScatter(merged_wta2, feature1 = "Area_num", feature2 = "MRC1", pt.size = 2,group.by = 'orig.ident',
               cols=dittoColors()) + geom_smooth(method = "loess", se = F, span = 2,size=2,
                                                 aes(color = merged_wta2$orig.ident))

p4 = FeatureScatter(merged_wta2, feature1 = "Area_num", feature2 = "MMP9", pt.size = 2,group.by = 'orig.ident',
               cols=dittoColors()) + geom_smooth(method = "loess", se = F, span = 2,size=2,
                                                 aes(color = merged_wta2$orig.ident))
pdf("nano_immune_markers.pdf", height = 12, width = 14)
p1+p4+p2+p3
dev.off()
```


# Plotting for the paper
```{r}
hsamgi <- buildMSIGDB(species="human",keytype="SYMBOL",anntype="GO")
hsakom <- buildAnnot(species = "human",keytype="SYMBOL", anntype = "KEGGM")

setwd("~/Aneu_paper")

id = unique(Idents(merged))
id = id[id !="LQ"]
merged2 = subset(merged, idents=id)
merged2$clusters[merged2$clusters == "FB3"] = "FBMC3"
merged2$clusters[merged2$clusters == "Artery"] = "EC1"
merged2$clusters[merged2$clusters == "Venule"] = "EC2"
merged2$clusters[merged2$clusters == "Venous"] = "EC3"
Idents(merged2) = merged2$clusters

merged2@assays$RNA@data <- shifted_log_transform(merged2@assays$RNA@counts, 
                                                overdispersion = 0.01, on_disk = TRUE)


saveRDS(merged2,file="/media/chang/HDD-1/chang/fastqs/aneu/merged.rds")

pdf("UMAP_merged.pdf",width = 12)
DimPlot_scCustom(merged2, label=F,group.by = 'clusters', raster = F, pt.size = .1) + DimPlot_scCustom(merged2, group.by = 'Condition', colors_use = c("red","lightgray"),raster = F, pt.size = .1) & NoAxes() & NoLegend()
dev.off()

pdf("UMAP_legend_merged.pdf",width = 12)
DimPlot_scCustom(merged2, label=F, group.by = 'clusters', raster = T, pt.size = .2) + DimPlot_scCustom(merged2, group.by = 'Condition', colors_use = c("red","lightgray"),raster = T, pt.size = .2) & NoAxes()
dev.off()


pdf("bar_clusters_condition.pdf")
dittoBarPlot(merged2, var='Condition', group.by = 'clusters', color.panel = c("red","lightgray"))
dev.off()

write.table(data.frame(rbind(table(merged2$Sample, merged2$clusters))), sep=',',
            "sample_cluster_cell_count.csv",row.names =TRUE, quote=F)

markers = wilcoxauc(merged2, seurat_assay = 'SCT',group_by = "clusters")
top5 = top_markers(markers, n = 3)
top5 = unlist(top5)
top5 = unique(top5[4:length(top5)])
top5 = top5[!grepl("RPS",top5)]
top5 = top5[!grepl("RPL",top5)]
top5 = top5[!grepl("ENSG",top5)]
top5 = top5[!grepl("HSPA",top5)]
top5 = top5[!grepl("HLA",top5)]

pdf('dotplot.pdf', width = 14, height = 10)
top5 = c("DCN",'LUM',"KCNJ8",'ACTA2','CNN1','LYZ','MRC1','F13A1','CD1C','CD8A','CD8B','KCNT2',
         'HIGD1B','CLDN5','ARL15','MECOM','VWF','MFSD2A','IGKC','PTPRC','MMP9','ACP5',"CD3D",
         'SPP1','GNLY','PLP1','PTPRZ1','NRXN1','GPC6','LAMA2','CCL19','IGFBP5','TSHZ2',"GFAP",
         'HBB','HBA1','HBA2','APOD','MYH11','DLC1','NKG7','CCL5','IGFBP7','COL1A2','OLIG1',"GPNMB",
         'SPARC',"BGN","PTGDS","JCHAIN","ZFHX3",'CXCR4',"CD79A","CCL2","PTPRK","VCAN","P2RY13")

pdf('dotplot.pdf', width = 12, height = 8)

top5 = c("CLDN5",'PODXL',"ABCC9","HIGD1B","CNN1","MYH11",'KCNT2','CCL19',"GPC6",
         "APOD","LAMA2",'COL1A1',"LUM",'POSTN',"CD3D",'CD8A','CD8B','GNLY',"IGLC2",
         "BACH2",'JCHAIN','CD1C','MRC1',"LYVE1","EREG",'LYZ',"ACP5","MMP9",'GFAP',
         "AQP4","SCN7A","CDH19","PLP1","OLIG1","HBB","HBA2")
ord = c("EC1", "EC2", "EC3", "PC", "SMC1","SMC2","SMC3","FBMC1","FBMC2","FBMC3",
        "FB1","FB2","Aneu1","Aneu2","Ly1","Ly2","Ly3","Ly4","BC","pDC","cDC","PVM1",
        "PVM2","PVM3","PVM4",'OC',"AC",'Neu','OL','RBC')
levels(merged2) = ord
DotPlot(merged2, features = top5) + scale_color_viridis_c() + RotatedAxis()
dev.off()

Clustered_DotPlot(merged2, features = top5, k = 9,row_label_size = 13)
DotPlot_scCustom(seurat_object = merged2, features = top5, x_lab_rotate = TRUE)
DotPlot(merged2, features = top5) + scale_color_viridis_c() + RotatedAxis()
dev.off()


merged2$group = merged2$Condition
merged2$sample = merged2$Sample
meta2 = propeller(merged2)
pdf('prop_volcano.pdf', width = 14, height = 10)
EnhancedVolcano(meta2,
    lab = rownames(meta2),
    x = 'Tstatistic',
    y = 'FDR',
    title = 'Aneu versus CTRL',
    pCutoff = .05,
    FCcutoff = 1,
    pointSize = 3.0,
    labSize = 5.0, ylim=c(0,6),xlim=c(-8,8),
    col=c('black', 'black', 'black', 'red3'),
    colAlpha = 1)
dev.off()

pdf("sample_cell_proportion.pdf")
dittoBarPlot(merged2, var='ident' ,group.by = 'Sample', color.panel = 
               DiscretePalette_scCustomize(num_colors = 36, palette = "polychrome"))
dev.off()



pdf("aneu1_2_vln.pdf", height = 5)
test = subset(merged2, idents=c("Aneu1","Aneu2"))

VlnPlot(test, features = c("CNN1","POSTN"), assay = 'RNA', pt.size = 0,cols=dittoColors())
dev.off()


test = subset(merged2, subset=Condition == "Aneu")
test[['percent.mt']] = PercentageFeatureSet(test, pattern = "MT-",assay = 'RNA')
pdf('qc.pdf')
VlnPlot(test, features = c("nCount_RNA","nFeature_RNA", "percent.mt"), 
        ncol = 3, group.by = 'Sample', pt.size = 0, cols=dittoColors())
dev.off()

pdf("barplot_individual_clusters.pdf")
dittoBarPlot(merged2, var='clusters', group.by = 'Sample')
dev.off()

pdf("umap_individuals.pdf")
DimPlot(merged2, group.by = 'Sample', cols = dittoColors()) + NoAxes()
dev.off()



DefaultAssay(merged2) = "SCT"
avg = AverageExpression(merged2, assays = 'SCT', group.by = 'clusters')
avg = avg$SCT[,c("SMC1","SMC2","SMC3","PC","FB1","FB2","FBMC1","FBMC2","FBMC3",
                 "Aneu1","Aneu2")]
cor.matrix = cor(avg)
pdf("heatmap_cor.pdf", width = 10)
pheatmap::pheatmap(cor.matrix, cluster_rows = T, cluster_cols = T, color=viridis(30)) 
dev.off()

fb1.markers = FindMarkers(merged2, ident.1 = "Aneu1", ident.2 = "FB1", only.pos = T)
fb1.markers = fb1.markers[!grepl("RPS", rownames(fb1.markers)),]
fb1.markers = fb1.markers[!grepl("MT-", rownames(fb1.markers)),]
fb1.markers = fb1.markers[!grepl("RPL", rownames(fb1.markers)),]
fb1.markers$gene = rownames(fb1.markers)
output_df <- run_pathfindR(fb1.markers[,c('gene','avg_log2FC','p_val_adj')])

aneu2.markers = FindMarkers(merged2, ident.1 = "Aneu2", ident.2 = "SMC2", only.pos = T)
aneu2.markers = aneu2.markers[!grepl("RPS", rownames(aneu2.markers)),]
aneu2.markers = aneu2.markers[!grepl("MT-", rownames(aneu2.markers)),]
aneu2.markers = aneu2.markers[!grepl("RPL", rownames(aneu2.markers)),]
aneu2.markers$gene = rownames(aneu2.markers)
output_df2 <- run_pathfindR(aneu2.markers[,c('gene','avg_log2FC','p_val_adj')])

combined_df <- combine_pathfindR_results(result_A = output_df, 
                                         result_B = output_df2)
combined_df2 = combined_df[combined_df$combined_p < .000000001,]
combined_results_graph(combined_df2)

aneu.markers = FindMarkers(merged2, ident.1 = "Aneu1", ident.2 = "Aneu2")

aneu1.markers = FindMarkers(merged2, ident.1 = "Aneu1", only.pos = T)
aneu2.markers = FindMarkers(merged2, ident.1 = "Aneu2", only.pos = T)
aneu1.markers = aneu1.markers[aneu1.markers$p_val_adj < .05,]
aneu2.markers = aneu2.markers[aneu2.markers$p_val_adj < .05,]


aneu.markers = FindMarkers(merged2, ident.1 = "Pos", ident.2 = "Neg", 
                           subset.ident = "Aneu1", group.by = 'Scissor')
aneu.markers$gene = rownames(aneu.markers)
aneu.markers = aneu.markers[!grepl("RPS", rownames(aneu.markers)),]
aneu.markers = aneu.markers[!grepl("MT-", rownames(aneu.markers)),]
aneu.markers = aneu.markers[!grepl("RPL", rownames(aneu.markers)),]

resgo <- richGO(rownames(aneu.markers[aneu.markers$avg_log2FC > 0.25 & aneu.markers$p_val_adj < .01,]),godata = hsamgi,ontology ="BP")
resgo2 <- richGO(rownames(aneu.markers[aneu.markers$avg_log2FC < -.25 & aneu.markers$p_val_adj < .01,]),godata = hsamgi,ontology ="BP")

res <- compareResult(list(Aneu1=resgo,Aneu2=resgo2))
res2 = res[res$Padj < .000001,]

pdf("aneu_dotplot.pdf",width = 12, height = 10)
comparedot(res2,alpha = 1,low = 'lightgrey', high = 'red')
dev.off()

output_df3 <- run_pathfindR(aneu.markers[,c('gene','avg_log2FC','p_val_adj')])
EnhancedVolcano(aneu.markers,
    lab = rownames(aneu.markers),
    x = 'avg_log2FC',
    y = 'p_val_adj',
    title = 'Aneu1 versus Aneu2',
    pCutoff = 10e-50,
    FCcutoff = 1,xlim=c(-3,3),
    pointSize = 3.0,
    labSize = 6.0,
    col=c('black', 'black', 'black', 'red3'),
    colAlpha = 1)


DimPlot_scCustom(merged2, group.by = 'Scissor', pt.size = .2, 
                 colors_use = c("lightgrey","orange", "red")) + NoAxes()
pdf("scissor_barplot.pdf")
dittoBarPlot(merged2, var='Scissor', group.by = 'clusters', 
             color.panel  =  c("lightgrey","orange", "red"))
dev.off()

numbers <- length(infos$Scissor_pos) + length(infos$Scissor_neg)
result1 <- reliability.test(X, Y, network, alpha = 0.05, family = "cox", cell_num = numbers, n = 10, nfold = 10)


evaluate_summary = evaluate.cell('/media/chang/HDD-1/chang/fastqs/aneu/Scissor.RData', 
                                  infos, FDR = 0.05, bootstrap_n = 100)
evaluate_summary[1:5,1:4]

meta3 = merged@meta.data[rownames(evaluate_summary),]
evaluate_summary$clusters = meta3$clusters
evaluate_summary$Coefficient= evaluate_summary$Coefficient * -1
colnames(evaluate_summary)[[1]] = "Mean"
colnames(evaluate_summary)[[4]] = "Sig"
colnames(evaluate_summary)[[11]] = "Probability"
polychrome_pal <- DiscretePalette_scCustomize(num_colors = 30, palette = "polychrome")

pdf("scissors_stats.pdf")
ggplot(evaluate_summary, aes(x = Mean, y = Coefficient, color=clusters)) +geom_point() + theme_bw() + RotatedAxis() +scale_color_manual(values=polychrome_pal) + theme(text = element_text(size=20))
dev.off()


merged$Scissor2 = merged$Scissor
merged$Scissor2[merged$Scissor2 == "BG"] = "Neg"
aneu.upset = FindMarkers(merged, subset.ident = 'Aneu1', group.by = "Scissor2", 
                   ident.1 = "Pos", ident.2 = 'Neg')
cdc.upset = FindMarkers(merged, subset.ident = 'cDC', group.by = "Scissor2", 
                   ident.1 = "Pos", ident.2 = 'Neg')
fb1.upset = FindMarkers(merged, subset.ident = 'FB1', group.by = "Scissor2", 
                   ident.1 = "Pos", ident.2 = 'Neg')
fb2.upset = FindMarkers(merged, subset.ident = 'FB2', group.by = "Scissor2", 
                   ident.1 = "Pos", ident.2 = 'Neg')
oc.upset = FindMarkers(merged, subset.ident = 'OC', group.by = "Scissor2", 
                   ident.1 = "Pos", ident.2 = 'Neg')
pvm1.upset = FindMarkers(merged, subset.ident = 'PVM1', group.by = "Scissor2", 
                   ident.1 = "Pos", ident.2 = 'Neg')
pvm2.upset = FindMarkers(merged, subset.ident = 'PVM2', group.by = "Scissor2", 
                   ident.1 = "Pos", ident.2 = 'Neg')
pvm3.upset = FindMarkers(merged, subset.ident = 'PVM3', group.by = "Scissor2", 
                   ident.1 = "Pos", ident.2 = 'Neg')
pvm4.upset = FindMarkers(merged, subset.ident = 'PVM4', group.by = "Scissor2", 
                   ident.1 = "Pos", ident.2 = 'Neg')
rbc.upset = FindMarkers(merged, subset.ident = 'RBC', group.by = "Scissor2", 
                   ident.1 = "Pos", ident.2 = 'Neg')
listInput <- list(Aneu = rownames(aneu.upset), cDC = rownames(cdc.upset), FB1 = rownames(fb1.upset),
                  FB2=rownames(fb2.upset),OC=rownames(oc.upset),PVM1 = rownames(pvm1.upset),
                  PVM2=rownames(pvm2.upset),PVM3=rownames(pvm3.upset),PVM4=rownames(pvm4.upset))

pdf("deg_scissor_upset.pdf",width = 10)
upset(fromList(listInput), order.by = "freq",nsets = length(listInput), text.scale =2, nintersects = 22)
dev.off()

EnhancedVolcano(aneu,
    lab = rownames(aneu),
    x = 'avg_log2FC',
    y = 'p_val_adj',
    title = 'Pos versus Neg',
    pCutoff = 1e-5,
    FCcutoff = .5,
    pointSize = 3.0, ylim=c(0,80),
    labSize = 6.0,xlim=c(-2,1),
    col=c('black', 'black', 'black', 'red3'),
    colAlpha = 1)

pdf("cellchat_aneu.pdf", height = 10, width = 5)
rankNet(cellchat, mode = "comparison", stacked = T, do.stat = TRUE,color.use = c("Red",'lightgrey'))
dev.off()

pdf("cellchat_aneu_scissor.pdf", height = 9, width = 6)
rankNet(cellchat2, mode = "comparison", stacked = T, do.stat = TRUE,color.use = c("Red",'lightgrey'), cutoff.pvalue = .2,font.size = 15)
dev.off()

pdf("postn_cellchat.pdf")
pathways.show <- c("PERIOSTIN") 
weight.max <- getMaxWeight(object.list, slot.name = c("netP"), 
                           attribute = pathways.show) 
par(mfrow = c(1,1), xpd=TRUE)
for (i in 1:length(object.list)) {
  netVisual_aggregate(object.list[[i]], signaling = pathways.show, layout = "circle", 
                      edge.width.max = 10, signaling.name = paste(pathways.show, names(object.list)[i]))
}
dev.off()


scissor_markers = FindMarkers(merged, group.by = 'Scissor', ident.1 = "Pos", ident.2 = "Neg")
EnhancedVolcano(scissor_markers,
    lab = rownames(scissor_markers),
    x = 'avg_log2FC',
    y = 'p_val_adj',
    title = 'Pos versus Neg',
    pCutoff = 1e-5,
    FCcutoff = .5,
    pointSize = 3.0, ylim=c(0,80),
    labSize = 6.0,xlim=c(-2,1),
    col=c('black', 'black', 'black', 'red3'),
    colAlpha = 1)

m_df<- msigdbr(species = "Homo sapiens", category = "H")
fgsea_sets<- m_df %>% split(x = .$gene_symbol, f = .$gs_name)
#oc.upset$score = oc.upset$avg_log2FC * -log(oc.upset$p_val_adj)
ranks<- unlist(scissor_markers$avg_log2FC)
names(ranks) = rownames(scissor_markers)
fgseaRes<- fgsea(fgsea_sets, stats = ranks, nperm = 10000)
fgseaResTidy <- fgseaRes %>%
  as_tibble() %>%
  arrange(desc(NES))
pdf("GSEA_scissor.pdf")
ggplot(fgseaResTidy %>% filter(padj < 0.05) %>% head(n= 20), aes(reorder(pathway, NES), NES)) +
    geom_col(aes(fill= NES < 0)) +
    coord_flip() +
    labs(x="Pathway", y="Normalized Enrichment Score",
         title="Hallmark pathways NES from GSEA") + 
    theme_minimal()
dev.off()

pdf('complement.pdf')
pathways.show <- c("COMPLEMENT") 
weight.max <- getMaxWeight(object.list2, slot.name = c("netP"), 
                           attribute = pathways.show) 
par(mfrow = c(1,1), xpd=TRUE)
for (i in 1:length(object.list)) {
  netVisual_aggregate(object.list2[[i]], signaling = pathways.show, layout = "circle", 
                      edge.width.max = 10, signaling.name = paste(pathways.show, names(object.list2)[i]))
}
dev.off()

pdf("interaction_rupture.pdf", width = 10, height = 5)
num.link <- sapply(object.list2, function(x) {rowSums(x@net$count) + colSums(x@net$count)-diag(x@net$count)})
weight.MinMax <- c(min(num.link), max(num.link)) # control the dot size in the different datasets
gg <- list()
for (i in 1:length(object.list2)) {
  gg[[i]] <- netAnalysis_signalingRole_scatter(object.list2[[i]], title = names(object.list2)[i], weight.MinMax = weight.MinMax) + ylim(0,7) + xlim(0,12)
}
patchwork::wrap_plots(plots = gg)
dev.off()


avg = avg$SCT[!grepl("ENSG", rownames(avg$SCT)),]
write.table(avg, "/media/chang/HDD-1/chang/fastqs/aneu/spatial/mean_exp.csv",sep=',', quote=F)



c("EC1","EC2","EC3","PC","SMC1","SMC2","SMC3","FBMC1","FBMC2","FBMC3","FB1","FB2",
  "BC","cDC","pDC","Ly1","Ly2","Ly3","Ly4","PVM1","PVM2","PVM4","PVM3","OC","AC",
  "Neur","OL")


mono1 = subset(merged2, idents=c("FB1","FB2","Aneu1"))
mono2 = subset(merged2, idents=c("SMC2","SMC3","PC","Aneu2"))
mono3 = subset(merged2, idents=c("SMC1","FBMC1","FBMC3"))

cds1 <- as.cell_data_set(mono1)
cds1 <- cluster_cells(cds = cds1, reduction_method = "UMAP")
cds1 <- learn_graph(cds1, use_partition = TRUE)

cds2 <- as.cell_data_set(mono2)
cds2 <- cluster_cells(cds = cds2, reduction_method = "UMAP")
cds2 <- learn_graph(cds2, use_partition = TRUE)

cds3 <- as.cell_data_set(mono3)
cds3 <- cluster_cells(cds = cds3, reduction_method = "UMAP")
cds3 <- learn_graph(cds3, use_partition = TRUE)

s1 = CellSelector(DimPlot(mono1))[[1]]
s2 = CellSelector(DimPlot(mono2))[[1]]
s3 = CellSelector(DimPlot(mono3))[[1]]

cds1 <- order_cells(cds1, reduction_method = "UMAP", root_cells = s1)
cds2 <- order_cells(cds2, reduction_method = "UMAP", root_cells = s2)
cds3 <- order_cells(cds3, reduction_method = "UMAP", root_cells = s3)

plot_cells(
  cds = cds1,
  color_cells_by = "pseudotime",
  show_trajectory_graph = TRUE
)

merged2 <- AddMetaData(
  object = merged2,
  metadata = cds1@principal_graph_aux@listData$UMAP$pseudotime,
  col.name = "Aneu1_Monocle"
)
merged2 <- AddMetaData(
  object = merged2,
  metadata = cds2@principal_graph_aux@listData$UMAP$pseudotime,
  col.name = "Aneu2_Monocle"
)
merged2 <- AddMetaData(
  object = merged2,
  metadata = cds3@principal_graph_aux@listData$UMAP$pseudotime,
  col.name = "SMC_Monocle"
)

FeaturePlot(merged2, c("Aneu1_Monocle", "Aneu2_Monocle","SMC_Monocle"), 
            pt.size = 0.1) & scale_color_viridis()


merged.monocle = subset(merged2, idents=c("FB1","FB2","Aneu1","SMC2",
                                          "SMC3","PC","Aneu2"))
FeaturePlot(merged.monocle, c("Aneu1_Monocle","Aneu2_Monocle"), 
            pt.size = 0.1) & scale_color_viridis()

merged.monocle$Monocle = merged.monocle$Aneu1_Monocle
merged.monocle$Monocle[is.na(merged.monocle$Aneu1_Monocle)] = merged.monocle$Aneu2_Monocle[!is.na(merged.monocle$Aneu2_Monocle)]

cells3 = CellSelector(FeaturePlot(merged.monocle, c("Monocle"), order=T, max.cutoff = 'q99',
            pt.size = 0.1) & scale_color_viridis())
merged.monocle2 = subset(merged.monocle, cells=cells3)

FeaturePlot(merged.monocle2, c("Monocle"), order=T, max.cutoff = 'q99',
            pt.size = 1) & scale_color_viridis()


pdf("monocle_pseudotime2.pdf", width = 14)
p1 = FeatureScatter(merged.monocle, feature1 = "Monocle", feature2 = "SERPINF1", pt.size = .5, 
               cols=dittoColors()) + geom_smooth(method = "loess", se = F, span = 2,size=2,
                                                 aes(color = merged.monocle$clusters))

p2 = FeatureScatter(merged.monocle, feature1 = "Monocle", feature2 = "SFRP2", pt.size = .5, 
               cols=dittoColors()) + geom_smooth(method = "loess", se = F, span = 2,size=2,
                                                 aes(color = merged.monocle$clusters))
p1+p2
dev.off()

pdf("monocle_umap.pdf")
DimPlot(merged.monocle2, label=F, pt.size = 2, cols=dittoColors()) + NoAxes()
dev.off()

dittoDimPlot(merged.monocle2, trajectory.cluster.meta = "clusters",
             add.trajectory.lineages = list(c("FB1","Aneu1"),c("SMC2","Aneu2"), c("FB1","FB2")), 
             var='Monocle') + scale_color_viridis() + NoAxes()

plot_genes_in_pseudotime(cds1,
                         color_cells_by="pseudotime",
                         min_expr=0.5)

gene_fits <- fit_models(cds1, model_formula_str = "~pseudotime")
fit_coefs <- coefficient_table(gene_fits)
emb_time_terms <- fit_coefs %>% filter(term == "pseudotime")
emb_time_terms = emb_time_terms %>% filter (q_value < 0.05) %>%
         select(gene_id, term, q_value, estimate)
emb_time_terms = emb_time_terms[order(emb_time_terms$q_value, decreasing = F),]

genes = head(emb_time_terms$gene_id, 100)
genes = genes[!grepl("MT-", genes)]
genes = genes[!grepl("RPS", genes)]
genes = genes[!grepl("RPL", genes)]
genes = genes[!grepl("ENSG", genes)]

pt.matrix <- as.matrix(mono1@assays$RNA@data[match(genes,rownames(mono1@assays$RNA@data)),order(pseudotime(cds1))])
pt.matrix <- t(apply(pt.matrix,1,function(x){smooth.spline(x,df=3)$y}))
pt.matrix <- t(apply(pt.matrix,1,function(x){(x-mean(x))/sd(x)}))
rownames(pt.matrix) <- genes;
ht <- Heatmap(
  pt.matrix,
  name                         = "z-score",
  col                          = colorRamp2(seq(from=-2,to=2,length=11),rev(brewer.pal(11, "Spectral"))),
  show_row_names               = TRUE,
  show_column_names            = FALSE,
  row_names_gp                 = gpar(fontsize = 10),
  km = 1,
  row_title_rot                = 0,
  cluster_rows                 = TRUE,
  cluster_row_slices           = FALSE,
  cluster_columns              = FALSE)

pdf("aneu1_monocle_heatmap.pdf", height = 10)
print(ht)
dev.off()

##
gene_fits2 <- fit_models(cds2, model_formula_str = "~pseudotime")
fit_coefs <- coefficient_table(gene_fits2)
emb_time_terms <- fit_coefs %>% filter(term == "pseudotime")
emb_time_terms = emb_time_terms %>% filter (q_value < 0.05) %>%
         select(gene_id, term, q_value, estimate)
emb_time_terms = emb_time_terms[order(emb_time_terms$q_value, decreasing = F),]


genes = head(emb_time_terms$gene_id, 100)
genes = genes[!grepl("MT-", genes)]
genes = genes[!grepl("RPS", genes)]
genes = genes[!grepl("RPL", genes)]
genes = genes[!grepl("ENSG", genes)]

pt.matrix <- as.matrix(mono2@assays$RNA@data[match(genes,rownames(mono2@assays$RNA@data)),order(pseudotime(cds2))])
pt.matrix <- t(apply(pt.matrix,1,function(x){smooth.spline(x,df=3)$y}))
pt.matrix <- t(apply(pt.matrix,1,function(x){(x-mean(x))/sd(x)}))
rownames(pt.matrix) <- genes;
ht <- Heatmap(
  pt.matrix,
  name                         = "z-score",
  col                          = colorRamp2(seq(from=-2,to=2,length=11),rev(brewer.pal(11, "Spectral"))),
  show_row_names               = TRUE,
  show_column_names            = FALSE,
  row_names_gp                 = gpar(fontsize = 10),
  km = 1,
  row_title_rot                = 0,
  cluster_rows                 = TRUE,
  cluster_row_slices           = FALSE,
  cluster_columns              = FALSE)
ht

pdf("aneu2_monocle_heatmap.pdf", height = 10)
print(ht)
dev.off()

pdf("monocle_pseudotime_umap.pdf")
FeaturePlot(merged.monocle2, order=T, max.cutoff = 'q99', features = 'Monocle', pt.size = 2) + NoAxes() + scale_color_viridis()
dev.off()



res =
  merged2 |>
  sccomp_glm( 
   formula_composition = ~ Scissor2, 
    formula_variability = ~ 1, 
    Sample, 
    clusters
  )
```






```{r}
ReadSTAR <- function(data.dir, sample='sample',clean=TRUE, meta) {
  data.dir.genefull = paste0(data.dir, "GeneFull_Ex50pAS/raw/")
  raw = ReadSTARsolo(data.dir.genefull)
  data.dir.filt = paste0(data.dir, "GeneFull_Ex50pAS/filtered/")
  filtered = ReadSTARsolo(data.dir.filt)
  seurat = CreateSeuratObject(filtered, project=sample)
  seurat$Sample = sample
  donor_meta = read.table(meta, sep='\t', header = T)
  seurat@meta.data = cbind(seurat@meta.data, donor_meta)
  if(clean == TRUE){
    sce <- SingleCellExperiment(list(counts = filtered))
    sce.raw <- SingleCellExperiment(list(counts = raw))
    
    sce <- decontX(sce, background = sce.raw)
    seurat[['cleaned']] = CreateAssayObject(round(decontXcounts(sce)))
    seurat[['percent.mt']] = PercentageFeatureSet(seurat, pattern = "MT-",assay = 'cleaned')
    seurat = subset(seurat, subset=nFeature_cleaned >= 500)
  }
  subset(seurat, subset=donor_id %in% c("donor0","donor1","donor2","donor3","donor4","donor5","donor6","unassigned"))
}
male = c("UTY","USP9Y","DDX3Y","RPS4Y1","ZFY","KDM5D","EIF1AY")

b1_p2 = ReadSTAR("/media/chang/HDD-5/chang/10x/b1_p2/B1_P2_Solo.out/", sample="B1_P2",
                 meta="/media/chang/HDD-5/chang/10x/b1_p2/vireo/donor_ids.tsv")
DefaultAssay(b1_p2) = "cleaned"
b1_p2$Sample = "Donor"
b1_p2$Sample[b1_p2$donor_id == "donor0"] = "Ax-12-13"
b1_p2$Sample[b1_p2$donor_id == "donor1"] = "UCSF-2021-009"
b1_p2$Sample[b1_p2$donor_id == "donor2"] = "UCSF-2022-006"

b1_p2 = NormalizeData(b1_p2, assay = "cleaned")
VlnPlot_scCustom(b1_p2, features = c("XIST","UTY","USP9Y","DDX3Y","RPS4Y1","ZFY","KDM5D",
                                     "EIF1AY","POSTN","CNN1","TAGLN","ACTA2"),
                                      num_columns = 4, split.by = "donor_id")


b1_p3 = ReadSTAR("/media/chang/HDD-5/chang/10x/b1_p3/B1_P3_Solo.out/", sample="B1_P3",
                 meta="/media/chang/HDD-5/chang/10x/b1_p3/vireo/donor_ids.tsv")
DefaultAssay(b1_p3) = "cleaned"
b1_p3$Sample = "Donor"
b1_p3$Sample[b1_p3$donor_id == "donor0"] = "TB045"
b1_p3$Sample[b1_p3$donor_id == "donor1"] = "Ax-12-13"
b1_p3$Sample[b1_p3$donor_id == "donor2"] = "Ax-12-1"
b1_p3 = NormalizeData(b1_p3, assay = "cleaned")
VlnPlot_scCustom(b1_p3, features = c("XIST","UTY","USP9Y","DDX3Y","RPS4Y1","ZFY","KDM5D",
                                     "EIF1AY","POSTN","CNN1","TAGLN","ACTA2"),
                                      num_columns = 4, split.by = "donor_id")
FeatureScatter(b1_p3, feature1 = "POSTN", feature2 = "CNN1", group.by = "donor_id", cols=dittoColors())

b1_p4 = ReadSTAR("/media/chang/HDD-5/chang/10x/b1_p4/B1_P4_Solo.out/", sample="B1_P4",
                 meta="/media/chang/HDD-5/chang/10x/b1_p4/vireo/donor_ids.tsv")
DefaultAssay(b1_p4) = "cleaned"
b1_p4$Sample = "Donor"
b1_p4$Sample[b1_p4$donor_id == "donor0"] = "UCSF-2023-003"
b1_p4$Sample[b1_p4$donor_id == "donor1"] = "Ax-12-13"
b1_p4$Sample[b1_p4$donor_id == "donor2"] = "Ax-3623"
b1_p4$Sample[b1_p4$donor_id == "donor3"] = "Ax-1-10"

b1_p4 = NormalizeData(b1_p4, assay = "cleaned")
VlnPlot_scCustom(b1_p4, features = c("XIST","UTY","USP9Y","DDX3Y","RPS4Y1","ZFY","KDM5D",
                                     "EIF1AY","POSTN","CNN1","TAGLN","ACTA2"),
                                      num_columns = 4, split.by = "donor_id")
FeatureScatter(b1_p4, feature1 = "POSTN", feature2 = "CNN1", group.by = "donor_id", cols=dittoColors())

b1_p5 = ReadSTAR("/media/chang/HDD-5/chang/10x/b1_p5/B1_P5_Solo.out/", sample="B1_P5",
                 meta="/media/chang/HDD-5/chang/10x/b1_p5/vireo/donor_ids.tsv")
DefaultAssay(b1_p5) = "cleaned"
b1_p5$Sample = "Donor"
b1_p5$Sample[b1_p5$donor_id == "donor0"] = "UCSF-2023-003"
b1_p5$Sample[b1_p5$donor_id == "donor1"] = "UCSF-2022-006"
b1_p5 = NormalizeData(b1_p5, assay = "cleaned")
b1_p5 = AddModuleScore_UCell(b1_p5, features = list(Male=male), assay = 'cleaned')
VlnPlot_scCustom(b1_p5, features = c("POSTN","CNN1"), split.by = "donor_id")
FeatureScatter(b1_p5, feature1 = "POSTN", feature2 = "CNN1", group.by = "donor_id", cols=dittoColors())

b1_p6 = ReadSTAR("/media/chang/HDD-5/chang/10x/b1_p6/B1_P6_Solo.out/", sample="B1_P6",
                 meta="/media/chang/HDD-5/chang/10x/b1_p6/vireo/donor_ids.tsv")
DefaultAssay(b1_p6) = "cleaned"
b1_p6$Sample = "Donor"
b1_p6$Sample[b1_p6$donor_id == "donor0"] = "UCSF-2021-009"
b1_p6$Sample[b1_p6$donor_id == "donor1"] = "Ax-12-13"
b1_p6$Sample[b1_p6$donor_id == "donor2"] = "Ax-1-10"

b1_p6 = NormalizeData(b1_p6, assay = "cleaned")
b1_p6 = FindVariableFeatures(b1_p6)
b1_p6 = ScaleData(b1_p6)
b1_p6 = RunPCA(b1_p6)
b1_p6 = RunUMAP(b1_p6, dims=1:10)
FeaturePlot(b1_p6, features = c("POSTN","CNN1"), order=T, max.cutoff = 'q99')& NoAxes()
cells = CellSelector(FeaturePlot(b1_p6, features = c("CNN1"), order=T, max.cutoff = 'q99'))
VlnPlot_scCustom(b1_p6, features = c("POSTN","CNN1","XIST"), split.by = "donor_id")


b2_p2 = ReadSTAR("/media/chang/HDD-5/chang/10x/b2_p2/B2_P2_Solo.out/", sample="B2_P2",
                 meta="/media/chang/HDD-5/chang/10x/b2_p2/vireo/donor_ids.tsv")
DefaultAssay(b2_p2) = "cleaned"
b2_p2 = NormalizeData(b2_p2, assay = "cleaned")
b2_p2 = AddModuleScore_UCell(b2_p2, features = list(Male=male), assay = 'cleaned')
VlnPlot_scCustom(b2_p2, features = c("XIST","UTY","USP9Y","DDX3Y","RPS4Y1","ZFY","KDM5D",
                                     "EIF1AY","POSTN","CNN1","TAGLN","Male_UCell"),
                                      num_columns = 4, split.by = "donor_id")
b2_p2_male = subset(b2_p2, subset=Male_UCell > 0.05)
b2_p2_male$Sample = "UCSF-2023-003"

b2_p3 = ReadSTAR("/media/chang/HDD-5/chang/10x/b2_p3/B2_P3_Solo.out/", sample="B2_P3",
                 meta="/media/chang/HDD-5/chang/10x/b2_p3/vireo/donor_ids.tsv")
DefaultAssay(b2_p3) = "cleaned"
b2_p3 = NormalizeData(b2_p3, assay = "cleaned")
b2_p3 = AddModuleScore_UCell(b2_p3, features = list(Male=male), assay = 'cleaned')
VlnPlot_scCustom(b2_p3, features = c("POSTN","CNN1","XIST"), split.by = "donor_id")
b2_p3_male = subset(b2_p3, subset=Male_UCell > 0.05)
b2_p3_male$Sample = "UCSF-2023-003"

b2_p4 = ReadSTAR("/media/chang/HDD-5/chang/10x/b2_p4/B2_P4_Solo.out/", sample="B2_P4",
                 meta="/media/chang/HDD-5/chang/10x/b2_p4/vireo/donor_ids.tsv")
DefaultAssay(b2_p4) = "cleaned"
b2_p4$Sample = "Donor"
b2_p4$Sample[b2_p4$donor_id == "donor5"] = "Ax1116"
b2_p4$Sample[b2_p4$donor_id == "donor1"] = "Ax4323"
b2_p4 = NormalizeData(b2_p4, assay = "cleaned")
b2_p4 = AddModuleScore_UCell(b2_p4, features = list(Male=male), assay = 'cleaned')
VlnPlot_scCustom(b2_p4, features = c("XIST","UTY","USP9Y","DDX3Y","RPS4Y1","ZFY","KDM5D",
                                     "EIF1AY","POSTN","CNN1","TAGLN","ACTA2"),
                                      num_columns = 4, split.by = "donor_id")
b2_p4_male = subset(b2_p4, subset=Male_UCell > 0.05)
b2_p4$Sample[colnames(b2_p4_male)] = "UCSF-2023-003"

b2_p6 = ReadSTAR("/media/chang/HDD-5/chang/10x/b2_p6/B2_P6_Solo.out/", sample="B2_P6",
                 meta="/media/chang/HDD-5/chang/10x/b2_p6/vireo/donor_ids.tsv")
DefaultAssay(b2_p6) = "cleaned"
b2_p6$Sample = "Donor"
b2_p6$Sample[b2_p6$donor_id == "donor1"] = "Ax2623_TB039"
b2_p6 = NormalizeData(b2_p6, assay = "cleaned")
b2_p6 = AddModuleScore_UCell(b2_p6, features = list(Male=male), assay = 'cleaned')
VlnPlot_scCustom(b2_p6, features = c("POSTN","CNN1","XIST"), split.by = "donor_id")
b2_p6_male = subset(b2_p6, subset=Male_UCell > 0.05)
b2_p6$Sample[colnames(b2_p6_male)] = "UCSF-2023-003"

merged_aneu = merge(b1_p2,c(b1_p3,b1_p4,b1_p5,b1_p6,
                            b2_p2_male,b2_p3_male,b2_p4,b2_p6)) 
merged_aneu = subset(merged_aneu, subset=Sample != "Donor")

merged_aneu$Dx = "Ruptured"
merged_aneu$Dx[grepl("UCSF", merged_aneu$Sample)] = "CTRL"
merged_aneu$Dx[grepl("Ax1116", merged_aneu$Sample)] = "Unrupted_Unstable"
merged_aneu$Dx[grepl("Ax-12-13", merged_aneu$Sample)] = "Unrupted_Unstable"
merged_aneu$Dx[grepl("TB045", merged_aneu$Sample)] = "Unrupted_Unstable"
merged_aneu$Dx[grepl("TB039", merged_aneu$Sample)] = "Unrupted_Unstable"
merged_aneu$Dx[grepl("Ax4323", merged_aneu$Sample)] = "Unrupted_Stable"
merged_aneu$Dx[grepl("Ax-3623", merged_aneu$Sample)] = "Unrupted_Stable"
merged_aneu$Dissociation = "Nuclei"
saveRDS(merged_aneu,"/media/chang/HDD-5/chang/10x/aneu.rds")

merged_cell_aneu = readRDS("/media/chang/HDD-10/chang/fastqs/aneu/merged.rds")
merged_cell_aneu$Dx = merged_cell_aneu$Condition
merged_cell_aneu$Condition = NULL
merged_cell_aneu$Dx[merged_cell_aneu$Sample == "a219"] = "Ruptured"
merged_cell_aneu$Dx[merged_cell_aneu$Sample == "a1211"] = "Unruptured_unstable"
merged_cell_aneu$Dx[merged_cell_aneu$Sample == "a1217"] = "Unruptured_unstable"
merged_cell_aneu$Dissociation = "Cell"

merged_aneu = merge(merged_aneu,merged_cell_aneu)
saveRDS(merged_aneu,"/media/chang/HDD-5/chang/10x/aneu_all.rds")

```

```{r}
g0 = Read10X("/media/chang/HDD-5/chang/CK_GLYOXAL-401395031/G0_L1-ds.fec8f193b24f4c869d67f4dba1c9c2be/G0/filtered_matrix/sensitivity_5/")
g3 = Read10X("/media/chang/HDD-5/chang/CK_GLYOXAL-401395031/G3_L1-ds.18c22ab2e90b456a913298bee951b5cc/G3/filtered_matrix/sensitivity_3/")
g4 = Read10X("/media/chang/HDD-5/chang/CK_GLYOXAL-401395031/G4_L1-ds.0f17b8c71e4341799c767240551a831a/G4/filtered_matrix/sensitivity_2/")
g5 = Read10X("/media/chang/HDD-5/chang/CK_GLYOXAL-401395031/G5_L1-ds.d447b3b8959b46e2976da8ac322e9640/G5/filtered_matrix/sensitivity_2/")

g0 = CreateSeuratObject(g0, project = 'G0')
g3 = CreateSeuratObject(g3, project = 'G3')
g4 = CreateSeuratObject(g4, project = 'G4')
g5 = CreateSeuratObject(g5, project = 'G5')

g = merge(g0, c(g3,g4,g5))
g = subset(g, subset=nCount_RNA > 1000 & percent.mt<20)
g = SCTransform(g, vst.flavor='v2')
g = RunPCA(g,npcs = 10)
g = RunUMAP(g, dims=1:10)
g = FindNeighbors(g, dims=1:10)
g = FindClusters(g, algorithm = 2, resolution = .3)
DimPlot(g, group.by = c("seurat_clusters","orig.ident"), label=T)&NoAxes()

g$Fix = "Yes"
g$Fix[g$orig.ident == "G0"] = "No"

g = RunHarmony(g, group.by.vars = "Fix", assay.use = "SCT")
g = RunUMAP(g, dims=1:10, reduction = 'harmony')
g = FindNeighbors(g, dims=1:10, reduction = 'harmony')
g = FindClusters(g, algorithm = 2, resolution = .3)
DimPlot(g, group.by = c("seurat_clusters","orig.ident"), 
        label=T, reduction = 'umap')&NoAxes()

VlnPlot(subset(g, idents= c(4,5)), 
        features = c("nCount_RNA", "nFeature_RNA"))
VlnPlot(g, group.by = "orig.ident", pt.size = 0,
        features = c("nCount_RNA", "nFeature_RNA")) & ylim(700,4000)
VlnPlot(g, group.by = "orig.ident", pt.size = 0,ncol=4,
        features = c("percent.mt", "percent.rpl","percent.rps","percent.cox")) 
g[["percent.mt"]] <- PercentageFeatureSet(g, pattern = "^MT-")
g[["percent.rpl"]] <- PercentageFeatureSet(g, pattern = "^RPL")
g[["percent.rps"]] <- PercentageFeatureSet(g, pattern = "^RPS")
g[["percent.cox"]] <- PercentageFeatureSet(g, pattern = "^COX")

dittoBarPlot(subset(g, subset=orig.ident %in% c("G0","G3")), var='orig.ident', group.by = "ident", scale='count')


test = FindMarkers(g, group.by = 'orig.ident', ident.1 = 'G3', ident.2 = "G0")
EnhancedVolcano(test,x="avg_log2FC",y="p_val_adj", lab=rownames(test),
                xlim = c(-1.5,1.5))
genes =  rownames(g@assays$RNA@counts)[!grepl("^MT-|^RPS|^RPS", rownames(g@assays$RNA@counts))]

g2 = CreateSeuratObject(g@assays$RNA@counts[genes,])
g2$orig.ident = g$orig.ident
VlnPlot(g2, group.by = "orig.ident", pt.size = 0,
        features = c("nCount_RNA", "nFeature_RNA")) & ylim(700,4000)
g2 = subset(g2, subset=nCount_RNA > 1000)
g2 = SCTransform(g2, vst.flavor='v2')
g2 = RunPCA(g2)
g2 = RunUMAP(g2, dims=1:10)
g2 = FindNeighbors(g2, dims=1:10)
g2 = FindClusters(g2, algorithm = 2, resolution = .3)
DimPlot(g2, group.by = c("seurat_clusters","orig.ident"), label=T)&NoAxes()

test2 = FindMarkers(g2, group.by = 'orig.ident', ident.1 = 'G3', ident.2 = "G0")
EnhancedVolcano(test2,x="avg_log2FC",y="p_val_adj", lab=rownames(test2),
                xlim = c(-1.5,1.5))

FeaturePlot(g, reduction='umap',
            features = c("GLI3","EOMES","HOPX","SATB2","MKI67","LRRC7","DLX1","SSTR2",
                         "ARX","KIRREL3","CHL1","MEF2C"), 
            order=T,max.cutoff = 'q99') & NoAxes()


markers = wilcoxauc(g2)
top_markers(markers)
```